Hello all!

I just completed my undergrad senior thesis recently and I wanted to
share a little bit with you. All in all, it was a amazing experience! I
had a lot of help with my thesis from my advisors and other students but
it was a very independent project for me also, and I feel very
privileged to have gotten involved in such a cool field--bioinformatics!

The thesis is titled "Comparison of transcription factor binding sites
from ChIP-seq experiments" and I can make it available on request or
once it's published on my university's website. It contains a lot of
stuff that I have
[already](http://searchvoidstar.tumblr.com/post/46138839023/comparison-of-chip-seq-data)
[talked](http://searchvoidstar.tumblr.com/post/44474299692/analysis-of-chip-seq-data)
about on this blog already, but a lot of new stuff too!

For my final project, one of the main things that I did was to use the
limma package from Bioconductor as a way to compare ChIP-seq data. My
original goal was to use some standard t-tests to do comparisons, but I
found digging into the literature that there are improvements including
the moderated t-test from the limma package.

As a starting point, I use the binned read counts that are outputted as
wiggle files from MACS, find the bins that are differentially bound by
comparing the entire datasets using limma (Figure 1). This reveals
intuitively that the read counts that are high in one experiment but low
in the other are identified as differential.

![](https://64.media.tumblr.com/04c09c8fd00e773cb6fcaba1f3dbf21c/fd8ac0b514f0d0be-3a/s540x810/d7f54407189d1460ba3f63a63cacd623a789a1b0.png)

**Figure 1. Differential ChIP-seq read counts from two ChIP-seq
experiments that are identified using limma**

The next thing that I did was to merge all the nearby binding sites
together. I used a very simple merge approach using BEDTools by merging
nearby positions only \<20bp apart. It turns out that the merging
performs very well and many nearby binding sites are merged together
from the high-confidence sites that limma identifies (Figure 2)

![](https://64.media.tumblr.com/64c92c49c15639d366e2026afcf63f27/fd8ac0b514f0d0be-3f/s540x810/182d1f4f0619d8e8caf3ef80a6d8be6a7aa801b5.png)

**Figure 2. The ratio of the merged peaks (output) to the inputted bins
(input)** The decreasing ratio is a sign of many nearby bins that are
identified by limma being merged together at different p-value
thresholds.

The next thing I did was inspect the 'signals' to show that these peaks
resemble intuitively differential peaks. Indeed, many of these these
peaks are highly significant and highly differentially bound (Figure 3).
I attribute the success of this approach to the limma package. In my
paper, I also looked at how my method compares with other software such
as MACS and DIME for finding differential binding sites.

![](https://64.media.tumblr.com/fbb195d0b4e3be71f2f43f4516757f8e/fd8ac0b514f0d0be-bf/s540x810/53bbe7e5cb1a2f2150b53974e77deeb40d8a8949.png)

**Figure 3 An example of a differential binding site**, showing a peak
that is reproduced in the replicates from one strain but that is not
present in the  replicates of the other strain.

These differential binding sites appear in many different forms, and one
question is what effect do these binding sites have? These binding sites
essentially regulate the gene expression of nearby genes, so I looked at
the gene expression data that was also made available and assessed the
differential gene expression vs the differential binding (Figure 4).
This method is inspired by plotMM from the [Rcade
package](http://www.bioconductor.org/packages/2.12/bioc/html/Rcade.html)
from Bioconductor. The gene expression data was obtained with help from
the [GEO2R interface](http://www.ncbi.nlm.nih.gov/geo/geo2r/) that is
fairly new interface for GEOquery on the NCBI website!

![](https://64.media.tumblr.com/f4eb4cb3d0f6f029b4a281ebc9e6dbae/fd8ac0b514f0d0be-0f/s540x810/585ae9f846b904b086266c945763ddbc8faf39a1.png)

**Figure 4. The differences in binding sites are associated with changes
to gene expression also. **Comparison of the log ratio of differential
binding sites vs the log ratio of the gene expression data. The lightly
colored points are the gene expression from a random gene, and the dark
colored points are gene expression for the nearest gene to the binding
site

Overall, I was able to really deeply explore one approach for comparing
ChIP-seq data. I think the availability of high-throughput sequencing
data is very cool on the GEO database, and many new types of
bioinformatics tools are needed to help analyze it! I don't think my
work will stop just because I finished my project, and I will hope to
find new and interesting things for the future!

PS: there are some [interesting
comparisons](http://statsandgenomes.wordpress.com/2011/11/23/bioinformatics-seminar-professor-gordon-smyth-variance-models-for-rna-seq/)
of limma and edgeR that I did not consider in my project but which would
be interesting to analyze.

::: {#footer}
[ May 12th, 2013 4:53pm ]{#timestamp} [bioinformatics]{.tag}
[statistics]{.tag} [chip-seq]{.tag} [Rstats]{.tag} [genomics]{.tag}
:::
