<!DOCTYPE html><html><head><meta charSet="utf-8"/><meta name="viewport" content="width=device-width"/><meta name="next-head-count" content="2"/><noscript data-n-css=""></noscript><script defer="" nomodule="" src="/_next/static/chunks/polyfills-a40ef1678bae11e696dba45124eadd70.js"></script><script src="/_next/static/chunks/webpack-7c4c4c05609ee1ee2453.js" defer=""></script><script src="/_next/static/chunks/framework-7915ac2dcad08ccffd1a.js" defer=""></script><script src="/_next/static/chunks/main-d0187c47f4d0d1b19d72.js" defer=""></script><script src="/_next/static/chunks/pages/_app-51f94a4fd678a8c1801c.js" defer=""></script><script src="/_next/static/chunks/pages/2013-02-08-1657c8fa0d38f29d9919.js" defer=""></script><script src="/_next/static/ir3stmELHQRvEOT9d_k92/_buildManifest.js" defer=""></script><script src="/_next/static/ir3stmELHQRvEOT9d_k92/_ssgManifest.js" defer=""></script></head><body><div id="__next"><div><a href="/">Return home</a><div style="margin:0 auto;width:75%"><h1>Data analysis with ShortRead</h1><p>I have wanted to do more exploratory data analysis on my datasets after
taking the Coursera class, Data Analysis. I used a short guide I/0 and
Quality Assessment using ShortRead by martin Morgan. I have already
aligned reads so I can&#x27;t do much quality analysis but I thought the data
was interesting. I also needed to generally try to start using R also!</p><pre><code>opts_knit$set(upload.fun = imgur_upload, base.url = NULL)
opts_chunk$set(fig.width = 5, fig.height = 5, cache = TRUE)
</code></pre><p>Here is the setup for reading ELAND result files</p><pre><code>library(EMBL2011)
library(Biostrings)
library(R.utils)
library(xtable)
library(ShortRead)
rd &lt;- readAligned(&quot;HS959ChIP_rep1u_eland_result.txt&quot;, type = &quot;SolexaResult&quot;)
</code></pre><p>We can calculate a table according to which strand reads are aligned to</p><pre><code>whichStrand &lt;- strand(rd)
table(whichStrand)

## whichStrand
##      +      -      * 
## 530693 517883      0
</code></pre><p>Additionally, a table which gives us the nucleotide frequency</p><pre><code>sr &lt;- sread(rd)
af &lt;- alphabetFrequency(sr, baseOnly = TRUE, collapse = TRUE)
barchart(af)
</code></pre><p><img src="http://i.imgur.com/Gr7TKmi.png" alt="plot of chunk unnamed-chunk-1"/></p><p>The dinucleotide frequency can also show some interesting information.
The prevalence of certain terms shows the primers being overrepresented
here in the AA column</p><pre><code>cs &lt;- colSums(dinucleotideFrequency(sread(rd)))
barchart(cs)
</code></pre><p><img src="http://i.imgur.com/yx3rEbJ.png" alt="plot of chunk din"/></p><p>The GC content calculated as</p><pre><code>(af[&quot;G&quot;] + af[&quot;C&quot;])/(af[&quot;A&quot;] + af[&quot;G&quot;] + af[&quot;C&quot;] + af[&quot;T&quot;])

##      G 
## 0.3842
</code></pre><p>Observing the position of nucleotides in each read can be informative
too. We don&#x27;t have the read quality but these would show the read
quality at each position which could be correlated with the nucleotide
read cycles</p><pre><code>abc &lt;- alphabetByCycle(sread(rd))
abc &lt;- abc[rowSums(abc) != 0, ]
df &lt;- as.data.frame(t(abc[1:4, ]))
xyplot(A + C + G + T ~ 1:nrow(df), df, type = &quot;l&quot;, auto.key = list(x = 0.75, 
    y = 0.95, points = FALSE, lines = TRUE), xlab = &quot;Cycle&quot;, ylab = &quot;Count&quot;)
</code></pre><p><img src="http://i.imgur.com/KEeWT05.png" alt="plot of chunk cycle"/></p><p>\
Note I didn&#x27;t write any of the code here. Refer to
<a href="http://www.bioconductor.org/help/course-materials/2009/SeattleJan09/ShortRead/IOQA.pdf">http://www.bioconductor.org/help/course-materials/2009/SeattleJan09/ShortRead/IOQA.pdf</a></p><p>More to come in future analyses</p><p>::: {#footer}
<!-- -->[ February 8th, 2013 9:50pm ]<!-- -->{#timestamp}
:::</p></div></div></div><script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{}},"page":"/2013-02-08","query":{},"buildId":"ir3stmELHQRvEOT9d_k92","nextExport":true,"autoExport":true,"isFallback":false,"scriptLoader":[]}</script></body></html>