<!DOCTYPE html><html><head><meta charSet="utf-8"/><meta name="viewport" content="width=device-width"/><meta name="next-head-count" content="2"/><link rel="preload" href="/_next/static/css/39537df1dfa69610d89e.css" as="style"/><link rel="stylesheet" href="/_next/static/css/39537df1dfa69610d89e.css" data-n-g=""/><noscript data-n-css=""></noscript><script defer="" nomodule="" src="/_next/static/chunks/polyfills-a40ef1678bae11e696dba45124eadd70.js"></script><script src="/_next/static/chunks/webpack-1a8a258926ecde76681b.js" defer=""></script><script src="/_next/static/chunks/framework-b97a0ed4f13ff8397343.js" defer=""></script><script src="/_next/static/chunks/main-1f2c591c5d3bfcfc95e6.js" defer=""></script><script src="/_next/static/chunks/pages/_app-72a03db9ecdc52193d45.js" defer=""></script><script src="/_next/static/chunks/pages/2016-06-17-4ce1398e3aec2f9caa5b.js" defer=""></script><script src="/_next/static/RBCVMiHHi1f270ktUSjIs/_buildManifest.js" defer=""></script><script src="/_next/static/RBCVMiHHi1f270ktUSjIs/_ssgManifest.js" defer=""></script></head><body><div id="__next"><div><div><a href="/">Return home</a><img src="/avatar.png" style="height:2em;margin-left:1em"/></div><div class="blog"><h1>Querying InterMine databases using R</h1><p>In the past, I had found some ways to do simple queries on InterMine web
services using basic HTTP commands with R (see
<a href="https://gist.github.com/cmdcolin/4758167bdd89e6c9c055">https://gist.github.com/cmdcolin/4758167bdd89e6c9c055</a>)</p><p>However, the InterMineR (<a href="https://github.com/intermine/intermineR">https://github.com/intermine/intermineR</a>)
package automates some of these features and makes it easier to load the
data in R.</p><p><strong>Installation</strong></p><p>One way to install InterMineR is to install from github with
hadley/devtools</p><pre><code>install.packages(&quot;devtools&quot;)
devtools::install_github(&quot;hadley/devtools&quot;)
devtools::install_github(&quot;intermine/intermineR&quot;)
</code></pre><p><strong>Usage</strong></p><p>Basic usage includes loading the &quot;intermine URL&quot; using the initInterMine
function. Then various functions can be called on this result.</p><pre><code>library(InterMineR)
mine=initInterMine(&quot;http://bovinegenome.org/bovinemine/&quot;)
getVersion(mine) #18, intermine API version
getRelease(mine) #1.0, our data release version
getTemplates(mine) # lists all templates on interminer
</code></pre><p><strong>Run a template query</strong></p><p>From the getTemplates function, if you see a template query that you
want to run, you can use the getTemplateQuery function with it&#x27;s name,
and run it with the runQuery function</p><pre><code>getTemplateQuery(mine,&quot;TQ_protein_to_gene&quot;) # see what template looks like
template=getTemplateQuery(mine,&quot;TQ_protein_to_gene&quot;) # save template
runQuery(mine,template) # run the template query with default params, receive data.frame
</code></pre><p>This method is good, but some improvement could be added to change
default parameters in the template query, etc.</p><p><strong>Run query XML</strong></p><p>Another option for running queries is to use the query XML that you can
download from the InterMine query result pages. </p><pre><code> # get all Ensembl genes on chr28 from bovinemine
 query=&#x27;&lt;query model=&quot;genomic&quot; view=&quot;Gene.primaryIdentifier
 Gene.secondaryIdentifier Gene.symbol Gene.name Gene.source
 Gene.organism.shortName Gene.chromosome.primaryIdentifier&quot;
 sortOrder=&quot;Gene.primaryIdentifier ASC&quot; &gt;&lt;constraint
 path=&quot;Gene.organism.shortName&quot; op=&quot;=&quot; value=&quot;B. taurus&quot;
 /&gt;&lt;constraint path=&quot;Gene.chromosome.primaryIdentifier&quot; op=&quot;=&quot;
 value=&quot;GK000028.2&quot; /&gt;&lt;/query&gt;&#x27;

 results=runQuery(mine, query)

 head(results)
</code></pre><p><strong>Conclusion</strong></p><p>The InterMineR package has a couple of nice features for getting
InterMine data with a couple of functions for looking at templates. For
many use cases, copying the Query XML from a InterMine webpage and
pasting that into the runQuery function is sufficient and produces a
data frame that can be analyzed.</p><p>PS it is not easy to post XML on tumblr after editing the post in
markdown mode. You have to add the lt and gt shortcuts and even after
that it gets filtered?!</p><p>::: {#footer}
<!-- -->[ June 17th, 2016 11:29am ]<!-- -->{#timestamp} <!-- -->[intermine]<!-- -->{.tag}
<!-- -->[bioinformatics]<!-- -->{.tag} <!-- -->[R]<!-- -->{.tag} <!-- -->[rstats]<!-- -->{.tag}
:::</p></div></div></div><script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{}},"page":"/2016-06-17","query":{},"buildId":"RBCVMiHHi1f270ktUSjIs","nextExport":true,"autoExport":true,"isFallback":false,"scriptLoader":[]}</script></body></html>