<!DOCTYPE html><html lang="en"><head><meta charSet="utf-8"/><meta name="viewport" content="width=device-width"/><title>Querying InterMine databases using R</title><meta name="next-head-count" content="3"/><link rel="shortcut icon" href="favicon.ico"/><link rel="alternate" type="application/rss+xml" href="/rss.xml"/><meta name="description" content="Blogging for the future"/><script src="https://cdnjs.cloudflare.com/ajax/libs/prism/1.25.0/prism.min.js" integrity="sha512-hpZ5pDCF2bRCweL5WoA0/N1elet1KYL5mx3LP555Eg/0ZguaHawxNvEjF6O3rufAChs16HVNhEc6blF/rZoowQ==" crossorigin="anonymous" referrerPolicy="no-referrer"></script><link rel="preload" href="/_next/static/css/c32b3d2522b06381.css" as="style"/><link rel="stylesheet" href="/_next/static/css/c32b3d2522b06381.css" data-n-g=""/><noscript data-n-css=""></noscript><script defer="" nomodule="" src="/_next/static/chunks/polyfills-c67a75d1b6f99dc8.js"></script><script src="/_next/static/chunks/webpack-1b2cde50635286d1.js" defer=""></script><script src="/_next/static/chunks/framework-c0d8f0fd2eea5ac1.js" defer=""></script><script src="/_next/static/chunks/main-b12cd062888056b2.js" defer=""></script><script src="/_next/static/chunks/pages/_app-d9139de5c3de3b6b.js" defer=""></script><script src="/_next/static/chunks/996-3d2a4318c0ac6f1a.js" defer=""></script><script src="/_next/static/chunks/pages/posts/%5Bslug%5D-b6f09b4db8bfc341.js" defer=""></script><script src="/_next/static/i9vmI0eBx7RPb5Yk124js/_buildManifest.js" defer=""></script><script src="/_next/static/i9vmI0eBx7RPb5Yk124js/_ssgManifest.js" defer=""></script></head><body><div id="__next"><div><main><div><div style="margin-bottom:100px"><a href="/">Misc scribbles</a></div><article><div><h1>Querying InterMine databases using R</h1><h4>2016-06-17</h4></div><div><p>In the past, I had found some ways to do simple queries on InterMine web
services using basic HTTP commands with R (see
<a href="https://gist.github.com/cmdcolin/4758167bdd89e6c9c055">https://gist.github.com/cmdcolin/4758167bdd89e6c9c055</a>)</p>
<p>However, the InterMineR (<a href="https://github.com/intermine/intermineR">https://github.com/intermine/intermineR</a>)
package automates some of these features and makes it easier to load the
data in R.</p>
<p><strong>Installation</strong></p>
<p>One way to install InterMineR is to install from github with
hadley/devtools</p>
<pre><code>install.packages("devtools")
devtools::install_github("hadley/devtools")
devtools::install_github("intermine/intermineR")
</code></pre>
<p><strong>Usage</strong></p>
<p>Basic usage includes loading the "intermine URL" using the initInterMine
function. Then various functions can be called on this result.</p>
<pre><code>library(InterMineR)
mine=initInterMine("http://bovinegenome.org/bovinemine/")
getVersion(mine) #18, intermine API version
getRelease(mine) #1.0, our data release version
getTemplates(mine) # lists all templates on interminer
</code></pre>
<p><strong>Run a template query</strong></p>
<p>From the getTemplates function, if you see a template query that you
want to run, you can use the getTemplateQuery function with it's name,
and run it with the runQuery function</p>
<pre><code>getTemplateQuery(mine,"TQ_protein_to_gene") # see what template looks like
template=getTemplateQuery(mine,"TQ_protein_to_gene") # save template
runQuery(mine,template) # run the template query with default params, receive data.frame
</code></pre>
<p>This method is good, but some improvement could be added to change
default parameters in the template query, etc.</p>
<p><strong>Run query XML</strong></p>
<p>Another option for running queries is to use the query XML that you can
download from the InterMine query result pages.</p>
<pre><code> # get all Ensembl genes on chr28 from bovinemine
 query='&#x3C;query model="genomic" view="Gene.primaryIdentifier
 Gene.secondaryIdentifier Gene.symbol Gene.name Gene.source
 Gene.organism.shortName Gene.chromosome.primaryIdentifier"
 sortOrder="Gene.primaryIdentifier ASC" >&#x3C;constraint
 path="Gene.organism.shortName" op="=" value="B. taurus"
 />&#x3C;constraint path="Gene.chromosome.primaryIdentifier" op="="
 value="GK000028.2" />&#x3C;/query>'

 results=runQuery(mine, query)

 head(results)
</code></pre>
<p><strong>Conclusion</strong></p>
<p>The InterMineR package has a couple of nice features for getting
InterMine data with a couple of functions for looking at templates. For
many use cases, copying the Query XML from a InterMine webpage and
pasting that into the runQuery function is sufficient and produces a
data frame that can be analyzed.</p>
<p>PS it is not easy to post XML on tumblr after editing the post in
markdown mode. You have to add the lt and gt shortcuts and even after
that it gets filtered?!</p></div><div style="margin-top:200px"></div></article></div></main></div><footer style="margin-top:100px"><a href="/">Home</a> <a href="/archive">Blog archive</a> <a href="https://github.com/cmdcolin/">Github</a> <a href="https://twitter.com/cmdcolin">Twitter</a> <a href="/projects">Projects</a> <a href="/photos">Photos</a> <a href="/rss.xml">RSS</a><a href="/about">About</a> </footer></div><script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{"post":{"title":"Querying InterMine databases using R","date":"2016-06-17","slug":"2016-06-17","html":"\u003cp\u003eIn the past, I had found some ways to do simple queries on InterMine web\nservices using basic HTTP commands with R (see\n\u003ca href=\"https://gist.github.com/cmdcolin/4758167bdd89e6c9c055\"\u003ehttps://gist.github.com/cmdcolin/4758167bdd89e6c9c055\u003c/a\u003e)\u003c/p\u003e\n\u003cp\u003eHowever, the InterMineR (\u003ca href=\"https://github.com/intermine/intermineR\"\u003ehttps://github.com/intermine/intermineR\u003c/a\u003e)\npackage automates some of these features and makes it easier to load the\ndata in R.\u003c/p\u003e\n\u003cp\u003e\u003cstrong\u003eInstallation\u003c/strong\u003e\u003c/p\u003e\n\u003cp\u003eOne way to install InterMineR is to install from github with\nhadley/devtools\u003c/p\u003e\n\u003cpre\u003e\u003ccode\u003einstall.packages(\"devtools\")\ndevtools::install_github(\"hadley/devtools\")\ndevtools::install_github(\"intermine/intermineR\")\n\u003c/code\u003e\u003c/pre\u003e\n\u003cp\u003e\u003cstrong\u003eUsage\u003c/strong\u003e\u003c/p\u003e\n\u003cp\u003eBasic usage includes loading the \"intermine URL\" using the initInterMine\nfunction. Then various functions can be called on this result.\u003c/p\u003e\n\u003cpre\u003e\u003ccode\u003elibrary(InterMineR)\nmine=initInterMine(\"http://bovinegenome.org/bovinemine/\")\ngetVersion(mine) #18, intermine API version\ngetRelease(mine) #1.0, our data release version\ngetTemplates(mine) # lists all templates on interminer\n\u003c/code\u003e\u003c/pre\u003e\n\u003cp\u003e\u003cstrong\u003eRun a template query\u003c/strong\u003e\u003c/p\u003e\n\u003cp\u003eFrom the getTemplates function, if you see a template query that you\nwant to run, you can use the getTemplateQuery function with it's name,\nand run it with the runQuery function\u003c/p\u003e\n\u003cpre\u003e\u003ccode\u003egetTemplateQuery(mine,\"TQ_protein_to_gene\") # see what template looks like\ntemplate=getTemplateQuery(mine,\"TQ_protein_to_gene\") # save template\nrunQuery(mine,template) # run the template query with default params, receive data.frame\n\u003c/code\u003e\u003c/pre\u003e\n\u003cp\u003eThis method is good, but some improvement could be added to change\ndefault parameters in the template query, etc.\u003c/p\u003e\n\u003cp\u003e\u003cstrong\u003eRun query XML\u003c/strong\u003e\u003c/p\u003e\n\u003cp\u003eAnother option for running queries is to use the query XML that you can\ndownload from the InterMine query result pages.\u003c/p\u003e\n\u003cpre\u003e\u003ccode\u003e # get all Ensembl genes on chr28 from bovinemine\n query='\u0026#x3C;query model=\"genomic\" view=\"Gene.primaryIdentifier\n Gene.secondaryIdentifier Gene.symbol Gene.name Gene.source\n Gene.organism.shortName Gene.chromosome.primaryIdentifier\"\n sortOrder=\"Gene.primaryIdentifier ASC\" \u003e\u0026#x3C;constraint\n path=\"Gene.organism.shortName\" op=\"=\" value=\"B. taurus\"\n /\u003e\u0026#x3C;constraint path=\"Gene.chromosome.primaryIdentifier\" op=\"=\"\n value=\"GK000028.2\" /\u003e\u0026#x3C;/query\u003e'\n\n results=runQuery(mine, query)\n\n head(results)\n\u003c/code\u003e\u003c/pre\u003e\n\u003cp\u003e\u003cstrong\u003eConclusion\u003c/strong\u003e\u003c/p\u003e\n\u003cp\u003eThe InterMineR package has a couple of nice features for getting\nInterMine data with a couple of functions for looking at templates. For\nmany use cases, copying the Query XML from a InterMine webpage and\npasting that into the runQuery function is sufficient and produces a\ndata frame that can be analyzed.\u003c/p\u003e\n\u003cp\u003ePS it is not easy to post XML on tumblr after editing the post in\nmarkdown mode. You have to add the lt and gt shortcuts and even after\nthat it gets filtered?!\u003c/p\u003e"}},"__N_SSG":true},"page":"/posts/[slug]","query":{"slug":"2016-06-17"},"buildId":"i9vmI0eBx7RPb5Yk124js","isFallback":false,"gsp":true,"scriptLoader":[]}</script></body></html>