<!DOCTYPE html><html lang="en"><head><meta charSet="utf-8"/><meta name="viewport" content="width=device-width"/><title>Querying InterMine databases using R</title><meta name="next-head-count" content="3"/><link rel="shortcut icon" href="favicon.ico"/><link rel="alternate" type="application/rss+xml" href="/rss.xml"/><meta name="description" content="Blogging for the future"/><link rel="preload" href="/_next/static/css/97dc5fe527f5d592.css" as="style"/><link rel="stylesheet" href="/_next/static/css/97dc5fe527f5d592.css" data-n-g=""/><noscript data-n-css=""></noscript><script defer="" nomodule="" src="/_next/static/chunks/polyfills-c67a75d1b6f99dc8.js"></script><script src="/_next/static/chunks/webpack-8c0202d61bc61a66.js" defer=""></script><script src="/_next/static/chunks/framework-46611630e39cfdeb.js" defer=""></script><script src="/_next/static/chunks/main-731659734a610bcb.js" defer=""></script><script src="/_next/static/chunks/pages/_app-577aa7b9aa74ad1d.js" defer=""></script><script src="/_next/static/chunks/996-d948af01b202b8dc.js" defer=""></script><script src="/_next/static/chunks/pages/posts/%5Bslug%5D-59c11154d2baa387.js" defer=""></script><script src="/_next/static/lQtcHFVEFBTvZQE7L4r8A/_buildManifest.js" defer=""></script><script src="/_next/static/lQtcHFVEFBTvZQE7L4r8A/_ssgManifest.js" defer=""></script></head><body><div id="__next"><div><main><div><div style="margin-bottom:100px"><a href="/">Misc scribbles</a></div><article><div><h1>Querying InterMine databases using R</h1><h4>2016-06-17</h4></div><div><p>In the past, I had found some ways to do simple queries on InterMine web
services using basic HTTP commands with R (see
<a href="https://gist.github.com/cmdcolin/4758167bdd89e6c9c055">https://gist.github.com/cmdcolin/4758167bdd89e6c9c055</a>)</p>
<p>However, the InterMineR (<a href="https://github.com/intermine/intermineR">https://github.com/intermine/intermineR</a>) package
automates some of these features and makes it easier to load the data in R.</p>
<h2 id="installation"><a aria-hidden="true" tabindex="-1" href="#installation"><a href="#installation" style="margin-right: 10px">#</a></a>Installation</h2>
<p>One way to install InterMineR is to install from github with hadley/devtools</p>
<pre class="shiki poimandres" style="background-color: #1b1e28"><code><span class="line"><span style="color: #A6ACCD">install.packages(</span><span style="color: #A6ACCD">"</span><span style="color: #5DE4C7">devtools</span><span style="color: #A6ACCD">"</span><span style="color: #A6ACCD">)</span></span>
<span class="line"><span style="color: #A6ACCD">devtools</span><span style="color: #5DE4C7">::</span><span style="color: #E4F0FBD0">install_github</span><span style="color: #A6ACCD">(</span><span style="color: #A6ACCD">"</span><span style="color: #5DE4C7">hadley/devtools</span><span style="color: #A6ACCD">"</span><span style="color: #A6ACCD">)</span></span>
<span class="line"><span style="color: #A6ACCD">devtools</span><span style="color: #5DE4C7">::</span><span style="color: #E4F0FBD0">install_github</span><span style="color: #A6ACCD">(</span><span style="color: #A6ACCD">"</span><span style="color: #5DE4C7">intermine/intermineR</span><span style="color: #A6ACCD">"</span><span style="color: #A6ACCD">)</span></span></code></pre>
<h2 id="usage"><a aria-hidden="true" tabindex="-1" href="#usage"><a href="#usage" style="margin-right: 10px">#</a></a>Usage</h2>
<p>Basic usage includes loading theÂ "intermine URL" using the initInterMine
function. Then various functions can be called on this result.</p>
<pre class="shiki poimandres" style="background-color: #1b1e28"><code><span class="line"><span style="color: #A6ACCD">library(</span><span style="color: #E4F0FB">InterMineR</span><span style="color: #A6ACCD">)</span></span>
<span class="line"><span style="color: #E4F0FB">mine</span><span style="color: #91B4D5">=</span><span style="color: #E4F0FBD0">initInterMine</span><span style="color: #A6ACCD">(</span><span style="color: #A6ACCD">"</span><span style="color: #5DE4C7">http://bovinegenome.org/bovinemine/</span><span style="color: #A6ACCD">"</span><span style="color: #A6ACCD">)</span></span>
<span class="line"><span style="color: #E4F0FBD0">getVersion</span><span style="color: #A6ACCD">(</span><span style="color: #E4F0FB">mine</span><span style="color: #A6ACCD">) </span><span style="color: #767C9DB0; font-style: italic">#18, intermine API version</span></span>
<span class="line"><span style="color: #E4F0FBD0">getRelease</span><span style="color: #A6ACCD">(</span><span style="color: #E4F0FB">mine</span><span style="color: #A6ACCD">) </span><span style="color: #767C9DB0; font-style: italic">#1.0, our data release version</span></span>
<span class="line"><span style="color: #E4F0FBD0">getTemplates</span><span style="color: #A6ACCD">(</span><span style="color: #E4F0FB">mine</span><span style="color: #A6ACCD">) </span><span style="color: #767C9DB0; font-style: italic"># lists all templates on interminer</span></span></code></pre>
<h2 id="run-a-template-query"><a aria-hidden="true" tabindex="-1" href="#run-a-template-query"><a href="#run-a-template-query" style="margin-right: 10px">#</a></a>Run a template query</h2>
<p>From the getTemplates function, if you see a template query that you want to
run, you can use the getTemplateQuery function with it's name, and run it with
the runQuery function</p>
<pre class="shiki poimandres" style="background-color: #1b1e28"><code><span class="line"><span style="color: #E4F0FBD0">getTemplateQuery</span><span style="color: #A6ACCD">(</span><span style="color: #E4F0FB">mine</span><span style="color: #A6ACCD">,</span><span style="color: #A6ACCD">"</span><span style="color: #5DE4C7">TQ_protein_to_gene</span><span style="color: #A6ACCD">"</span><span style="color: #A6ACCD">) </span><span style="color: #767C9DB0; font-style: italic"># see what template looks like</span></span>
<span class="line"><span style="color: #E4F0FB">template</span><span style="color: #91B4D5">=</span><span style="color: #E4F0FBD0">getTemplateQuery</span><span style="color: #A6ACCD">(</span><span style="color: #E4F0FB">mine</span><span style="color: #A6ACCD">,</span><span style="color: #A6ACCD">"</span><span style="color: #5DE4C7">TQ_protein_to_gene</span><span style="color: #A6ACCD">"</span><span style="color: #A6ACCD">) </span><span style="color: #767C9DB0; font-style: italic"># save template</span></span>
<span class="line"><span style="color: #E4F0FBD0">runQuery</span><span style="color: #A6ACCD">(</span><span style="color: #E4F0FB">mine</span><span style="color: #A6ACCD">,</span><span style="color: #E4F0FB">template</span><span style="color: #A6ACCD">) </span><span style="color: #767C9DB0; font-style: italic"># run the template query with default params, receive data.frame</span></span></code></pre>
<p>This method is good, but some improvement could be added to change default
parameters in the template query, etc.</p>
<h2 id="run-query-xml"><a aria-hidden="true" tabindex="-1" href="#run-query-xml"><a href="#run-query-xml" style="margin-right: 10px">#</a></a>Run query XML</h2>
<p>Another option for running queries is to use the query XML that you can download
from the InterMine query result pages.</p>
<pre class="shiki poimandres" style="background-color: #1b1e28"><code><span class="line"><span style="color: #767C9DB0; font-style: italic"># get all Ensembl genes on chr28 from bovinemine</span></span>
<span class="line"><span style="color: #E4F0FB">query</span><span style="color: #91B4D5">=</span><span style="color: #A6ACCD">'</span><span style="color: #5DE4C7">&#x3C;query model="genomic" view="Gene.primaryIdentifier</span></span>
<span class="line"><span style="color: #5DE4C7">Gene.secondaryIdentifier Gene.symbol Gene.name Gene.source</span></span>
<span class="line"><span style="color: #5DE4C7">Gene.organism.shortName Gene.chromosome.primaryIdentifier"</span></span>
<span class="line"><span style="color: #5DE4C7">sortOrder="Gene.primaryIdentifier ASC" >&#x3C;constraint</span></span>
<span class="line"><span style="color: #5DE4C7">path="Gene.organism.shortName" op="=" value="B. taurus"</span></span>
<span class="line"><span style="color: #5DE4C7">/>&#x3C;constraint path="Gene.chromosome.primaryIdentifier" op="="</span></span>
<span class="line"><span style="color: #5DE4C7">value="GK000028.2" />&#x3C;/query></span><span style="color: #A6ACCD">'</span></span>
<span class="line"></span>
<span class="line"><span style="color: #E4F0FB">results</span><span style="color: #91B4D5">=</span><span style="color: #E4F0FBD0">runQuery</span><span style="color: #A6ACCD">(</span><span style="color: #E4F0FB">mine</span><span style="color: #A6ACCD">, </span><span style="color: #E4F0FB">query</span><span style="color: #A6ACCD">)</span></span>
<span class="line"></span>
<span class="line"><span style="color: #A6ACCD">head(</span><span style="color: #E4F0FB">results</span><span style="color: #A6ACCD">)</span></span></code></pre>
<h2 id="conclusion"><a aria-hidden="true" tabindex="-1" href="#conclusion"><a href="#conclusion" style="margin-right: 10px">#</a></a>Conclusion</h2>
<p>The InterMineR package has a couple of nice features for getting InterMine data
with a couple of functions for looking at templates. For many use cases, copying
the Query XML from a InterMine webpage and pasting that into the runQuery
function is sufficient and produces a data frame that can be analyzed.</p>
<p>PS it is not easy to post XML on tumblr after editing the post in markdown mode.
You have to add the lt and gt shortcuts and even after that it gets filtered?!</p></div><div style="margin-top:200px"></div></article></div></main></div><footer style="margin-top:100px"><a href="/">Home</a> <a href="/archive">Blog archive</a> <a href="https://github.com/cmdcolin/">Github</a> <a href="https://twitter.com/cmdcolin">Twitter</a> <a href="/projects">Projects</a><a href="/photos">Photos</a> <a href="/rss.xml">RSS</a><a href="/about">About</a></footer></div><script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{"post":{"title":"Querying InterMine databases using R","date":"2016-06-17","slug":"2016-06-17","html":"\u003cp\u003eIn the past, I had found some ways to do simple queries on InterMine web\nservices using basic HTTP commands with R (see\n\u003ca href=\"https://gist.github.com/cmdcolin/4758167bdd89e6c9c055\"\u003ehttps://gist.github.com/cmdcolin/4758167bdd89e6c9c055\u003c/a\u003e)\u003c/p\u003e\n\u003cp\u003eHowever, the InterMineR (\u003ca href=\"https://github.com/intermine/intermineR\"\u003ehttps://github.com/intermine/intermineR\u003c/a\u003e) package\nautomates some of these features and makes it easier to load the data in R.\u003c/p\u003e\n\u003ch2 id=\"installation\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#installation\"\u003e\u003ca href=\"#installation\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eInstallation\u003c/h2\u003e\n\u003cp\u003eOne way to install InterMineR is to install from github with hadley/devtools\u003c/p\u003e\n\u003cpre class=\"shiki poimandres\" style=\"background-color: #1b1e28\"\u003e\u003ccode\u003e\u003cspan class=\"line\"\u003e\u003cspan style=\"color: #A6ACCD\"\u003einstall.packages(\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e\"\u003c/span\u003e\u003cspan style=\"color: #5DE4C7\"\u003edevtools\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e\"\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e)\u003c/span\u003e\u003c/span\u003e\n\u003cspan class=\"line\"\u003e\u003cspan style=\"color: #A6ACCD\"\u003edevtools\u003c/span\u003e\u003cspan style=\"color: #5DE4C7\"\u003e::\u003c/span\u003e\u003cspan style=\"color: #E4F0FBD0\"\u003einstall_github\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e(\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e\"\u003c/span\u003e\u003cspan style=\"color: #5DE4C7\"\u003ehadley/devtools\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e\"\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e)\u003c/span\u003e\u003c/span\u003e\n\u003cspan class=\"line\"\u003e\u003cspan style=\"color: #A6ACCD\"\u003edevtools\u003c/span\u003e\u003cspan style=\"color: #5DE4C7\"\u003e::\u003c/span\u003e\u003cspan style=\"color: #E4F0FBD0\"\u003einstall_github\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e(\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e\"\u003c/span\u003e\u003cspan style=\"color: #5DE4C7\"\u003eintermine/intermineR\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e\"\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e)\u003c/span\u003e\u003c/span\u003e\u003c/code\u003e\u003c/pre\u003e\n\u003ch2 id=\"usage\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#usage\"\u003e\u003ca href=\"#usage\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eUsage\u003c/h2\u003e\n\u003cp\u003eBasic usage includes loading theÂ \"intermine URL\" using the initInterMine\nfunction. Then various functions can be called on this result.\u003c/p\u003e\n\u003cpre class=\"shiki poimandres\" style=\"background-color: #1b1e28\"\u003e\u003ccode\u003e\u003cspan class=\"line\"\u003e\u003cspan style=\"color: #A6ACCD\"\u003elibrary(\u003c/span\u003e\u003cspan style=\"color: #E4F0FB\"\u003eInterMineR\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e)\u003c/span\u003e\u003c/span\u003e\n\u003cspan class=\"line\"\u003e\u003cspan style=\"color: #E4F0FB\"\u003emine\u003c/span\u003e\u003cspan style=\"color: #91B4D5\"\u003e=\u003c/span\u003e\u003cspan style=\"color: #E4F0FBD0\"\u003einitInterMine\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e(\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e\"\u003c/span\u003e\u003cspan style=\"color: #5DE4C7\"\u003ehttp://bovinegenome.org/bovinemine/\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e\"\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e)\u003c/span\u003e\u003c/span\u003e\n\u003cspan class=\"line\"\u003e\u003cspan style=\"color: #E4F0FBD0\"\u003egetVersion\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e(\u003c/span\u003e\u003cspan style=\"color: #E4F0FB\"\u003emine\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e) \u003c/span\u003e\u003cspan style=\"color: #767C9DB0; font-style: italic\"\u003e#18, intermine API version\u003c/span\u003e\u003c/span\u003e\n\u003cspan class=\"line\"\u003e\u003cspan style=\"color: #E4F0FBD0\"\u003egetRelease\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e(\u003c/span\u003e\u003cspan style=\"color: #E4F0FB\"\u003emine\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e) \u003c/span\u003e\u003cspan style=\"color: #767C9DB0; font-style: italic\"\u003e#1.0, our data release version\u003c/span\u003e\u003c/span\u003e\n\u003cspan class=\"line\"\u003e\u003cspan style=\"color: #E4F0FBD0\"\u003egetTemplates\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e(\u003c/span\u003e\u003cspan style=\"color: #E4F0FB\"\u003emine\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e) \u003c/span\u003e\u003cspan style=\"color: #767C9DB0; font-style: italic\"\u003e# lists all templates on interminer\u003c/span\u003e\u003c/span\u003e\u003c/code\u003e\u003c/pre\u003e\n\u003ch2 id=\"run-a-template-query\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#run-a-template-query\"\u003e\u003ca href=\"#run-a-template-query\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eRun a template query\u003c/h2\u003e\n\u003cp\u003eFrom the getTemplates function, if you see a template query that you want to\nrun, you can use the getTemplateQuery function with it's name, and run it with\nthe runQuery function\u003c/p\u003e\n\u003cpre class=\"shiki poimandres\" style=\"background-color: #1b1e28\"\u003e\u003ccode\u003e\u003cspan class=\"line\"\u003e\u003cspan style=\"color: #E4F0FBD0\"\u003egetTemplateQuery\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e(\u003c/span\u003e\u003cspan style=\"color: #E4F0FB\"\u003emine\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e,\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e\"\u003c/span\u003e\u003cspan style=\"color: #5DE4C7\"\u003eTQ_protein_to_gene\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e\"\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e) \u003c/span\u003e\u003cspan style=\"color: #767C9DB0; font-style: italic\"\u003e# see what template looks like\u003c/span\u003e\u003c/span\u003e\n\u003cspan class=\"line\"\u003e\u003cspan style=\"color: #E4F0FB\"\u003etemplate\u003c/span\u003e\u003cspan style=\"color: #91B4D5\"\u003e=\u003c/span\u003e\u003cspan style=\"color: #E4F0FBD0\"\u003egetTemplateQuery\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e(\u003c/span\u003e\u003cspan style=\"color: #E4F0FB\"\u003emine\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e,\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e\"\u003c/span\u003e\u003cspan style=\"color: #5DE4C7\"\u003eTQ_protein_to_gene\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e\"\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e) \u003c/span\u003e\u003cspan style=\"color: #767C9DB0; font-style: italic\"\u003e# save template\u003c/span\u003e\u003c/span\u003e\n\u003cspan class=\"line\"\u003e\u003cspan style=\"color: #E4F0FBD0\"\u003erunQuery\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e(\u003c/span\u003e\u003cspan style=\"color: #E4F0FB\"\u003emine\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e,\u003c/span\u003e\u003cspan style=\"color: #E4F0FB\"\u003etemplate\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e) \u003c/span\u003e\u003cspan style=\"color: #767C9DB0; font-style: italic\"\u003e# run the template query with default params, receive data.frame\u003c/span\u003e\u003c/span\u003e\u003c/code\u003e\u003c/pre\u003e\n\u003cp\u003eThis method is good, but some improvement could be added to change default\nparameters in the template query, etc.\u003c/p\u003e\n\u003ch2 id=\"run-query-xml\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#run-query-xml\"\u003e\u003ca href=\"#run-query-xml\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eRun query XML\u003c/h2\u003e\n\u003cp\u003eAnother option for running queries is to use the query XML that you can download\nfrom the InterMine query result pages.\u003c/p\u003e\n\u003cpre class=\"shiki poimandres\" style=\"background-color: #1b1e28\"\u003e\u003ccode\u003e\u003cspan class=\"line\"\u003e\u003cspan style=\"color: #767C9DB0; font-style: italic\"\u003e# get all Ensembl genes on chr28 from bovinemine\u003c/span\u003e\u003c/span\u003e\n\u003cspan class=\"line\"\u003e\u003cspan style=\"color: #E4F0FB\"\u003equery\u003c/span\u003e\u003cspan style=\"color: #91B4D5\"\u003e=\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e'\u003c/span\u003e\u003cspan style=\"color: #5DE4C7\"\u003e\u0026#x3C;query model=\"genomic\" view=\"Gene.primaryIdentifier\u003c/span\u003e\u003c/span\u003e\n\u003cspan class=\"line\"\u003e\u003cspan style=\"color: #5DE4C7\"\u003eGene.secondaryIdentifier Gene.symbol Gene.name Gene.source\u003c/span\u003e\u003c/span\u003e\n\u003cspan class=\"line\"\u003e\u003cspan style=\"color: #5DE4C7\"\u003eGene.organism.shortName Gene.chromosome.primaryIdentifier\"\u003c/span\u003e\u003c/span\u003e\n\u003cspan class=\"line\"\u003e\u003cspan style=\"color: #5DE4C7\"\u003esortOrder=\"Gene.primaryIdentifier ASC\" \u003e\u0026#x3C;constraint\u003c/span\u003e\u003c/span\u003e\n\u003cspan class=\"line\"\u003e\u003cspan style=\"color: #5DE4C7\"\u003epath=\"Gene.organism.shortName\" op=\"=\" value=\"B. taurus\"\u003c/span\u003e\u003c/span\u003e\n\u003cspan class=\"line\"\u003e\u003cspan style=\"color: #5DE4C7\"\u003e/\u003e\u0026#x3C;constraint path=\"Gene.chromosome.primaryIdentifier\" op=\"=\"\u003c/span\u003e\u003c/span\u003e\n\u003cspan class=\"line\"\u003e\u003cspan style=\"color: #5DE4C7\"\u003evalue=\"GK000028.2\" /\u003e\u0026#x3C;/query\u003e\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e'\u003c/span\u003e\u003c/span\u003e\n\u003cspan class=\"line\"\u003e\u003c/span\u003e\n\u003cspan class=\"line\"\u003e\u003cspan style=\"color: #E4F0FB\"\u003eresults\u003c/span\u003e\u003cspan style=\"color: #91B4D5\"\u003e=\u003c/span\u003e\u003cspan style=\"color: #E4F0FBD0\"\u003erunQuery\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e(\u003c/span\u003e\u003cspan style=\"color: #E4F0FB\"\u003emine\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e, \u003c/span\u003e\u003cspan style=\"color: #E4F0FB\"\u003equery\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e)\u003c/span\u003e\u003c/span\u003e\n\u003cspan class=\"line\"\u003e\u003c/span\u003e\n\u003cspan class=\"line\"\u003e\u003cspan style=\"color: #A6ACCD\"\u003ehead(\u003c/span\u003e\u003cspan style=\"color: #E4F0FB\"\u003eresults\u003c/span\u003e\u003cspan style=\"color: #A6ACCD\"\u003e)\u003c/span\u003e\u003c/span\u003e\u003c/code\u003e\u003c/pre\u003e\n\u003ch2 id=\"conclusion\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#conclusion\"\u003e\u003ca href=\"#conclusion\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eConclusion\u003c/h2\u003e\n\u003cp\u003eThe InterMineR package has a couple of nice features for getting InterMine data\nwith a couple of functions for looking at templates. For many use cases, copying\nthe Query XML from a InterMine webpage and pasting that into the runQuery\nfunction is sufficient and produces a data frame that can be analyzed.\u003c/p\u003e\n\u003cp\u003ePS it is not easy to post XML on tumblr after editing the post in markdown mode.\nYou have to add the lt and gt shortcuts and even after that it gets filtered?!\u003c/p\u003e"}},"__N_SSG":true},"page":"/posts/[slug]","query":{"slug":"2016-06-17"},"buildId":"lQtcHFVEFBTvZQE7L4r8A","isFallback":false,"gsp":true,"scriptLoader":[]}</script></body></html>