<!DOCTYPE html><html lang="en"><head><meta charSet="utf-8"/><meta name="viewport" content="width=device-width, initial-scale=1"/><link rel="stylesheet" href="/_next/static/css/d19cfa26cd1ef694.css" data-precedence="next"/><link rel="preload" as="script" fetchPriority="low" href="/_next/static/chunks/webpack-dea5254e1907c5df.js"/><script src="/_next/static/chunks/4bd1b696-7cc559629682b47d.js" async=""></script><script src="/_next/static/chunks/215-0c2c111511fa65fd.js" async=""></script><script src="/_next/static/chunks/main-app-d92e72a2e865503d.js" async=""></script><script src="/_next/static/chunks/231-fc74e6e791c0837d.js" async=""></script><script src="/_next/static/chunks/app/page-95d4a906a5a795f0.js" async=""></script><script src="/_next/static/chunks/app/posts/%5Bid%5D/page-e50cd2f34f19f050.js" async=""></script><title>Querying InterMine databases using R</title><meta name="description" content="A blog"/><script src="/_next/static/chunks/polyfills-42372ed130431b0a.js" noModule=""></script></head><body><div class="mb-8"><a href="/">Misc scribbles</a></div><div class="lg:w-1/2"><div><h1>Querying InterMine databases using R</h1><h4>2016-06-17</h4></div><div><p>In the past, I had found some ways to do simple queries on InterMine web
services using basic HTTP commands with R (see
<a href="https://gist.github.com/cmdcolin/4758167bdd89e6c9c055">https://gist.github.com/cmdcolin/4758167bdd89e6c9c055</a>)</p>
<p>However, the InterMineR (<a href="https://github.com/intermine/intermineR">https://github.com/intermine/intermineR</a>) package
automates some of these features and makes it easier to load the data in R.</p>
<h2 id="installation"><a aria-hidden="true" tabindex="-1" href="#installation"><a href="#installation" style="margin-right: 10px">#</a></a>Installation</h2>
<p>One way to install InterMineR is to install from github with hadley/devtools</p>
<figure data-rehype-pretty-code-figure=""><pre style="background-color:#282c34;color:#abb2bf" tabindex="0" data-language="r" data-theme="one-dark-pro"><code data-language="r" data-theme="one-dark-pro" style="display: grid;"><span data-line=""><span style="color:#56B6C2">install.packages</span><span style="color:#ABB2BF">(</span><span style="color:#98C379">"devtools"</span><span style="color:#ABB2BF">)</span></span>
<span data-line=""><span style="color:#ABB2BF">devtools</span><span style="color:#C678DD">::</span><span style="color:#61AFEF">install_github</span><span style="color:#ABB2BF">(</span><span style="color:#98C379">"hadley/devtools"</span><span style="color:#ABB2BF">)</span></span>
<span data-line=""><span style="color:#ABB2BF">devtools</span><span style="color:#C678DD">::</span><span style="color:#61AFEF">install_github</span><span style="color:#ABB2BF">(</span><span style="color:#98C379">"intermine/intermineR"</span><span style="color:#ABB2BF">)</span></span></code></pre></figure>
<h2 id="usage"><a aria-hidden="true" tabindex="-1" href="#usage"><a href="#usage" style="margin-right: 10px">#</a></a>Usage</h2>
<p>Basic usage includes loading the "intermine URL" using the initInterMine
function. Then various functions can be called on this result.</p>
<figure data-rehype-pretty-code-figure=""><pre style="background-color:#282c34;color:#abb2bf" tabindex="0" data-language="r" data-theme="one-dark-pro"><code data-language="r" data-theme="one-dark-pro" style="display: grid;"><span data-line=""><span style="color:#56B6C2">library</span><span style="color:#ABB2BF">(InterMineR)</span></span>
<span data-line=""><span style="color:#E06C75;font-style:italic">mine</span><span style="color:#56B6C2">=</span><span style="color:#61AFEF">initInterMine</span><span style="color:#ABB2BF">(</span><span style="color:#98C379">"http://bovinegenome.org/bovinemine/"</span><span style="color:#ABB2BF">)</span></span>
<span data-line=""><span style="color:#61AFEF">getVersion</span><span style="color:#ABB2BF">(mine) </span><span style="color:#7F848E;font-style:italic">#18, intermine API version</span></span>
<span data-line=""><span style="color:#61AFEF">getRelease</span><span style="color:#ABB2BF">(mine) </span><span style="color:#7F848E;font-style:italic">#1.0, our data release version</span></span>
<span data-line=""><span style="color:#61AFEF">getTemplates</span><span style="color:#ABB2BF">(mine) </span><span style="color:#7F848E;font-style:italic"># lists all templates on interminer</span></span></code></pre></figure>
<h2 id="run-a-template-query"><a aria-hidden="true" tabindex="-1" href="#run-a-template-query"><a href="#run-a-template-query" style="margin-right: 10px">#</a></a>Run a template query</h2>
<p>From the getTemplates function, if you see a template query that you want to
run, you can use the getTemplateQuery function with it's name, and run it with
the runQuery function</p>
<figure data-rehype-pretty-code-figure=""><pre style="background-color:#282c34;color:#abb2bf" tabindex="0" data-language="r" data-theme="one-dark-pro"><code data-language="r" data-theme="one-dark-pro" style="display: grid;"><span data-line=""><span style="color:#61AFEF">getTemplateQuery</span><span style="color:#ABB2BF">(mine,</span><span style="color:#98C379">"TQ_protein_to_gene"</span><span style="color:#ABB2BF">) </span><span style="color:#7F848E;font-style:italic"># see what template looks like</span></span>
<span data-line=""><span style="color:#E06C75;font-style:italic">template</span><span style="color:#56B6C2">=</span><span style="color:#61AFEF">getTemplateQuery</span><span style="color:#ABB2BF">(mine,</span><span style="color:#98C379">"TQ_protein_to_gene"</span><span style="color:#ABB2BF">) </span><span style="color:#7F848E;font-style:italic"># save template</span></span>
<span data-line=""><span style="color:#61AFEF">runQuery</span><span style="color:#ABB2BF">(mine,template) </span><span style="color:#7F848E;font-style:italic"># run the template query with default params, receive data.frame</span></span></code></pre></figure>
<p>This method is good, but some improvement could be added to change default
parameters in the template query, etc.</p>
<h2 id="run-query-xml"><a aria-hidden="true" tabindex="-1" href="#run-query-xml"><a href="#run-query-xml" style="margin-right: 10px">#</a></a>Run query XML</h2>
<p>Another option for running queries is to use the query XML that you can download
from the InterMine query result pages.</p>
<figure data-rehype-pretty-code-figure=""><pre style="background-color:#282c34;color:#abb2bf" tabindex="0" data-language="r" data-theme="one-dark-pro"><code data-language="r" data-theme="one-dark-pro" style="display: grid;"><span data-line=""><span style="color:#7F848E;font-style:italic"># get all Ensembl genes on chr28 from bovinemine</span></span>
<span data-line=""><span style="color:#E06C75;font-style:italic">query</span><span style="color:#56B6C2">=</span><span style="color:#98C379">'&#x3C;query model="genomic" view="Gene.primaryIdentifier</span></span>
<span data-line=""><span style="color:#98C379">Gene.secondaryIdentifier Gene.symbol Gene.name Gene.source</span></span>
<span data-line=""><span style="color:#98C379">Gene.organism.shortName Gene.chromosome.primaryIdentifier"</span></span>
<span data-line=""><span style="color:#98C379">sortOrder="Gene.primaryIdentifier ASC" >&#x3C;constraint</span></span>
<span data-line=""><span style="color:#98C379">path="Gene.organism.shortName" op="=" value="B. taurus"</span></span>
<span data-line=""><span style="color:#98C379">/>&#x3C;constraint path="Gene.chromosome.primaryIdentifier" op="="</span></span>
<span data-line=""><span style="color:#98C379">value="GK000028.2" />&#x3C;/query>'</span></span>
<span data-line=""> </span>
<span data-line=""><span style="color:#E06C75;font-style:italic">results</span><span style="color:#56B6C2">=</span><span style="color:#61AFEF">runQuery</span><span style="color:#ABB2BF">(mine, query)</span></span>
<span data-line=""> </span>
<span data-line=""><span style="color:#56B6C2">head</span><span style="color:#ABB2BF">(results)</span></span></code></pre></figure>
<h2 id="conclusion"><a aria-hidden="true" tabindex="-1" href="#conclusion"><a href="#conclusion" style="margin-right: 10px">#</a></a>Conclusion</h2>
<p>The InterMineR package has a couple of nice features for getting InterMine data
with a couple of functions for looking at templates. For many use cases, copying
the Query XML from a InterMine webpage and pasting that into the runQuery
function is sufficient and produces a data frame that can be analyzed.</p>
<p>PS it is not easy to post XML on tumblr after editing the post in markdown mode.
You have to add the lt and gt shortcuts and even after that it gets filtered?!</p></div><div class="mt-5"></div></div><footer class="mt-16"><a class="m-2" href="/">Home</a><a class="m-2" href="/archive">Blog archive</a><a class="m-2" href="https://github.com/cmdcolin/">Github</a><a class="m-2" href="/projects">Projects</a><a class="m-2" href="/photos">Photos</a><a class="m-2" href="/books">Books</a><a class="m-2" href="/about">About</a></footer><script src="/_next/static/chunks/webpack-dea5254e1907c5df.js" async=""></script><script>(self.__next_f=self.__next_f||[]).push([0])</script><script>self.__next_f.push([1,"2:\"$Sreact.fragment\"\n3:I[231,[\"231\",\"static/chunks/231-fc74e6e791c0837d.js\",\"931\",\"static/chunks/app/page-95d4a906a5a795f0.js\"],\"\"]\n4:I[9275,[],\"\"]\n5:I[1343,[],\"\"]\n7:I[3120,[],\"OutletBoundary\"]\n9:I[3120,[],\"MetadataBoundary\"]\nb:I[3120,[],\"ViewportBoundary\"]\nd:I[6130,[],\"\"]\n1:HL[\"/_next/static/css/d19cfa26cd1ef694.css\",\"style\"]\n"])</script><script>self.__next_f.push([1,"0:{\"P\":null,\"b\":\"aB5tvoBeFOkGEMABtnoeL\",\"p\":\"\",\"c\":[\"\",\"posts\",\"2016-06-17\"],\"i\":false,\"f\":[[[\"\",{\"children\":[\"posts\",{\"children\":[[\"id\",\"2016-06-17\",\"d\"],{\"children\":[\"__PAGE__?{\\\"id\\\":\\\"2016-06-17\\\"}\",{}]}]}]},\"$undefined\",\"$undefined\",true],[\"\",[\"$\",\"$2\",\"c\",{\"children\":[[[\"$\",\"link\",\"0\",{\"rel\":\"stylesheet\",\"href\":\"/_next/static/css/d19cfa26cd1ef694.css\",\"precedence\":\"next\",\"crossOrigin\":\"$undefined\",\"nonce\":\"$undefined\"}]],[\"$\",\"html\",null,{\"lang\":\"en\",\"children\":[\"$\",\"body\",null,{\"children\":[[\"$\",\"div\",null,{\"className\":\"mb-8\",\"children\":[\"$\",\"$L3\",null,{\"href\":\"/\",\"children\":\"Misc scribbles\"}]}],[\"$\",\"$L4\",null,{\"parallelRouterKey\":\"children\",\"segmentPath\":[\"children\"],\"error\":\"$undefined\",\"errorStyles\":\"$undefined\",\"errorScripts\":\"$undefined\",\"template\":[\"$\",\"$L5\",null,{}],\"templateStyles\":\"$undefined\",\"templateScripts\":\"$undefined\",\"notFound\":[[\"$\",\"title\",null,{\"children\":\"404: This page could not be found.\"}],[\"$\",\"div\",null,{\"style\":{\"fontFamily\":\"system-ui,\\\"Segoe UI\\\",Roboto,Helvetica,Arial,sans-serif,\\\"Apple Color Emoji\\\",\\\"Segoe UI Emoji\\\"\",\"height\":\"100vh\",\"textAlign\":\"center\",\"display\":\"flex\",\"flexDirection\":\"column\",\"alignItems\":\"center\",\"justifyContent\":\"center\"},\"children\":[\"$\",\"div\",null,{\"children\":[[\"$\",\"style\",null,{\"dangerouslySetInnerHTML\":{\"__html\":\"body{color:#000;background:#fff;margin:0}.next-error-h1{border-right:1px solid rgba(0,0,0,.3)}@media (prefers-color-scheme:dark){body{color:#fff;background:#000}.next-error-h1{border-right:1px solid rgba(255,255,255,.3)}}\"}}],[\"$\",\"h1\",null,{\"className\":\"next-error-h1\",\"style\":{\"display\":\"inline-block\",\"margin\":\"0 20px 0 0\",\"padding\":\"0 23px 0 0\",\"fontSize\":24,\"fontWeight\":500,\"verticalAlign\":\"top\",\"lineHeight\":\"49px\"},\"children\":\"404\"}],[\"$\",\"div\",null,{\"style\":{\"display\":\"inline-block\"},\"children\":[\"$\",\"h2\",null,{\"style\":{\"fontSize\":14,\"fontWeight\":400,\"lineHeight\":\"49px\",\"margin\":0},\"children\":\"This page could not be found.\"}]}]]}]}]],\"notFoundStyles\":[]}],[\"$\",\"footer\",null,{\"className\":\"mt-16\",\"children\":[[\"$\",\"$L3\",null,{\"className\":\"m-2\",\"href\":\"/\",\"children\":\"Home\"}],[\"$\",\"$L3\",null,{\"className\":\"m-2\",\"href\":\"/archive\",\"children\":\"Blog archive\"}],[\"$\",\"$L3\",null,{\"className\":\"m-2\",\"href\":\"https://github.com/cmdcolin/\",\"children\":\"Github\"}],[\"$\",\"$L3\",null,{\"className\":\"m-2\",\"href\":\"/projects\",\"children\":\"Projects\"}],[\"$\",\"$L3\",null,{\"className\":\"m-2\",\"href\":\"/photos\",\"children\":\"Photos\"}],[\"$\",\"$L3\",null,{\"className\":\"m-2\",\"href\":\"/books\",\"children\":\"Books\"}],[\"$\",\"$L3\",null,{\"className\":\"m-2\",\"href\":\"/about\",\"children\":\"About\"}]]}]]}]}]]}],{\"children\":[\"posts\",[\"$\",\"$2\",\"c\",{\"children\":[null,[\"$\",\"$L4\",null,{\"parallelRouterKey\":\"children\",\"segmentPath\":[\"children\",\"posts\",\"children\"],\"error\":\"$undefined\",\"errorStyles\":\"$undefined\",\"errorScripts\":\"$undefined\",\"template\":[\"$\",\"$L5\",null,{}],\"templateStyles\":\"$undefined\",\"templateScripts\":\"$undefined\",\"notFound\":\"$undefined\",\"notFoundStyles\":\"$undefined\"}]]}],{\"children\":[[\"id\",\"2016-06-17\",\"d\"],[\"$\",\"$2\",\"c\",{\"children\":[null,[\"$\",\"$L4\",null,{\"parallelRouterKey\":\"children\",\"segmentPath\":[\"children\",\"posts\",\"children\",\"$0:f:0:1:2:children:2:children:0\",\"children\"],\"error\":\"$undefined\",\"errorStyles\":\"$undefined\",\"errorScripts\":\"$undefined\",\"template\":[\"$\",\"$L5\",null,{}],\"templateStyles\":\"$undefined\",\"templateScripts\":\"$undefined\",\"notFound\":\"$undefined\",\"notFoundStyles\":\"$undefined\"}]]}],{\"children\":[\"__PAGE__\",[\"$\",\"$2\",\"c\",{\"children\":[\"$L6\",null,[\"$\",\"$L7\",null,{\"children\":\"$L8\"}]]}],{},null]},null]},null]},null],[\"$\",\"$2\",\"h\",{\"children\":[null,[\"$\",\"$2\",\"-YT6F48bMDHPY4pshRzp4\",{\"children\":[[\"$\",\"$L9\",null,{\"children\":\"$La\"}],[\"$\",\"$Lb\",null,{\"children\":\"$Lc\"}],null]}]]}]]],\"m\":\"$undefined\",\"G\":[\"$d\",\"$undefined\"],\"s\":false,\"S\":true}\n"])</script><script>self.__next_f.push([1,"c:[[\"$\",\"meta\",\"0\",{\"name\":\"viewport\",\"content\":\"width=device-width, initial-scale=1\"}]]\n"])</script><script>self.__next_f.push([1,"f:I[1458,[\"722\",\"static/chunks/app/posts/%5Bid%5D/page-e50cd2f34f19f050.js\"],\"default\"]\ne:T1d39,"])</script><script>self.__next_f.push([1,"\u003cp\u003eIn the past, I had found some ways to do simple queries on InterMine web\nservices using basic HTTP commands with R (see\n\u003ca href=\"https://gist.github.com/cmdcolin/4758167bdd89e6c9c055\"\u003ehttps://gist.github.com/cmdcolin/4758167bdd89e6c9c055\u003c/a\u003e)\u003c/p\u003e\n\u003cp\u003eHowever, the InterMineR (\u003ca href=\"https://github.com/intermine/intermineR\"\u003ehttps://github.com/intermine/intermineR\u003c/a\u003e) package\nautomates some of these features and makes it easier to load the data in R.\u003c/p\u003e\n\u003ch2 id=\"installation\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#installation\"\u003e\u003ca href=\"#installation\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eInstallation\u003c/h2\u003e\n\u003cp\u003eOne way to install InterMineR is to install from github with hadley/devtools\u003c/p\u003e\n\u003cfigure data-rehype-pretty-code-figure=\"\"\u003e\u003cpre style=\"background-color:#282c34;color:#abb2bf\" tabindex=\"0\" data-language=\"r\" data-theme=\"one-dark-pro\"\u003e\u003ccode data-language=\"r\" data-theme=\"one-dark-pro\" style=\"display: grid;\"\u003e\u003cspan data-line=\"\"\u003e\u003cspan style=\"color:#56B6C2\"\u003einstall.packages\u003c/span\u003e\u003cspan style=\"color:#ABB2BF\"\u003e(\u003c/span\u003e\u003cspan style=\"color:#98C379\"\u003e\"devtools\"\u003c/span\u003e\u003cspan style=\"color:#ABB2BF\"\u003e)\u003c/span\u003e\u003c/span\u003e\n\u003cspan data-line=\"\"\u003e\u003cspan style=\"color:#ABB2BF\"\u003edevtools\u003c/span\u003e\u003cspan style=\"color:#C678DD\"\u003e::\u003c/span\u003e\u003cspan style=\"color:#61AFEF\"\u003einstall_github\u003c/span\u003e\u003cspan style=\"color:#ABB2BF\"\u003e(\u003c/span\u003e\u003cspan style=\"color:#98C379\"\u003e\"hadley/devtools\"\u003c/span\u003e\u003cspan style=\"color:#ABB2BF\"\u003e)\u003c/span\u003e\u003c/span\u003e\n\u003cspan data-line=\"\"\u003e\u003cspan style=\"color:#ABB2BF\"\u003edevtools\u003c/span\u003e\u003cspan style=\"color:#C678DD\"\u003e::\u003c/span\u003e\u003cspan style=\"color:#61AFEF\"\u003einstall_github\u003c/span\u003e\u003cspan style=\"color:#ABB2BF\"\u003e(\u003c/span\u003e\u003cspan style=\"color:#98C379\"\u003e\"intermine/intermineR\"\u003c/span\u003e\u003cspan style=\"color:#ABB2BF\"\u003e)\u003c/span\u003e\u003c/span\u003e\u003c/code\u003e\u003c/pre\u003e\u003c/figure\u003e\n\u003ch2 id=\"usage\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#usage\"\u003e\u003ca href=\"#usage\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eUsage\u003c/h2\u003e\n\u003cp\u003eBasic usage includes loading the \"intermine URL\" using the initInterMine\nfunction. Then various functions can be called on this result.\u003c/p\u003e\n\u003cfigure data-rehype-pretty-code-figure=\"\"\u003e\u003cpre style=\"background-color:#282c34;color:#abb2bf\" tabindex=\"0\" data-language=\"r\" data-theme=\"one-dark-pro\"\u003e\u003ccode data-language=\"r\" data-theme=\"one-dark-pro\" style=\"display: grid;\"\u003e\u003cspan data-line=\"\"\u003e\u003cspan style=\"color:#56B6C2\"\u003elibrary\u003c/span\u003e\u003cspan style=\"color:#ABB2BF\"\u003e(InterMineR)\u003c/span\u003e\u003c/span\u003e\n\u003cspan data-line=\"\"\u003e\u003cspan style=\"color:#E06C75;font-style:italic\"\u003emine\u003c/span\u003e\u003cspan style=\"color:#56B6C2\"\u003e=\u003c/span\u003e\u003cspan style=\"color:#61AFEF\"\u003einitInterMine\u003c/span\u003e\u003cspan style=\"color:#ABB2BF\"\u003e(\u003c/span\u003e\u003cspan style=\"color:#98C379\"\u003e\"http://bovinegenome.org/bovinemine/\"\u003c/span\u003e\u003cspan style=\"color:#ABB2BF\"\u003e)\u003c/span\u003e\u003c/span\u003e\n\u003cspan data-line=\"\"\u003e\u003cspan style=\"color:#61AFEF\"\u003egetVersion\u003c/span\u003e\u003cspan style=\"color:#ABB2BF\"\u003e(mine) \u003c/span\u003e\u003cspan style=\"color:#7F848E;font-style:italic\"\u003e#18, intermine API version\u003c/span\u003e\u003c/span\u003e\n\u003cspan data-line=\"\"\u003e\u003cspan style=\"color:#61AFEF\"\u003egetRelease\u003c/span\u003e\u003cspan style=\"color:#ABB2BF\"\u003e(mine) \u003c/span\u003e\u003cspan style=\"color:#7F848E;font-style:italic\"\u003e#1.0, our data release version\u003c/span\u003e\u003c/span\u003e\n\u003cspan data-line=\"\"\u003e\u003cspan style=\"color:#61AFEF\"\u003egetTemplates\u003c/span\u003e\u003cspan style=\"color:#ABB2BF\"\u003e(mine) \u003c/span\u003e\u003cspan style=\"color:#7F848E;font-style:italic\"\u003e# lists all templates on interminer\u003c/span\u003e\u003c/span\u003e\u003c/code\u003e\u003c/pre\u003e\u003c/figure\u003e\n\u003ch2 id=\"run-a-template-query\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#run-a-template-query\"\u003e\u003ca href=\"#run-a-template-query\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eRun a template query\u003c/h2\u003e\n\u003cp\u003eFrom the getTemplates function, if you see a template query that you want to\nrun, you can use the getTemplateQuery function with it's name, and run it with\nthe runQuery function\u003c/p\u003e\n\u003cfigure data-rehype-pretty-code-figure=\"\"\u003e\u003cpre style=\"background-color:#282c34;color:#abb2bf\" tabindex=\"0\" data-language=\"r\" data-theme=\"one-dark-pro\"\u003e\u003ccode data-language=\"r\" data-theme=\"one-dark-pro\" style=\"display: grid;\"\u003e\u003cspan data-line=\"\"\u003e\u003cspan style=\"color:#61AFEF\"\u003egetTemplateQuery\u003c/span\u003e\u003cspan style=\"color:#ABB2BF\"\u003e(mine,\u003c/span\u003e\u003cspan style=\"color:#98C379\"\u003e\"TQ_protein_to_gene\"\u003c/span\u003e\u003cspan style=\"color:#ABB2BF\"\u003e) \u003c/span\u003e\u003cspan style=\"color:#7F848E;font-style:italic\"\u003e# see what template looks like\u003c/span\u003e\u003c/span\u003e\n\u003cspan data-line=\"\"\u003e\u003cspan style=\"color:#E06C75;font-style:italic\"\u003etemplate\u003c/span\u003e\u003cspan style=\"color:#56B6C2\"\u003e=\u003c/span\u003e\u003cspan style=\"color:#61AFEF\"\u003egetTemplateQuery\u003c/span\u003e\u003cspan style=\"color:#ABB2BF\"\u003e(mine,\u003c/span\u003e\u003cspan style=\"color:#98C379\"\u003e\"TQ_protein_to_gene\"\u003c/span\u003e\u003cspan style=\"color:#ABB2BF\"\u003e) \u003c/span\u003e\u003cspan style=\"color:#7F848E;font-style:italic\"\u003e# save template\u003c/span\u003e\u003c/span\u003e\n\u003cspan data-line=\"\"\u003e\u003cspan style=\"color:#61AFEF\"\u003erunQuery\u003c/span\u003e\u003cspan style=\"color:#ABB2BF\"\u003e(mine,template) \u003c/span\u003e\u003cspan style=\"color:#7F848E;font-style:italic\"\u003e# run the template query with default params, receive data.frame\u003c/span\u003e\u003c/span\u003e\u003c/code\u003e\u003c/pre\u003e\u003c/figure\u003e\n\u003cp\u003eThis method is good, but some improvement could be added to change default\nparameters in the template query, etc.\u003c/p\u003e\n\u003ch2 id=\"run-query-xml\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#run-query-xml\"\u003e\u003ca href=\"#run-query-xml\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eRun query XML\u003c/h2\u003e\n\u003cp\u003eAnother option for running queries is to use the query XML that you can download\nfrom the InterMine query result pages.\u003c/p\u003e\n\u003cfigure data-rehype-pretty-code-figure=\"\"\u003e\u003cpre style=\"background-color:#282c34;color:#abb2bf\" tabindex=\"0\" data-language=\"r\" data-theme=\"one-dark-pro\"\u003e\u003ccode data-language=\"r\" data-theme=\"one-dark-pro\" style=\"display: grid;\"\u003e\u003cspan data-line=\"\"\u003e\u003cspan style=\"color:#7F848E;font-style:italic\"\u003e# get all Ensembl genes on chr28 from bovinemine\u003c/span\u003e\u003c/span\u003e\n\u003cspan data-line=\"\"\u003e\u003cspan style=\"color:#E06C75;font-style:italic\"\u003equery\u003c/span\u003e\u003cspan style=\"color:#56B6C2\"\u003e=\u003c/span\u003e\u003cspan style=\"color:#98C379\"\u003e'\u0026#x3C;query model=\"genomic\" view=\"Gene.primaryIdentifier\u003c/span\u003e\u003c/span\u003e\n\u003cspan data-line=\"\"\u003e\u003cspan style=\"color:#98C379\"\u003eGene.secondaryIdentifier Gene.symbol Gene.name Gene.source\u003c/span\u003e\u003c/span\u003e\n\u003cspan data-line=\"\"\u003e\u003cspan style=\"color:#98C379\"\u003eGene.organism.shortName Gene.chromosome.primaryIdentifier\"\u003c/span\u003e\u003c/span\u003e\n\u003cspan data-line=\"\"\u003e\u003cspan style=\"color:#98C379\"\u003esortOrder=\"Gene.primaryIdentifier ASC\" \u003e\u0026#x3C;constraint\u003c/span\u003e\u003c/span\u003e\n\u003cspan data-line=\"\"\u003e\u003cspan style=\"color:#98C379\"\u003epath=\"Gene.organism.shortName\" op=\"=\" value=\"B. taurus\"\u003c/span\u003e\u003c/span\u003e\n\u003cspan data-line=\"\"\u003e\u003cspan style=\"color:#98C379\"\u003e/\u003e\u0026#x3C;constraint path=\"Gene.chromosome.primaryIdentifier\" op=\"=\"\u003c/span\u003e\u003c/span\u003e\n\u003cspan data-line=\"\"\u003e\u003cspan style=\"color:#98C379\"\u003evalue=\"GK000028.2\" /\u003e\u0026#x3C;/query\u003e'\u003c/span\u003e\u003c/span\u003e\n\u003cspan data-line=\"\"\u003e \u003c/span\u003e\n\u003cspan data-line=\"\"\u003e\u003cspan style=\"color:#E06C75;font-style:italic\"\u003eresults\u003c/span\u003e\u003cspan style=\"color:#56B6C2\"\u003e=\u003c/span\u003e\u003cspan style=\"color:#61AFEF\"\u003erunQuery\u003c/span\u003e\u003cspan style=\"color:#ABB2BF\"\u003e(mine, query)\u003c/span\u003e\u003c/span\u003e\n\u003cspan data-line=\"\"\u003e \u003c/span\u003e\n\u003cspan data-line=\"\"\u003e\u003cspan style=\"color:#56B6C2\"\u003ehead\u003c/span\u003e\u003cspan style=\"color:#ABB2BF\"\u003e(results)\u003c/span\u003e\u003c/span\u003e\u003c/code\u003e\u003c/pre\u003e\u003c/figure\u003e\n\u003ch2 id=\"conclusion\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#conclusion\"\u003e\u003ca href=\"#conclusion\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eConclusion\u003c/h2\u003e\n\u003cp\u003eThe InterMineR package has a couple of nice features for getting InterMine data\nwith a couple of functions for looking at templates. For many use cases, copying\nthe Query XML from a InterMine webpage and pasting that into the runQuery\nfunction is sufficient and produces a data frame that can be analyzed.\u003c/p\u003e\n\u003cp\u003ePS it is not easy to post XML on tumblr after editing the post in markdown mode.\nYou have to add the lt and gt shortcuts and even after that it gets filtered?!\u003c/p\u003e"])</script><script>self.__next_f.push([1,"6:[\"$\",\"div\",null,{\"className\":\"lg:w-1/2\",\"children\":[[\"$\",\"div\",null,{\"children\":[[\"$\",\"h1\",null,{\"children\":\"Querying InterMine databases using R\"}],[\"$\",\"h4\",null,{\"children\":\"2016-06-17\"}]]}],[\"$\",\"div\",null,{\"dangerouslySetInnerHTML\":{\"__html\":\"$e\"}}],[\"$\",\"$Lf\",null,{}]]}]\na:[[\"$\",\"meta\",\"0\",{\"charSet\":\"utf-8\"}],[\"$\",\"title\",\"1\",{\"children\":\"Querying InterMine databases using R\"}],[\"$\",\"meta\",\"2\",{\"name\":\"description\",\"content\":\"A blog\"}]]\n8:null\n"])</script></body></html>