<!DOCTYPE html><html lang="en"><head><link rel="shortcut icon" href="favicon.ico"/><link rel="alternate" type="application/rss+xml" href="/rss.xml"/><meta name="description" content="Blogging for the future"/><script src="https://cdnjs.cloudflare.com/ajax/libs/prism/1.25.0/prism.min.js" integrity="sha512-hpZ5pDCF2bRCweL5WoA0/N1elet1KYL5mx3LP555Eg/0ZguaHawxNvEjF6O3rufAChs16HVNhEc6blF/rZoowQ==" crossorigin="anonymous" referrerPolicy="no-referrer"></script><meta name="viewport" content="width=device-width"/><meta charSet="utf-8"/><title>Querying InterMine databases using R</title><meta name="next-head-count" content="3"/><link rel="preload" href="/_next/static/css/314c24908ac760cf.css" as="style"/><link rel="stylesheet" href="/_next/static/css/314c24908ac760cf.css" data-n-g=""/><noscript data-n-css=""></noscript><script defer="" nomodule="" src="/_next/static/chunks/polyfills-5cd94c89d3acac5f.js"></script><script src="/_next/static/chunks/webpack-514908bffb652963.js" defer=""></script><script src="/_next/static/chunks/framework-91d7f78b5b4003c8.js" defer=""></script><script src="/_next/static/chunks/main-d98b4a7f39fdfc80.js" defer=""></script><script src="/_next/static/chunks/pages/_app-76cd57bf65f05d70.js" defer=""></script><script src="/_next/static/chunks/pages/posts/%5Bslug%5D-6db65d5e18d8108f.js" defer=""></script><script src="/_next/static/HaI7tQbHT8lCPqt5Zh7pO/_buildManifest.js" defer=""></script><script src="/_next/static/HaI7tQbHT8lCPqt5Zh7pO/_ssgManifest.js" defer=""></script><script src="/_next/static/HaI7tQbHT8lCPqt5Zh7pO/_middlewareManifest.js" defer=""></script></head><body><div id="__next" data-reactroot=""><div><main><div><div style="margin-bottom:100px"><a href="/">Misc scribbles</a></div><article><div><h1>Querying InterMine databases using R</h1><h4>2016-06-17</h4></div><p>In the past, I had found some ways to do simple queries on InterMine web
services using basic HTTP commands with R (see
<a href="https://gist.github.com/cmdcolin/4758167bdd89e6c9c055">https://gist.github.com/cmdcolin/4758167bdd89e6c9c055</a>)</p><p>However, the InterMineR (<a href="https://github.com/intermine/intermineR">https://github.com/intermine/intermineR</a>)
package automates some of these features and makes it easier to load the
data in R.</p><p><strong>Installation</strong></p><p>One way to install InterMineR is to install from github with
hadley/devtools</p><pre><code>install.packages(&quot;devtools&quot;)
devtools::install_github(&quot;hadley/devtools&quot;)
devtools::install_github(&quot;intermine/intermineR&quot;)
</code></pre><p><strong>Usage</strong></p><p>Basic usage includes loading theÂ &quot;intermine URL&quot; using the initInterMine
function. Then various functions can be called on this result.</p><pre><code>library(InterMineR)
mine=initInterMine(&quot;http://bovinegenome.org/bovinemine/&quot;)
getVersion(mine) #18, intermine API version
getRelease(mine) #1.0, our data release version
getTemplates(mine) # lists all templates on interminer
</code></pre><p><strong>Run a template query</strong></p><p>From the getTemplates function, if you see a template query that you
want to run, you can use the getTemplateQuery function with it&#x27;s name,
and run it with the runQuery function</p><pre><code>getTemplateQuery(mine,&quot;TQ_protein_to_gene&quot;) # see what template looks like
template=getTemplateQuery(mine,&quot;TQ_protein_to_gene&quot;) # save template
runQuery(mine,template) # run the template query with default params, receive data.frame
</code></pre><p>This method is good, but some improvement could be added to change
default parameters in the template query, etc.</p><p><strong>Run query XML</strong></p><p>Another option for running queries is to use the query XML that you can
download from the InterMine query result pages.</p><pre><code> # get all Ensembl genes on chr28 from bovinemine
 query=&#x27;&lt;query model=&quot;genomic&quot; view=&quot;Gene.primaryIdentifier
 Gene.secondaryIdentifier Gene.symbol Gene.name Gene.source
 Gene.organism.shortName Gene.chromosome.primaryIdentifier&quot;
 sortOrder=&quot;Gene.primaryIdentifier ASC&quot; &gt;&lt;constraint
 path=&quot;Gene.organism.shortName&quot; op=&quot;=&quot; value=&quot;B. taurus&quot;
 /&gt;&lt;constraint path=&quot;Gene.chromosome.primaryIdentifier&quot; op=&quot;=&quot;
 value=&quot;GK000028.2&quot; /&gt;&lt;/query&gt;&#x27;

 results=runQuery(mine, query)

 head(results)
</code></pre><p><strong>Conclusion</strong></p><p>The InterMineR package has a couple of nice features for getting
InterMine data with a couple of functions for looking at templates. For
many use cases, copying the Query XML from a InterMine webpage and
pasting that into the runQuery function is sufficient and produces a
data frame that can be analyzed.</p><p>PS it is not easy to post XML on tumblr after editing the post in
markdown mode. You have to add the lt and gt shortcuts and even after
that it gets filtered?!</p></article></div></main></div><footer style="margin-top:100px"><a href="/">Home</a> <a href="/archive">Blog archive</a> <a href="https://github.com/cmdcolin">Github</a> <a href="https://twitter.com/cmdcolin">Twitter</a> <a href="/projects">Projects</a> <a href="https://scholar.google.com/citations?view_op=list_works&amp;hl=en&amp;user=--FwzsgAAAAJ">Publications</a> <a href="/sketches">Sketches</a> <a href="/kitty">Kitty</a> <a href="/rss.xml">RSS</a><a href="/about">About</a> </footer></div><script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{"post":{"title":"Querying InterMine databases using R","date":"2016-06-17","slug":"2016-06-17","mdxSource":{"compiledSource":"var l=Object.defineProperty,d=Object.defineProperties;var c=Object.getOwnPropertyDescriptors;var o=Object.getOwnPropertySymbols;var i=Object.prototype.hasOwnProperty,s=Object.prototype.propertyIsEnumerable;var p=(e,t,n)=\u003et in e?l(e,t,{enumerable:!0,configurable:!0,writable:!0,value:n}):e[t]=n,r=(e,t)=\u003e{for(var n in t||(t={}))i.call(t,n)\u0026\u0026p(e,n,t[n]);if(o)for(var n of o(t))s.call(t,n)\u0026\u0026p(e,n,t[n]);return e},m=(e,t)=\u003ed(e,c(t));var u=(e,t)=\u003e{var n={};for(var a in e)i.call(e,a)\u0026\u0026t.indexOf(a)\u003c0\u0026\u0026(n[a]=e[a]);if(e!=null\u0026\u0026o)for(var a of o(e))t.indexOf(a)\u003c0\u0026\u0026s.call(e,a)\u0026\u0026(n[a]=e[a]);return n};const layoutProps={},MDXLayout=\"wrapper\";function MDXContent(n){var a=n,{components:e}=a,t=u(a,[\"components\"]);return mdx(MDXLayout,m(r(r({},layoutProps),t),{components:e,mdxType:\"MDXLayout\"}),mdx(\"p\",null,`In the past, I had found some ways to do simple queries on InterMine web\nservices using basic HTTP commands with R (see\n`,mdx(\"a\",r({parentName:\"p\"},{href:\"https://gist.github.com/cmdcolin/4758167bdd89e6c9c055\"}),\"https://gist.github.com/cmdcolin/4758167bdd89e6c9c055\"),\")\"),mdx(\"p\",null,\"However, the InterMineR (\",mdx(\"a\",r({parentName:\"p\"},{href:\"https://github.com/intermine/intermineR\"}),\"https://github.com/intermine/intermineR\"),`)\npackage automates some of these features and makes it easier to load the\ndata in R.`),mdx(\"p\",null,mdx(\"strong\",{parentName:\"p\"},\"Installation\")),mdx(\"p\",null,`One way to install InterMineR is to install from github with\nhadley/devtools`),mdx(\"pre\",null,mdx(\"code\",r({parentName:\"pre\"},{}),`install.packages(\"devtools\")\ndevtools::install_github(\"hadley/devtools\")\ndevtools::install_github(\"intermine/intermineR\")\n`)),mdx(\"p\",null,mdx(\"strong\",{parentName:\"p\"},\"Usage\")),mdx(\"p\",null,`Basic usage includes loading the\\xA0\"intermine URL\" using the initInterMine\nfunction. Then various functions can be called on this result.`),mdx(\"pre\",null,mdx(\"code\",r({parentName:\"pre\"},{}),`library(InterMineR)\nmine=initInterMine(\"http://bovinegenome.org/bovinemine/\")\ngetVersion(mine) #18, intermine API version\ngetRelease(mine) #1.0, our data release version\ngetTemplates(mine) # lists all templates on interminer\n`)),mdx(\"p\",null,mdx(\"strong\",{parentName:\"p\"},\"Run a template query\")),mdx(\"p\",null,`From the getTemplates function, if you see a template query that you\nwant to run, you can use the getTemplateQuery function with it's name,\nand run it with the runQuery function`),mdx(\"pre\",null,mdx(\"code\",r({parentName:\"pre\"},{}),`getTemplateQuery(mine,\"TQ_protein_to_gene\") # see what template looks like\ntemplate=getTemplateQuery(mine,\"TQ_protein_to_gene\") # save template\nrunQuery(mine,template) # run the template query with default params, receive data.frame\n`)),mdx(\"p\",null,`This method is good, but some improvement could be added to change\ndefault parameters in the template query, etc.`),mdx(\"p\",null,mdx(\"strong\",{parentName:\"p\"},\"Run query XML\")),mdx(\"p\",null,`Another option for running queries is to use the query XML that you can\ndownload from the InterMine query result pages.`),mdx(\"pre\",null,mdx(\"code\",r({parentName:\"pre\"},{}),` # get all Ensembl genes on chr28 from bovinemine\n query='\u003cquery model=\"genomic\" view=\"Gene.primaryIdentifier\n Gene.secondaryIdentifier Gene.symbol Gene.name Gene.source\n Gene.organism.shortName Gene.chromosome.primaryIdentifier\"\n sortOrder=\"Gene.primaryIdentifier ASC\" \u003e\u003cconstraint\n path=\"Gene.organism.shortName\" op=\"=\" value=\"B. taurus\"\n /\u003e\u003cconstraint path=\"Gene.chromosome.primaryIdentifier\" op=\"=\"\n value=\"GK000028.2\" /\u003e\u003c/query\u003e'\n\n results=runQuery(mine, query)\n\n head(results)\n`)),mdx(\"p\",null,mdx(\"strong\",{parentName:\"p\"},\"Conclusion\")),mdx(\"p\",null,`The InterMineR package has a couple of nice features for getting\nInterMine data with a couple of functions for looking at templates. For\nmany use cases, copying the Query XML from a InterMine webpage and\npasting that into the runQuery function is sufficient and produces a\ndata frame that can be analyzed.`),mdx(\"p\",null,`PS it is not easy to post XML on tumblr after editing the post in\nmarkdown mode. You have to add the lt and gt shortcuts and even after\nthat it gets filtered?!`))}MDXContent.isMDXComponent=!0;\n","scope":{}}}},"__N_SSG":true},"page":"/posts/[slug]","query":{"slug":"2016-06-17"},"buildId":"HaI7tQbHT8lCPqt5Zh7pO","isFallback":false,"gsp":true,"scriptLoader":[]}</script></body></html>