<!DOCTYPE html><html lang="en"><head><meta charSet="utf-8"/><meta name="viewport" content="width=device-width"/><title>Querying InterMine databases using R</title><meta name="next-head-count" content="3"/><link rel="shortcut icon" href="favicon.ico"/><link rel="alternate" type="application/rss+xml" href="/rss.xml"/><meta name="description" content="Blogging for the future"/><script src="https://cdnjs.cloudflare.com/ajax/libs/prism/1.25.0/prism.min.js" integrity="sha512-hpZ5pDCF2bRCweL5WoA0/N1elet1KYL5mx3LP555Eg/0ZguaHawxNvEjF6O3rufAChs16HVNhEc6blF/rZoowQ==" crossorigin="anonymous" referrerPolicy="no-referrer"></script><link rel="preload" href="/_next/static/css/9d066bf523979bbf.css" as="style"/><link rel="stylesheet" href="/_next/static/css/9d066bf523979bbf.css" data-n-g=""/><noscript data-n-css=""></noscript><script defer="" nomodule="" src="/_next/static/chunks/polyfills-0d1b80a048d4787e.js"></script><script src="/_next/static/chunks/webpack-8b8836cfd4de0dcd.js" defer=""></script><script src="/_next/static/chunks/framework-c0d8f0fd2eea5ac1.js" defer=""></script><script src="/_next/static/chunks/main-ad18704b098b4264.js" defer=""></script><script src="/_next/static/chunks/pages/_app-d9139de5c3de3b6b.js" defer=""></script><script src="/_next/static/chunks/996-f78e375898643c6a.js" defer=""></script><script src="/_next/static/chunks/pages/posts/%5Bslug%5D-34e532adea304c66.js" defer=""></script><script src="/_next/static/C7V0DlGSdj0drWaU6MWzM/_buildManifest.js" defer=""></script><script src="/_next/static/C7V0DlGSdj0drWaU6MWzM/_ssgManifest.js" defer=""></script></head><body><div id="__next"><div><main><div><div style="margin-bottom:100px"><a href="/">Misc scribbles</a></div><article><div><h1>Querying InterMine databases using R</h1><h4>2016-06-17</h4></div><p>In the past, I had found some ways to do simple queries on InterMine web
services using basic HTTP commands with R (see
https://gist.github.com/cmdcolin/4758167bdd89e6c9c055)</p>
<p>However, the InterMineR (https://github.com/intermine/intermineR)
package automates some of these features and makes it easier to load the
data in R.</p>
<p><strong>Installation</strong></p>
<p>One way to install InterMineR is to install from github with
hadley/devtools</p>
<p>install.packages(&quot;devtools&quot;)
devtools::install_github(&quot;hadley/devtools&quot;)
devtools::install_github(&quot;intermine/intermineR&quot;)</p>
<p><strong>Usage</strong></p>
<p>Basic usage includes loading the &quot;intermine URL&quot; using the initInterMine
function. Then various functions can be called on this result.</p>
<p>library(InterMineR)
mine=initInterMine(&quot;http://bovinegenome.org/bovinemine/&quot;)
getVersion(mine) #18, intermine API version
getRelease(mine) #1.0, our data release version
getTemplates(mine) # lists all templates on interminer</p>
<p><strong>Run a template query</strong></p>
<p>From the getTemplates function, if you see a template query that you
want to run, you can use the getTemplateQuery function with it&#x27;s name,
and run it with the runQuery function</p>
<p>getTemplateQuery(mine,&quot;TQ_protein_to_gene&quot;) # see what template looks like
template=getTemplateQuery(mine,&quot;TQ_protein_to_gene&quot;) # save template
runQuery(mine,template) # run the template query with default params, receive data.frame</p>
<p>This method is good, but some improvement could be added to change
default parameters in the template query, etc.</p>
<p><strong>Run query XML</strong></p>
<p>Another option for running queries is to use the query XML that you can
download from the InterMine query result pages.</p>
<pre><code> # get all Ensembl genes on chr28 from bovinemine
 query=&#x27;&lt;query model=&quot;genomic&quot; view=&quot;Gene.primaryIdentifier
 Gene.secondaryIdentifier Gene.symbol Gene.name Gene.source
 Gene.organism.shortName Gene.chromosome.primaryIdentifier&quot;
 sortOrder=&quot;Gene.primaryIdentifier ASC&quot; &gt;&lt;constraint
 path=&quot;Gene.organism.shortName&quot; op=&quot;=&quot; value=&quot;B. taurus&quot;
 /&gt;&lt;constraint path=&quot;Gene.chromosome.primaryIdentifier&quot; op=&quot;=&quot;
 value=&quot;GK000028.2&quot; /&gt;&lt;/query&gt;&#x27;

 results=runQuery(mine, query)

 head(results)
</code></pre>
<p><strong>Conclusion</strong></p>
<p>The InterMineR package has a couple of nice features for getting
InterMine data with a couple of functions for looking at templates. For
many use cases, copying the Query XML from a InterMine webpage and
pasting that into the runQuery function is sufficient and produces a
data frame that can be analyzed.</p>
<p>PS it is not easy to post XML on tumblr after editing the post in
markdown mode. You have to add the lt and gt shortcuts and even after
that it gets filtered?!</p><div style="margin-top:200px"></div></article></div></main></div><footer style="margin-top:100px"><a href="/">Home</a> <a href="/archive">Blog archive</a> <a href="https://github.com/cmdcolin/">Github</a> <a href="https://twitter.com/cmdcolin">Twitter</a> <a href="/projects">Projects</a> <a href="/photos">Photos</a> <a href="/rss.xml">RSS</a><a href="/about">About</a> </footer></div><script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{"post":{"title":"Querying InterMine databases using R","date":"2016-06-17","slug":"2016-06-17","mdxSource":{"compiledSource":"/*@jsxRuntime automatic @jsxImportSource react*/\nconst {Fragment: _Fragment, jsx: _jsx, jsxs: _jsxs} = arguments[0];\nconst {useMDXComponents: _provideComponents} = arguments[0];\nfunction _createMdxContent(props) {\n  const _components = Object.assign({\n    p: \"p\",\n    strong: \"strong\",\n    pre: \"pre\",\n    code: \"code\"\n  }, _provideComponents(), props.components);\n  return _jsxs(_Fragment, {\n    children: [_jsx(_components.p, {\n      children: \"In the past, I had found some ways to do simple queries on InterMine web\\nservices using basic HTTP commands with R (see\\nhttps://gist.github.com/cmdcolin/4758167bdd89e6c9c055)\"\n    }), \"\\n\", _jsx(_components.p, {\n      children: \"However, the InterMineR (https://github.com/intermine/intermineR)\\npackage automates some of these features and makes it easier to load the\\ndata in R.\"\n    }), \"\\n\", _jsx(_components.p, {\n      children: _jsx(_components.strong, {\n        children: \"Installation\"\n      })\n    }), \"\\n\", _jsx(_components.p, {\n      children: \"One way to install InterMineR is to install from github with\\nhadley/devtools\"\n    }), \"\\n\", _jsx(_components.p, {\n      children: \"install.packages(\\\"devtools\\\")\\ndevtools::install_github(\\\"hadley/devtools\\\")\\ndevtools::install_github(\\\"intermine/intermineR\\\")\"\n    }), \"\\n\", _jsx(_components.p, {\n      children: _jsx(_components.strong, {\n        children: \"Usage\"\n      })\n    }), \"\\n\", _jsx(_components.p, {\n      children: \"Basic usage includes loading the \\\"intermine URL\\\" using the initInterMine\\nfunction. Then various functions can be called on this result.\"\n    }), \"\\n\", _jsx(_components.p, {\n      children: \"library(InterMineR)\\nmine=initInterMine(\\\"http://bovinegenome.org/bovinemine/\\\")\\ngetVersion(mine) #18, intermine API version\\ngetRelease(mine) #1.0, our data release version\\ngetTemplates(mine) # lists all templates on interminer\"\n    }), \"\\n\", _jsx(_components.p, {\n      children: _jsx(_components.strong, {\n        children: \"Run a template query\"\n      })\n    }), \"\\n\", _jsx(_components.p, {\n      children: \"From the getTemplates function, if you see a template query that you\\nwant to run, you can use the getTemplateQuery function with it's name,\\nand run it with the runQuery function\"\n    }), \"\\n\", _jsx(_components.p, {\n      children: \"getTemplateQuery(mine,\\\"TQ_protein_to_gene\\\") # see what template looks like\\ntemplate=getTemplateQuery(mine,\\\"TQ_protein_to_gene\\\") # save template\\nrunQuery(mine,template) # run the template query with default params, receive data.frame\"\n    }), \"\\n\", _jsx(_components.p, {\n      children: \"This method is good, but some improvement could be added to change\\ndefault parameters in the template query, etc.\"\n    }), \"\\n\", _jsx(_components.p, {\n      children: _jsx(_components.strong, {\n        children: \"Run query XML\"\n      })\n    }), \"\\n\", _jsx(_components.p, {\n      children: \"Another option for running queries is to use the query XML that you can\\ndownload from the InterMine query result pages.\"\n    }), \"\\n\", _jsx(_components.pre, {\n      children: _jsx(_components.code, {\n        children: \" # get all Ensembl genes on chr28 from bovinemine\\n query='\u003cquery model=\\\"genomic\\\" view=\\\"Gene.primaryIdentifier\\n Gene.secondaryIdentifier Gene.symbol Gene.name Gene.source\\n Gene.organism.shortName Gene.chromosome.primaryIdentifier\\\"\\n sortOrder=\\\"Gene.primaryIdentifier ASC\\\" \u003e\u003cconstraint\\n path=\\\"Gene.organism.shortName\\\" op=\\\"=\\\" value=\\\"B. taurus\\\"\\n /\u003e\u003cconstraint path=\\\"Gene.chromosome.primaryIdentifier\\\" op=\\\"=\\\"\\n value=\\\"GK000028.2\\\" /\u003e\u003c/query\u003e'\\n\\n results=runQuery(mine, query)\\n\\n head(results)\\n\"\n      })\n    }), \"\\n\", _jsx(_components.p, {\n      children: _jsx(_components.strong, {\n        children: \"Conclusion\"\n      })\n    }), \"\\n\", _jsx(_components.p, {\n      children: \"The InterMineR package has a couple of nice features for getting\\nInterMine data with a couple of functions for looking at templates. For\\nmany use cases, copying the Query XML from a InterMine webpage and\\npasting that into the runQuery function is sufficient and produces a\\ndata frame that can be analyzed.\"\n    }), \"\\n\", _jsx(_components.p, {\n      children: \"PS it is not easy to post XML on tumblr after editing the post in\\nmarkdown mode. You have to add the lt and gt shortcuts and even after\\nthat it gets filtered?!\"\n    })]\n  });\n}\nfunction MDXContent(props = {}) {\n  const {wrapper: MDXLayout} = Object.assign({}, _provideComponents(), props.components);\n  return MDXLayout ? _jsx(MDXLayout, Object.assign({}, props, {\n    children: _jsx(_createMdxContent, props)\n  })) : _createMdxContent(props);\n}\nreturn {\n  default: MDXContent\n};\n","frontmatter":{},"scope":{}}}},"__N_SSG":true},"page":"/posts/[slug]","query":{"slug":"2016-06-17"},"buildId":"C7V0DlGSdj0drWaU6MWzM","isFallback":false,"gsp":true,"scriptLoader":[]}</script></body></html>