<!DOCTYPE html><html lang="en"><head><meta charSet="utf-8"/><meta name="viewport" content="width=device-width"/><title>Structural variants and the SAM format - the long (reads) and short (reads) of it</title><meta name="next-head-count" content="3"/><link rel="shortcut icon" href="favicon.ico"/><link rel="alternate" type="application/rss+xml" href="/rss.xml"/><meta name="description" content="Blogging for the future"/><script src="https://cdnjs.cloudflare.com/ajax/libs/prism/1.25.0/prism.min.js" integrity="sha512-hpZ5pDCF2bRCweL5WoA0/N1elet1KYL5mx3LP555Eg/0ZguaHawxNvEjF6O3rufAChs16HVNhEc6blF/rZoowQ==" crossorigin="anonymous" referrerPolicy="no-referrer"></script><link rel="preload" href="/_next/static/css/50853c3f2e0364c3.css" as="style"/><link rel="stylesheet" href="/_next/static/css/50853c3f2e0364c3.css" data-n-g=""/><noscript data-n-css=""></noscript><script defer="" nomodule="" src="/_next/static/chunks/polyfills-c67a75d1b6f99dc8.js"></script><script src="/_next/static/chunks/webpack-93aa87d657ed1852.js" defer=""></script><script src="/_next/static/chunks/framework-c0d8f0fd2eea5ac1.js" defer=""></script><script src="/_next/static/chunks/main-ae0707941214d030.js" defer=""></script><script src="/_next/static/chunks/pages/_app-d9139de5c3de3b6b.js" defer=""></script><script src="/_next/static/chunks/996-06f747e212688e27.js" defer=""></script><script src="/_next/static/chunks/pages/posts/%5Bslug%5D-b6f09b4db8bfc341.js" defer=""></script><script src="/_next/static/hr7NfWqLWdv1Kvd3x6-Z_/_buildManifest.js" defer=""></script><script src="/_next/static/hr7NfWqLWdv1Kvd3x6-Z_/_ssgManifest.js" defer=""></script></head><body><div id="__next"><div><main><div><div style="margin-bottom:100px"><a href="/">Misc scribbles</a></div><article><div><h1>Structural variants and the SAM format - the long (reads) and short (reads) of it</h1><h4>2022-02-06</h4></div><div><p>The <code>SAM</code> specification is pretty amazing
(<a href="https://samtools.github.io/hts-specs/SAMv1.pdf">https://samtools.github.io/hts-specs/SAMv1.pdf</a>) but it is also fairly terse
and abstract. True understanding might come from playing with real world data.
I will try to relay some things I have learned over the years, with a bit of a
focus on how <code>SAM</code> file concepts can relate to structural variants.</p>
<p>Disclaimer: I'm a developer of JBrowse 2. This document has some screenshots
and links for it, feel free to try it at <a href="https://jbrowse.org">https://jbrowse.org</a>.</p>
<h2 id="basics"><a aria-hidden="true" tabindex="-1" href="#basics"><a href="#basics" style="margin-right: 10px">#</a></a>Basics</h2>
<h3 id="what-is-a-sam-file-and-how-does-it-relate-to-bam-and-cram"><a aria-hidden="true" tabindex="-1" href="#what-is-a-sam-file-and-how-does-it-relate-to-bam-and-cram"><a href="#what-is-a-sam-file-and-how-does-it-relate-to-bam-and-cram" style="margin-right: 10px">#</a></a>What is a <code>SAM</code> file and how does it relate to <code>BAM</code> and <code>CRAM</code>?</h3>
<ul>
<li>
<p>A <code>SAM</code> file <strong>generally</strong> contains "reads" from a sequencer, with information
about how they are mapped to a reference genome [1][2].</p>
</li>
<li>
<p>A <code>SAM</code> file is <strong>generally</strong> produced when an aligner takes in raw unaligned reads
(often stored in <code>FASTQ</code> format files) and aligns them to a reference genome [3].</p>
</li>
<li>
<p>A <code>SAM</code> file is a text format that you can read with your text editor. <code>BAM</code> and
<code>CRAM</code> are compressed representations of the <code>SAM</code> format.</p>
</li>
</ul>
<p>You can convert <code>SAM</code> to <code>BAM</code> with samtools</p>
<div class="highlight highlight-shell"><pre>samtools view file.sam -o file.bam
</pre></div>
<p>You can also convert a <code>BAM</code> back to <code>SAM</code> with samtools view</p>
<div class="highlight highlight-shell"><pre>samtools view -h file.bam -o file.sam
</pre></div>
<p>The -h just makes sure to preserve the header.</p>
<p>If you are converting <code>SAM</code> to <code>CRAM</code>, it may require the -T argument to
specify your reference sequence (this is because the <code>CRAM</code> is "reference
compressed")</p>
<div class="highlight highlight-shell"><pre>samtools view -T reference.fa file.sam -o file.cram
</pre></div>
<p>Also see Appendix C: piping FASTQ from <code>minimap2</code> directly to CRAM</p>
<p>[1] <code>SAM</code> can contain any type of sequence, not specifically reads. If you
created a <em>de novo</em> assembly, you could align the contigs of the <em>de novo</em> assembly
to a reference genome and store the results in <code>SAM</code>.</p>
<p>[2] Does not always have to have information about mapping to a reference
genome. You can also store unaligned data in <code>SAM</code>/<code>BAM</code>/<code>CRAM</code> (so-called
<code>uBAM</code> for example) but most of the time, the reads in <code>SAM</code> format are aligned
to a reference genome.</p>
<p>[3] Examples of programs that do alignment include <code>bwa</code>, <code>bowtie</code>, and <code>minimap2</code>
(there are many others). These programs all can produce <code>SAM</code> outputs</p>
<h3 id="what-is-in-a-sam-file"><a aria-hidden="true" tabindex="-1" href="#what-is-in-a-sam-file"><a href="#what-is-in-a-sam-file" style="margin-right: 10px">#</a></a>What is in a <code>SAM</code> file</h3>
<p>A <code>SAM</code> file contains a header (<code>BAM</code> and <code>CRAM</code> files also have the <code>SAM</code>
header) and a series of records. A record is a single line in a <code>SAM</code> file, and
it generally corresponds to a single read, but as we will see, a split
alignment may produce multiple records that refer to the same source read.</p>
<p>Note: if a read failed to align to the reference genome, it may still be in
your <code>SAM</code> file, marked as unmapped using the flag column. Sometimes, "dumpster
diving" (looking at the unmapped records from a <code>SAM</code> file) can be used to aid
structural variant searches (e.g. there may be novel sequence in there not from
the reference genome that could be assembled)</p>
<h3 id="what-are-tags-in-a-sam-file"><a aria-hidden="true" tabindex="-1" href="#what-are-tags-in-a-sam-file"><a href="#what-are-tags-in-a-sam-file" style="margin-right: 10px">#</a></a>What are tags in a <code>SAM</code> file</h3>
<p>A SAM file has a core set of required fields, and then an arbitrary list of
extra columns called tags. The tags have a two character abbreviation like <code>MQ</code>
(mapping quality) or many others. They can be upper or lower case. Upper case
are reserved for official usages (except those with X, Y, or Z prefixed). See
<a href="https://samtools.github.io/hts-specs/SAMtags.pdf">SAMtags.pdf</a> for more
details</p>
<h3 id="what-is-a-cigar-string"><a aria-hidden="true" tabindex="-1" href="#what-is-a-cigar-string"><a href="#what-is-a-cigar-string" style="margin-right: 10px">#</a></a>What is a <code>CIGAR</code> string</h3>
<p>A <code>CIGAR</code> string is a "compact idiosyncratic gapped alignment report". It tells
you about insertions, deletions, and clipping. It is a series of "operators"
with lengths.</p>
<p>Insertion example:</p>
<p><code>50M50I50M</code></p>
<p>That would be 50bp of matching bases (<code>50M</code>), followed by a 50bp insertion
(<code>50I</code>), followed by another 50bp of matches (<code>50M</code>). The 50bp insertion means
the read contains 50 bases in the middle which did not match the reference
genome that you are comparing the read to.</p>
<p>Clipping example:</p>
<p><code>50S50M50S</code></p>
<p>This means that 50bp matched (<code>50M</code> in the middle of the <code>CIGAR</code> string) and
both sides of the read are soft clipped. The clipping means the aligner was not
able to align the reads on either side. You could imagine clipping being like
"an insertion on either side of the read" if you like: basically those bases on
either side did not align.</p>
<p>Notes:</p>
<ul>
<li>
<p><em>Finding mismatches</em>: A <code>CIGAR</code> string match like <code>50M</code> means 50 bases
"matched" the reference genome, but that only means that there are no
insertions or deletions in those 50 bases. There could be underlying
mismatches in the read compared to the reference. Note: there is also
extended <code>CIGAR</code> that replaces <code>M</code> with <code>=</code> (exact match) and <code>X</code> (mismatch).
Also see Appendix D on the <code>MD</code> tag and finding where the mismatches are, but
note that <code>MD</code> tag is tricky</p>
</li>
<li>
<p><em>Ambiguity of representation</em>: A <code>CIGAR</code> string with insertions and deletions
could be <code>50M1D1I50M</code>. This string had a 1bp deletion and a 1bp insertion
back-to-back. This could be just a mismatch! There is ambiguity in sequence
alignment representations. Downstream programs must accomodate this.</p>
</li>
<li>
<p><em>Split records and soft-clipping</em>: A <code>CIGAR</code> string with soft-clipping
<code>500S50M</code> this means that 500 bases of the read were not aligned at this
position, but 50 bases were! Note that the alignment might have been a split
alignment (see section on split alignments below) so another record in the
<code>SAM</code> file, linked by the <code>SA</code> (supplementary alignment) tag, might contain
info on where the other 500 bases aligned! (or, they might not map anywhere).
The linked split or supplementary alignments all have the same read name
(<code>QNAME</code>).</p>
</li>
</ul>
<p>See <a href="https://samtools.github.io/hts-specs/SAMv1.pdf">SAMv1.pdf</a> for all the
CIGAR operators.</p>
<p>If you are working with <code>SAM</code> data, you will often write loops that directly
parse CIGAR strings. See Appendix B for handy functions for parsing <code>CIGAR</code>
strings. Don't fear the <code>CIGAR</code>!</p>
<h2 id="detecting-svs-from-long-reads"><a aria-hidden="true" tabindex="-1" href="#detecting-svs-from-long-reads"><a href="#detecting-svs-from-long-reads" style="margin-right: 10px">#</a></a>Detecting SVs from long reads</h2>
<p>Long reads offer a wide array of methods for detecting SVs</p>
<ul>
<li><em>Small insertions/deletions</em>: Long reads can completely span moderate sized
insertions and deletions, indicated by <code>I</code> or <code>D</code> in a <code>CIGAR</code> string.</li>
<li><em>Large insertions/deletions</em>: If a long read does not completely span an
insertion or deletion, it may be split aligned on either side of the SV or
could be soft/hard clipped where it can't align all the way through an
insertion.</li>
<li><em>Translocations</em>: A split long alignment can span long range or even
inter-chromosomal translocations, so part of the read maps to one chromosome
and one part maps to the other</li>
<li><em>Inversions</em>: A split alignment can span an inversion, the long read is split
into multiple parts, one part of it aligns in the reverse orientation, while
the other part aligns in the forward orientation</li>
</ul>
<p>Note that there are many different methods for detecting SVs from long reads,
e.g. not all use mapped reads from SAM files, some use <em>de novo</em> assembly, but
it is still useful to be familiar with mapped read methods.</p>
<h3 id="what-are-splitsupplementarychimeric-alignments"><a aria-hidden="true" tabindex="-1" href="#what-are-splitsupplementarychimeric-alignments"><a href="#what-are-splitsupplementarychimeric-alignments" style="margin-right: 10px">#</a></a>What are split/supplementary/chimeric alignments?</h3>
<p>Split alignments, or chimeric alignments, are alignments where part of the read
maps to one place, and another part to another. For example, part of a long
read may map to <code>chr1</code> and part of it maps to <code>chr4</code>. It is worth reading the
definition of "Chimeric alignment" from
<a href="https://samtools.github.io/hts-specs/SAMv1.pdf">SAMv1.pdf</a> when you get the
chance.</p>
<p>As <a href="https://samtools.github.io/hts-specs/SAMv1.pdf">SAMv1.pdf</a> tells us, one
record is marked as "representative", sometimes also called the "primary"
record, while the other components of the split read are marked
"supplementary", given the 2048 flag. The "primary" record generally has a
<code>SEQ</code> field that represents the entirety of the original read's sequence (with
CIGAR soft clipping operators saying which part of that sequence aligned), and
the "supplementary alignments" will have <code>SEQ</code> field but sometimes just
segments of the original read's sequence with CIGAR hard clipping operators
indicating that it is partial.</p>
<p>Supplementary alignments are especially common with long reads, and it can be a
signal for structural variants e.g. where two chromosomes are fused together,
and parts of the read align to multiple chromosomes, or the split alignment may
align to either side of a large deletion, or they may be split to align through
an inversion (part of it aligns to the forward strand, part of it to the
reverse strand, and again the forward strand)</p>
<p>There is no limitation on how many splits might occur so the split can align to
3, 4, or more different places. Each part of the split puts a new line in the
SAM file, and note that all the records also have the same read name, or
<code>QNAME</code> (first column of <code>SAM</code>).</p>
<h3 id="what-are-secondary-alignmentsmulti-mappers"><a aria-hidden="true" tabindex="-1" href="#what-are-secondary-alignmentsmulti-mappers"><a href="#what-are-secondary-alignmentsmulti-mappers" style="margin-right: 10px">#</a></a>What are secondary alignments/multi-mappers</h3>
<p>Secondary alignments generally come from "multi-mappers" where the entire read
maps equally well (or at least somewhat equally well) to, say, somewhere on
both <code>chr4</code> and <code>chr1</code>. "Multi mapping" results in secondary alignments, while
split reads result in supplementary alignments. See "Multiple mapping" in the
<a href="https://samtools.github.io/hts-specs/SAMv1.pdf">SAMv1.pdf</a> for the definition
of multi-mapping. Note also that secondary alignments sometimes are missing the
<code>SEQ</code> field entirely too, see
<a href="https://github.com/lh3/minimap2/issues/458#issuecomment-516661855">https://github.com/lh3/minimap2/issues/458#issuecomment-516661855</a></p>
<p>I wrote a tool called secondary_rewriter to add the <code>SEQ</code> field back to
secondary alignments, which may help in some cases</p>
<h3 id="what-is-the-sa-tag"><a aria-hidden="true" tabindex="-1" href="#what-is-the-sa-tag"><a href="#what-is-the-sa-tag" style="margin-right: 10px">#</a></a>What is the <code>SA</code> tag?</h3>
<p>The <code>SA</code> tag is outputted on each part of the supplementary/split/chimeric
alignment, e.g. the primary contains an <code>SA</code> tag that refers to information
(e.g. the location) of where all the supplementary alignments where placed, and
each of the supplementary alignments also contains an <code>SA</code> tag that refers to
the primary alignment and each other supplementary alignment.</p>
<p>Fun fact: The <code>SA</code> tag conceptually can result in a 'quadratic explosion' of
data, because each part of the split contains references to every other part.
For example, if a read is split into 4 pieces, then each record would would
have an <code>SA</code> tag with 3 segments, so 3*4 segments will be documented in the
<code>SA</code> tag. In many cases, this is not a problem, but if you imagine a finished
chromosome aligned to a draft assembly, it may get split so many times this
could be a factor.</p>
<p>See <a href="https://samtools.github.io/hts-specs/SAMtags.pdf">SAMtags.pdf</a> for more
info on the <code>SA</code> tag.</p>
<h3 id="visualizing-split-reads-across-a-breakend-or-translocation"><a aria-hidden="true" tabindex="-1" href="#visualizing-split-reads-across-a-breakend-or-translocation"><a href="#visualizing-split-reads-across-a-breakend-or-translocation" style="margin-right: 10px">#</a></a>Visualizing split reads across a breakend or translocation</h3>
<p>This is a specialized JBrowse 2 feature, but if there is an inter-chromosomal
translocation, you can load this into JBrowse and visualize support for this
event using our "breakpoint split view". This view shows the evidence for the
reads that are split aligned across an SV, and can show connections between
paired-end reads across an SV too.</p>
<p>We also have a workflow called the "SV inspector" that helps you setup the
"breakpoint split views"
(<a href="https://jbrowse.org/jb2/docs/user_guide/#sv-inspector">https://jbrowse.org/jb2/docs/user_guide/#sv-inspector</a>.
The SV inspector and Breakpoint split view work best on Breakends (e.g. VCF 4.3
section 5.4) and <code>&#x3C;TRA></code> (translocation) events from <code>VCF</code>, or <code>BEDPE</code>
formatted SV calls, and you can launch the "breakpoint split view" from the "SV
inspector"</p>
<p><img src="/media/breakpoint_split_view.png" alt=""></p>
<h3 id="visualizing-a-read-vs-reference-view-given-a-split-alignment"><a aria-hidden="true" tabindex="-1" href="#visualizing-a-read-vs-reference-view-given-a-split-alignment"><a href="#visualizing-a-read-vs-reference-view-given-a-split-alignment" style="margin-right: 10px">#</a></a>Visualizing a 'read vs reference' view given a split alignment</h3>
<p>If we are given the the primary alignment of an arbitrary split read, then we
can construct what that split looks like compared to the reference genome.</p>
<p>If we are not given the primary alignment (e.g. we are starting from a
supplementary alignment) then we can search the <code>SA</code> list for the one that is
primary, because at least one will be.</p>
<p>Now that we have the primary alignment, it will have the <code>SEQ</code> (of the entire
read, the supplementary alignments typically have a blank <code>SEQ</code>!) and the <code>SA</code>
tag containing the <code>CIGAR</code> of all the different parts of the split. We can then
construct how the entire read, not just a particular record of the split
alignment, compares to the genome. In JBrowse 2 we implemented this and it uses
a synteny-style rendering. [1]</p>
<p><img src="/media/linear_alignment.png" alt=""></p>
<p>Figure showing JBrowse 2 piecing together a long read vs the reference genome
from a single read</p>
<p>In order to do this reconstruction, JBrowse 2 takes the <code>CIGAR</code> strings of the
primary alignment and each of the pieces of the <code>SA</code> tag (it is a semi-colon
separated list of chunks), sort them by the amount of softclipping (the
softclipping values will progressively trim off more of the <code>SEQ</code> telling you
it aligned further and further on in the long read), and then this tells me
where each piece of the split alignment came from in the original <code>SEQ</code>, so we
can plot the alignments of the read vs the reference genome using synteny style
display.</p>
<p>[1] Similar functionality also exists in GenomeRibbon
<a href="https://genomeribbon.org">https://genomeribbon.org</a></p>
<h3 id="sam-vs-vcf---breakends-vs-split-alignments"><a aria-hidden="true" tabindex="-1" href="#sam-vs-vcf---breakends-vs-split-alignments"><a href="#sam-vs-vcf---breakends-vs-split-alignments" style="margin-right: 10px">#</a></a><code>SAM</code> vs <code>VCF</code> - Breakends vs split alignments</h3>
<p>An interesting outcome (to me) is that from a single record in a <code>SAM</code> file, I
can reconstruct the "derived" genome around a region of interest from a single
read.</p>
<p>If I was to try to do this with the <code>VCF</code> Breakend specification (section 5.4
of <a href="https://samtools.github.io/hts-specs/VCFv4.3.pdf">VCF4.3.pdf</a>), it may
actually be more challenging than from a <code>SAM</code> read. This is because a Breakend
in <code>VCF</code> format is only an edge in a graph (and the sequences are nodes).
Therefore, in order to properly reconstruct a structural variant from a <code>VCF</code>
with Breakends, I would have to construct a graph and decode paths through it.</p>
<p>I like the ability to reconstruct the derived genome from a single read, but
individual reads can be noisy (contain errors). That said, <em>de novo</em> assembled
contigs can also be stored in <code>SAM</code> format and is significantly less noisy
(being composed of the aggregate information of many reads).</p>
<p>The point though is that interpretation of the <code>VCF</code> breakend specification is
challenging due to imposing a sequence graph on the genome, while the <code>SA</code> tag
remains just a simple set of linear alignments that can easily be pieced
together, and you only need to refer to a single record in the <code>SAM</code> file to do
so.</p>
<p>I am not aware of a lot of tools that work on the <code>VCF</code> Breakend graph, and
expect more will need to be created to truly work with this standard. An
inversion for example may create 4 record in the <code>VCF</code> file (see section 5.4 in
the <a href="https://samtools.github.io/hts-specs/VCFv4.3.pdf">VCF4.3.pdf</a> for
example), and needs careful interpretation.</p>
<h3 id="haplotype-tagged-reads"><a aria-hidden="true" tabindex="-1" href="#haplotype-tagged-reads"><a href="#haplotype-tagged-reads" style="margin-right: 10px">#</a></a>Haplotype tagged reads</h3>
<p>A new trend has been to create <code>SAM</code>/<code>BAM</code>/<code>CRAM</code> files with tagged reads,
which tells us which haplotype a read was inferred to have come from. This is
commonly done with the <code>HP</code> tag, which might have <code>HP=0</code> and <code>HP=1</code> for a
diploid genome. Tools like <code>whatshap</code> can add these tags to a <code>SAM</code> file, and
IGV and JBrowse 2 can color and sort by these tags.</p>
<p><img src="/media/color_by_tag.png" alt=""></p>
<p>Screenshot of JBrowse 2 with the "Color by tag" and "Sort by tag" setting
enabled (coloring and sorting by the <code>HP</code> tag) letting us see that only one
haplotype has a deletion. Tutorial for how to do this in JBrowse 2 here
<a href="https://jbrowse.org/jb2/docs/user_guide/#sort-color-and-filter-by-tag">https://jbrowse.org/jb2/docs/user_guide/#sort-color-and-filter-by-tag</a></p>
<h2 id="how-do-you-detect-svs-with-paired-end-reads"><a aria-hidden="true" tabindex="-1" href="#how-do-you-detect-svs-with-paired-end-reads"><a href="#how-do-you-detect-svs-with-paired-end-reads" style="margin-right: 10px">#</a></a>How do you detect SVs with paired-end reads?</h2>
<p>Paired-end reads are short reads, e.g. 150bp each. This makes them unable to
recover some large structural variants.</p>
<p>However, paired-end reads have a number of attributes that can be used to
detect paired end reads</p>
<h3 id="distance-between-pairs-being-abnormally-large-or-short"><a aria-hidden="true" tabindex="-1" href="#distance-between-pairs-being-abnormally-large-or-short"><a href="#distance-between-pairs-being-abnormally-large-or-short" style="margin-right: 10px">#</a></a>Distance between pairs being abnormally large or short</h3>
<p>The distance between pairs is encoded by the <code>TLEN</code> column in the <code>SAM</code> format.
The distance between pairs with good mapping is relatively constant and called
the "insert length". This comes from how the sequencing is done: paired-end
sequencing performs sequencing on both ends of a fragment.</p>
<p>But, if you are mapping reads vs the reference genome, and you observe that
they are abnormally far apart, say 50kb apart instead of 1kb apart, this may
indicate there your sample contains a deletion relative to the reference.</p>
<p><img src="/media/aberrant_size.png" alt=""></p>
<p>Screenshot of JBrowse 1 with "View as pairs" enabled, and large insert size
colored as red (from
<a href="https://jbrowse.org/docs/paired_reads.html">https://jbrowse.org/docs/paired_reads.html</a>.
Note that some of JBrowse 1's View as pairs features are not yet available in
JBrowse 2</p>
<h3 id="an-abundance-of-reads-being-clipped-at-a-particular-position"><a aria-hidden="true" tabindex="-1" href="#an-abundance-of-reads-being-clipped-at-a-particular-position"><a href="#an-abundance-of-reads-being-clipped-at-a-particular-position" style="margin-right: 10px">#</a></a>An abundance of reads being "clipped" at a particular position</h3>
<p>This can indicate that part of the reads map well, but then there was an abrupt
stop to the mapping. This might mean that there is a sequence that was an
insertion at that position, or a deletion, or a translocation.</p>
<p>The clipping is indicated by the <code>CIGAR</code> string, either at the start or end of
it by an <code>S</code> or an <code>H</code>. The <code>S</code> indicates "soft clipping", and indicates that
the sequence of the clipped portion can be found in the <code>SEQ</code> field of the
primary alignment. The <code>H</code> is hard clipped, and the sequence that is hard
clipped will not appear in the <code>SEQ</code>.</p>
<p><img src="/media/clipping_pileup.png" alt=""></p>
<p>Screenshot of JBrowse 2 showing blue clipping indicator with a "pileup" of
soft-clipping at a particular position shown in blue. The clipping is an
"interbase" operation (it occurs between base pair coordinates) so it is
plotted separately from the normal coverage histogram.</p>
<p><img src="/media/show_soft_clipping.png" alt=""></p>
<p>Screenshot of JBrowse 2 showing an insertion with Nanopore (top), PacBio
(middle) and Illumina short reads. The long reads may completely span the
insertion, so the <code>CIGAR</code> string on those have an <code>I</code> operator and are indicated
by the purple triangle above the reads. For the short reads, the reads near the
insertion will be clipped since they will not properly map to the reference
genome and cannot span the sinsertion. The "Show soft clipping" setting in
JBrowse 2 and IGV can be used to show visually the bases that extend into the
insertion (shown on the bottom track).</p>
<h3 id="unexpected-pair-orientation"><a aria-hidden="true" tabindex="-1" href="#unexpected-pair-orientation"><a href="#unexpected-pair-orientation" style="margin-right: 10px">#</a></a>Unexpected pair orientation</h3>
<p>With standard paired end sequencing, the pairs normally point at each other</p>
<pre><code>forward reverse
 --->    &#x3C;---
</code></pre>
<p>If the stranded-ness of the pair is off, then it could indicate a structural
variant. See Appendix A for a handy function for calculating pair orientation.</p>
<p>This guide from IGV is helpful for interpreting the pair directionality with
patterns of SVs using "Color by pair orientation"</p>
<p><a href="https://software.broadinstitute.org/software/igv/interpreting_pair_orientations">https://software.broadinstitute.org/software/igv/interpreting_pair_orientations</a></p>
<p><img src="/media/inverted_duplication.png" alt=""></p>
<p>Figure: JBrowse 2 showing an inverted (tandem) duplication in 1000 genomes
data. It uses the same coloring as IGV for pair orientation. The tandem
duplication can produce green arrows which have reads pointing in opposite
directions e.g. <code>&#x3C;--</code> and <code>--></code>, while blue arrows which can indicate an
inversion point in the same direction e.g. <code>--></code> and <code>--></code></p>
<h3 id="caveat-about-tlen"><a aria-hidden="true" tabindex="-1" href="#caveat-about-tlen"><a href="#caveat-about-tlen" style="margin-right: 10px">#</a></a>Caveat about TLEN</h3>
<p>Note that <code>TLEN</code> is a field in the SAM format that is somewhat ill defined,
at least in the sense that different tools may use it differently
<a href="https://github.com/pysam-developers/pysam/issues/667#issuecomment-381521767">https://github.com/pysam-developers/pysam/issues/667#issuecomment-381521767</a></p>
<p>If needed, you can calculate <code>TLEN</code> yourself if you process the file yourself
(e.g. process all reads, get the actual records for the pairs, and calculate
distance) but I have not had trouble with relying on the <code>TLEN</code> from the data
files themselves.</p>
<h2 id="calling-copy-number-variants-with-your-short-or-long-reads"><a aria-hidden="true" tabindex="-1" href="#calling-copy-number-variants-with-your-short-or-long-reads"><a href="#calling-copy-number-variants-with-your-short-or-long-reads" style="margin-right: 10px">#</a></a>Calling copy number variants with your short or long reads</h2>
<p>Another type of SV that you can get from your <code>SAM</code> files are copy number
variants (CNVs). By looking at the depth-of-coverage for your data files, you
can look for abnormalities that may indicate copy number variants. By using a
tool like <code>mosdepth</code>, you can quickly get a file showing the coverage across
the genome.</p>
<p>Be aware that if you are comparing the coverage counts from different tools,
that they have different defaults that may affect comparison. Some discard
<code>QC_FAIL</code>, <code>DUP</code>, and <code>SECONDARY</code> flagged reads. This is probably appropriate,
and corresponds to what most genome browsers will display (see
<a href="https://gist.github.com/cmdcolin/9f677eca28448d8a7c5d6e9917fc56af">https://gist.github.com/cmdcolin/9f677eca28448d8a7c5d6e9917fc56af</a> for a short
summary of depth calculated from different tools)</p>
<p>Note that both long and short reads can be used for CNV detection. Long reads
may give more accurate measurements also, with their better ability to map
smoothly through difficult regions of the genome.</p>
<p><img src="/media/coverage_cnv.png" alt=""></p>
<p>Screenshot showing coverage in <code>BigWig</code> format from nanopore reads on normal
and tumor tissue from a melanoma cancer cell line (COLO829) plotted using
JBrowse 2. This coverage data is calculated from nanopore sequencing from
<a href="https://www.biorxiv.org/content/10.1101/2020.10.15.340497v1.full">here</a> using
<a href="https://github.com/brentp/mosdepth"><code>mosdepth</code></a>, converted from <code>BedGraph</code> to
<code>BigWig</code>, and loaded into JBrowse 2. See
(<a href="https://jbrowse.org/code/jb2/v1.6.4/?config=test_data%2Fconfig_demo.json&#x26;session=share-MZj3d18lzH&#x26;password=3X7bS">demo</a>
and
<a href="https://jbrowse.org/jb2/docs/user_guide/#viewing-whole-genome-coverage-for-profiling-cnv">tutorial</a>)</p>
<h2 id="the-future-with-graph-genomes-and-de-novo-assemblies"><a aria-hidden="true" tabindex="-1" href="#the-future-with-graph-genomes-and-de-novo-assemblies"><a href="#the-future-with-graph-genomes-and-de-novo-assemblies" style="margin-right: 10px">#</a></a>The future, with graph genomes and <em>de novo</em> assemblies</h2>
<p>Currently, SV visualization is highly based on comparing data versus a
reference genome (and the <code>SAM</code> format is a signature of this: it stores data
in terms of reference genome coordinates). In the future, SV visualization may
look more similar to comparative genomics, where we compare an SV to a
population specific reference from a graph genomes or something like this.</p>
<p>It is known that <em>de novo</em> assembly has more power to detect SVs than some read
operations (<a href="https://twitter.com/lh3lh3/status/1362921612690010118/photo/1">https://twitter.com/lh3lh3/status/1362921612690010118/photo/1</a>
as <em>de novo</em> assembled genomes improve and become more widespread, we may see a
shift in how SVs are called</p>
<p>I would also like to see improved ability to do fast or 'on the fly' gene
prediction on the <em>de novo</em> assembled genomes, and we can see what SNPs or
modified splicing might look like in copies of genes (e.g. derived regions of
the CNV duplications).</p>
<p>Fun fact: the
<a href="https://github.com/lh3/gfatools/blob/master/doc/rGFA.md#the-graph-alignment-format-gaf"><code>GAF</code></a>
(graphical alignment format) is a strict superset of
<a href="https://github.com/lh3/miniasm/blob/master/PAF.md"><code>PAF</code></a> (pairwise alignment
format) by storing graph node labels in the <code>target name</code> slot of <code>PAF</code>, and
can refer to an <code>rGFA</code> (reference genome graph)! Looking forward to the graph
genome world.</p>
<h2 id="conclusion"><a aria-hidden="true" tabindex="-1" href="#conclusion"><a href="#conclusion" style="margin-right: 10px">#</a></a>Conclusion</h2>
<p>Algorithms that actually call structural variants face many challenges, but
understanding how the reads are encoded in SAM format, and seeing what they
look like in the genome browser is a useful first step to gaining a better
understanding.</p>
<p>In summary, some of the signatures of SVs may include:</p>
<ul>
<li>Aberrant insert size (<code>TLEN</code>) detection (longer for deletion, shorter for
insertion)</li>
<li>Aberrant pair orientation (pairs are not pointing at each other)</li>
<li>Split-read detection (<code>SA</code> tag)</li>
<li><code>CIGAR</code> string processing (<code>D</code> operator for deletions, <code>I</code> operator for
insertions)</li>
<li>Over-abundance of clipping (<code>S</code> or <code>H</code> operators in <code>CIGAR</code>)</li>
<li>Depth of coverage changes for CNVs</li>
<li>Aligning <em>de novo</em> assembly vs a reference genome
(<a href="https://twitter.com/lh3lh3/status/1362921612690010118/photo/1">https://twitter.com/lh3lh3/status/1362921612690010118/photo/1</a>) which can
output <code>SAM</code>, but it can also output
<a href="https://github.com/lh3/miniasm/blob/master/PAF.md"><code>PAF</code></a> format (which can
be loaded in JBrowse 2 in the synteny views). Techniques of detecting SVs on
PAF will be fundamentally pretty similar to the techniques listed above but
may look a bit different (see <code>cs</code> tag in <code>PAF</code> for example, it is a modified
<code>CIGAR</code>-like string)</li>
</ul>
<p>If you have any ideas I should include here, let me know!</p>
<h3 id="appendix-a-parsing-cigar-strings"><a aria-hidden="true" tabindex="-1" href="#appendix-a-parsing-cigar-strings"><a href="#appendix-a-parsing-cigar-strings" style="margin-right: 10px">#</a></a>Appendix A: Parsing <code>CIGAR</code> strings</h3>
<p>This is code that can help determine the pair orientation from a single BAM
record. Might be too much detail but follow along</p>
<div class="highlight highlight-ts"><pre><span class="pl-c">// @param flags - flags from a single read</span>
<span class="pl-c">// @param ref - the string of the reference sequence, just used to determine if it matches rnext</span>
<span class="pl-c">// @param rnext - the string of the RNEXT, just used to determine if it matches ref</span>
<span class="pl-c">// @param tlen - the TLEN field from SAM</span>
<span class="pl-c">// @return e.g. F1R2 normal paired end orientation</span>
<span class="pl-k">function</span> <span class="pl-en">getPairOrientation</span>(
  <span class="pl-v">flags</span><span class="pl-k">:</span> <span class="pl-c1">number</span>,
  <span class="pl-v">ref</span><span class="pl-k">:</span> <span class="pl-c1">string</span>,
  <span class="pl-v">rnext</span><span class="pl-k">:</span> <span class="pl-c1">string</span>,
  <span class="pl-v">tlen</span><span class="pl-k">:</span> <span class="pl-c1">number</span>,
) {
  <span class="pl-c">// this read is not unmapped &#x26;&#x26;</span>
  <span class="pl-c">// this read's mate is also not unmapped &#x26;&#x26;</span>
  <span class="pl-c">// this read's mate is on the same reference genome</span>
  <span class="pl-k">if</span> (<span class="pl-k">!</span><span class="pl-smi">flags</span> <span class="pl-k">&#x26;</span> <span class="pl-c1">4</span> <span class="pl-k">&#x26;&#x26;</span> <span class="pl-k">!</span><span class="pl-smi">flags</span> <span class="pl-k">&#x26;</span> <span class="pl-c1">8</span> <span class="pl-k">&#x26;&#x26;</span> <span class="pl-smi">ref</span> <span class="pl-k">===</span> <span class="pl-smi">rnext</span>) {
    <span class="pl-k">const</span> <span class="pl-c1">s1</span> <span class="pl-k">=</span> <span class="pl-smi">flags</span> <span class="pl-k">&#x26;</span> <span class="pl-c1">16</span> <span class="pl-k">?</span> <span class="pl-s"><span class="pl-pds">'</span>R<span class="pl-pds">'</span></span> <span class="pl-k">:</span> <span class="pl-s"><span class="pl-pds">'</span>F<span class="pl-pds">'</span></span>
    <span class="pl-k">const</span> <span class="pl-c1">s2</span> <span class="pl-k">=</span> <span class="pl-smi">flags</span> <span class="pl-k">&#x26;</span> <span class="pl-c1">32</span> <span class="pl-k">?</span> <span class="pl-s"><span class="pl-pds">'</span>R<span class="pl-pds">'</span></span> <span class="pl-k">:</span> <span class="pl-s"><span class="pl-pds">'</span>F<span class="pl-pds">'</span></span>
    <span class="pl-k">let</span> <span class="pl-smi">o1</span> <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">'</span> <span class="pl-pds">'</span></span>
    <span class="pl-k">let</span> <span class="pl-smi">o2</span> <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">'</span> <span class="pl-pds">'</span></span>

    <span class="pl-c">// if first in pair</span>
    <span class="pl-k">if</span> (<span class="pl-smi">flags</span> <span class="pl-k">&#x26;</span> <span class="pl-c1">64</span>) {
      <span class="pl-smi">o1</span> <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">'</span>1<span class="pl-pds">'</span></span>
      <span class="pl-smi">o2</span> <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">'</span>2<span class="pl-pds">'</span></span>
    }

    <span class="pl-c">// else if second in pair</span>
    <span class="pl-k">else</span> <span class="pl-k">if</span> (<span class="pl-smi">flags</span> <span class="pl-k">&#x26;</span> <span class="pl-c1">128</span>) {
      <span class="pl-smi">o1</span> <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">'</span>2<span class="pl-pds">'</span></span>
      <span class="pl-smi">o2</span> <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">'</span>1<span class="pl-pds">'</span></span>
    }

    <span class="pl-k">const</span> <span class="pl-c1">tmp</span> <span class="pl-k">=</span> []
    <span class="pl-k">if</span> (<span class="pl-smi">tlen</span> <span class="pl-k">></span> <span class="pl-c1">0</span>) {
      <span class="pl-smi">tmp</span>[<span class="pl-c1">0</span>] <span class="pl-k">=</span> <span class="pl-smi">s1</span>
      <span class="pl-smi">tmp</span>[<span class="pl-c1">1</span>] <span class="pl-k">=</span> <span class="pl-smi">o1</span>
      <span class="pl-smi">tmp</span>[<span class="pl-c1">2</span>] <span class="pl-k">=</span> <span class="pl-smi">s2</span>
      <span class="pl-smi">tmp</span>[<span class="pl-c1">3</span>] <span class="pl-k">=</span> <span class="pl-smi">o2</span>
    } <span class="pl-k">else</span> {
      <span class="pl-smi">tmp</span>[<span class="pl-c1">2</span>] <span class="pl-k">=</span> <span class="pl-smi">s1</span>
      <span class="pl-smi">tmp</span>[<span class="pl-c1">3</span>] <span class="pl-k">=</span> <span class="pl-smi">o1</span>
      <span class="pl-smi">tmp</span>[<span class="pl-c1">0</span>] <span class="pl-k">=</span> <span class="pl-smi">s2</span>
      <span class="pl-smi">tmp</span>[<span class="pl-c1">1</span>] <span class="pl-k">=</span> <span class="pl-smi">o2</span>
    }
    <span class="pl-k">return</span> <span class="pl-smi">tmp</span>.<span class="pl-c1">join</span>(<span class="pl-s"><span class="pl-pds">''</span></span>)
  }
  <span class="pl-k">return</span> <span class="pl-c1">null</span>
}
</pre></div>
<p>Then this can be broken down further by orientation type</p>
<p>Paired end reads are "fr"
Mate pair reads are "rf"</p>
<p>So you can interpret e.g. F1R2 in relation to being a paired end read (fr) or mate pair (rf) below and with this link <a href="https://software.broadinstitute.org/software/igv/interpreting_pair_orientations">https://software.broadinstitute.org/software/igv/interpreting_pair_orientations</a></p>
<div class="highlight highlight-json"><pre>{
  <span class="pl-ent">"fr"</span>: {
    <span class="pl-ent">"F1R2"</span>: <span class="pl-s"><span class="pl-pds">"</span>LR<span class="pl-pds">"</span></span>,
    <span class="pl-ent">"F2R1"</span>: <span class="pl-s"><span class="pl-pds">"</span>LR<span class="pl-pds">"</span></span>,

    <span class="pl-ent">"F1F2"</span>: <span class="pl-s"><span class="pl-pds">"</span>LL<span class="pl-pds">"</span></span>,
    <span class="pl-ent">"F2F1"</span>: <span class="pl-s"><span class="pl-pds">"</span>LL<span class="pl-pds">"</span></span>,

    <span class="pl-ent">"R1R2"</span>: <span class="pl-s"><span class="pl-pds">"</span>RR<span class="pl-pds">"</span></span>,
    <span class="pl-ent">"R2R1"</span>: <span class="pl-s"><span class="pl-pds">"</span>RR<span class="pl-pds">"</span></span>,

    <span class="pl-ent">"R1F2"</span>: <span class="pl-s"><span class="pl-pds">"</span>RL<span class="pl-pds">"</span></span>,
    <span class="pl-ent">"R2F1"</span>: <span class="pl-s"><span class="pl-pds">"</span>RL<span class="pl-pds">"</span></span>
  },

  <span class="pl-ent">"rf"</span>: {
    <span class="pl-ent">"R1F2"</span>: <span class="pl-s"><span class="pl-pds">"</span>LR<span class="pl-pds">"</span></span>,
    <span class="pl-ent">"R2F1"</span>: <span class="pl-s"><span class="pl-pds">"</span>LR<span class="pl-pds">"</span></span>,

    <span class="pl-ent">"R1R2"</span>: <span class="pl-s"><span class="pl-pds">"</span>LL<span class="pl-pds">"</span></span>,
    <span class="pl-ent">"R2R1"</span>: <span class="pl-s"><span class="pl-pds">"</span>LL<span class="pl-pds">"</span></span>,

    <span class="pl-ent">"F1F2"</span>: <span class="pl-s"><span class="pl-pds">"</span>RR<span class="pl-pds">"</span></span>,
    <span class="pl-ent">"F2F1"</span>: <span class="pl-s"><span class="pl-pds">"</span>RR<span class="pl-pds">"</span></span>,

    <span class="pl-ent">"F1R2"</span>: <span class="pl-s"><span class="pl-pds">"</span>RL<span class="pl-pds">"</span></span>,
    <span class="pl-ent">"F2R1"</span>: <span class="pl-s"><span class="pl-pds">"</span>RL<span class="pl-pds">"</span></span>
  }
}
</pre></div>
<h3 id="appendix-b---cigar-parsing"><a aria-hidden="true" tabindex="-1" href="#appendix-b---cigar-parsing"><a href="#appendix-b---cigar-parsing" style="margin-right: 10px">#</a></a>Appendix B - <code>CIGAR</code> parsing</h3>
<div class="highlight highlight-ts"><pre><span class="pl-c">// @param cigar: CIGAR string in text form</span>
<span class="pl-c">// @returns an array of elements like ['30','M', '2','I', '50','M', '40','D']</span>
<span class="pl-c">// which you can consume in a loop two elements at a time</span>
<span class="pl-k">function</span> <span class="pl-en">parseCigar</span>(<span class="pl-v">cigar</span><span class="pl-k">:</span> <span class="pl-c1">string</span>) {
  <span class="pl-k">return</span> <span class="pl-smi">cigar</span>.<span class="pl-c1">split</span>(<span class="pl-s"><span class="pl-sr"><span class="pl-pds">/</span>(<span class="pl-c1">[MIDNSHPX=]</span>)<span class="pl-pds">/</span></span></span>)
}
</pre></div>
<p>Then parse the returned array two at a time</p>
<div class="highlight highlight-ts"><pre><span class="pl-c">// this function does nothing, but is informative for how to parse interpret a</span>
<span class="pl-c">// CIGAR string</span>
<span class="pl-c">// @param cigar:CIGAR string from record</span>
<span class="pl-c">// @param readSeq: the SEQ from record</span>
<span class="pl-c">// @param refSeq: the reference sequence underlying the read</span>
<span class="pl-k">function</span> <span class="pl-en">interpretCigar</span>(<span class="pl-v">cigar</span><span class="pl-k">:</span> <span class="pl-c1">string</span>, <span class="pl-v">readSeq</span><span class="pl-k">:</span> <span class="pl-c1">string</span>, <span class="pl-v">refSeq</span><span class="pl-k">:</span> <span class="pl-c1">string</span>) {
  <span class="pl-k">const</span> <span class="pl-c1">opts</span> <span class="pl-k">=</span> <span class="pl-en">parseCigar</span>(<span class="pl-smi">cigar</span>)
  <span class="pl-k">let</span> <span class="pl-smi">qpos</span> <span class="pl-k">=</span> <span class="pl-c1">0</span> <span class="pl-c">// query position, position on the read</span>
  <span class="pl-k">let</span> <span class="pl-smi">tpos</span> <span class="pl-k">=</span> <span class="pl-c1">0</span> <span class="pl-c">// target position, position on the reference sequence</span>
  <span class="pl-k">for</span> (<span class="pl-k">let</span> <span class="pl-smi">i</span> <span class="pl-k">=</span> <span class="pl-c1">0</span>; <span class="pl-smi">i</span> <span class="pl-k">&#x3C;</span> <span class="pl-smi">ops</span>.<span class="pl-c1">length</span>; <span class="pl-smi">i</span> <span class="pl-k">+=</span> <span class="pl-c1">2</span>) {
    <span class="pl-k">const</span> <span class="pl-c1">length</span> <span class="pl-k">=</span> <span class="pl-k">+</span><span class="pl-smi">opts</span>[<span class="pl-smi">i</span>]
    <span class="pl-k">const</span> <span class="pl-c1">operator</span> <span class="pl-k">=</span> <span class="pl-smi">opts</span>[<span class="pl-smi">i</span> <span class="pl-k">+</span> <span class="pl-c1">1</span>]
    <span class="pl-c">// do things. refer to the CIGAR chart in SAMv1.pdf for which operators</span>
    <span class="pl-c">// "consume reference" to see whether to increment</span>
    <span class="pl-k">if</span> (<span class="pl-smi">op</span> <span class="pl-k">===</span> <span class="pl-s"><span class="pl-pds">'</span>M<span class="pl-pds">'</span></span> <span class="pl-k">||</span> <span class="pl-smi">op</span> <span class="pl-k">===</span> <span class="pl-s"><span class="pl-pds">'</span>=<span class="pl-pds">'</span></span>) {
      <span class="pl-c">// matches consume query and reference</span>
      <span class="pl-k">const</span> <span class="pl-c1">refMatch</span> <span class="pl-k">=</span> <span class="pl-smi">refSeq</span>.<span class="pl-c1">slice</span>(<span class="pl-smi">tpos</span>, <span class="pl-smi">tpos</span> <span class="pl-k">+</span> <span class="pl-smi">len</span>)
      <span class="pl-k">const</span> <span class="pl-c1">readMatch</span> <span class="pl-k">=</span> <span class="pl-smi">readSeq</span>.<span class="pl-c1">slice</span>(<span class="pl-smi">qpos</span>, <span class="pl-smi">qpos</span> <span class="pl-k">+</span> <span class="pl-smi">len</span>)
      <span class="pl-k">for</span> (<span class="pl-k">let</span> <span class="pl-smi">i</span> <span class="pl-k">=</span> <span class="pl-c1">0</span>; <span class="pl-smi">i</span> <span class="pl-k">&#x3C;</span> <span class="pl-smi">len</span>; <span class="pl-smi">i</span><span class="pl-k">++</span>) {
        <span class="pl-k">if</span> (<span class="pl-smi">refMatch</span>[<span class="pl-smi">i</span>] <span class="pl-k">!==</span> <span class="pl-smi">readMatch</span>[<span class="pl-smi">i</span>]) {
          <span class="pl-c">// SNP at this position</span>
        }
      }
      <span class="pl-smi">qpos</span> <span class="pl-k">+=</span> <span class="pl-smi">len</span>
      <span class="pl-smi">tpos</span> <span class="pl-k">+=</span> <span class="pl-smi">len</span>
    }
    <span class="pl-k">if</span> (<span class="pl-smi">op</span> <span class="pl-k">===</span> <span class="pl-s"><span class="pl-pds">'</span>I<span class="pl-pds">'</span></span>) {
      <span class="pl-c">// insertions only consume query</span>
      <span class="pl-c">// sequence of the insertion from the read is</span>
      <span class="pl-k">const</span> <span class="pl-c1">insSeq</span> <span class="pl-k">=</span> <span class="pl-smi">readSeq</span>.<span class="pl-c1">slice</span>(<span class="pl-smi">qpos</span>, <span class="pl-smi">qpos</span> <span class="pl-k">+</span> <span class="pl-smi">len</span>)
      <span class="pl-smi">qpos</span> <span class="pl-k">+=</span> <span class="pl-smi">len</span>
    }
    <span class="pl-k">if</span> (<span class="pl-smi">op</span> <span class="pl-k">===</span> <span class="pl-s"><span class="pl-pds">'</span>D<span class="pl-pds">'</span></span>) {
      <span class="pl-c">// deletions only consume reference</span>
      <span class="pl-c">// sequence of the deletion from the reference is</span>
      <span class="pl-k">const</span> <span class="pl-c1">delSeq</span> <span class="pl-k">=</span> <span class="pl-smi">refSeq</span>.<span class="pl-c1">slice</span>(<span class="pl-smi">tpos</span>, <span class="pl-smi">tpos</span> <span class="pl-k">+</span> <span class="pl-smi">len</span>)
      <span class="pl-smi">tpad</span> <span class="pl-k">+=</span> <span class="pl-smi">len</span>
    }
    <span class="pl-k">if</span> (<span class="pl-smi">op</span> <span class="pl-k">===</span> <span class="pl-s"><span class="pl-pds">'</span>N<span class="pl-pds">'</span></span>) {
      <span class="pl-c">// skips only consume reference</span>
      <span class="pl-c">// skips are similar to deletions but are related to spliced alignments</span>
      <span class="pl-smi">tpad</span> <span class="pl-k">+=</span> <span class="pl-smi">len</span>
    }
    <span class="pl-k">if</span> (<span class="pl-smi">op</span> <span class="pl-k">===</span> <span class="pl-s"><span class="pl-pds">'</span>X<span class="pl-pds">'</span></span>) {
      <span class="pl-c">// mismatch using the extended CIGAR format</span>
      <span class="pl-c">// could lookup the mismatch letter in a string containing the reference</span>
      <span class="pl-k">const</span> <span class="pl-c1">mismatch</span> <span class="pl-k">=</span> <span class="pl-smi">refSeq</span>.<span class="pl-c1">slice</span>(<span class="pl-smi">tpos</span>, <span class="pl-smi">tpos</span> <span class="pl-k">+</span> <span class="pl-smi">len</span>)
      <span class="pl-smi">qpos</span> <span class="pl-k">+=</span> <span class="pl-smi">len</span>
      <span class="pl-smi">tpos</span> <span class="pl-k">+=</span> <span class="pl-smi">len</span>
    }
    <span class="pl-k">if</span> (<span class="pl-smi">op</span> <span class="pl-k">===</span> <span class="pl-s"><span class="pl-pds">'</span>H<span class="pl-pds">'</span></span>) {
      <span class="pl-c">// does not consume query or reference</span>
      <span class="pl-c">// hardclip is just an indicator</span>
    }
    <span class="pl-k">if</span> (<span class="pl-smi">op</span> <span class="pl-k">===</span> <span class="pl-s"><span class="pl-pds">'</span>S<span class="pl-pds">'</span></span>) {
      <span class="pl-c">// softclip consumes query</span>
      <span class="pl-c">// below gets the entire soft clipped portion</span>
      <span class="pl-k">const</span> <span class="pl-c1">softClipStr</span> <span class="pl-k">=</span> <span class="pl-smi">readSeq</span>.<span class="pl-c1">slice</span>(<span class="pl-smi">qpos</span>, <span class="pl-smi">qpos</span> <span class="pl-k">+</span> <span class="pl-smi">len</span>)
      <span class="pl-smi">qpos</span> <span class="pl-k">+=</span> <span class="pl-smi">len</span>
    }
  }
}
</pre></div>
<p>Note for example, that to determine how long a record is on the reference
sequence, you have to combine the records start position with the CIGAR string,
basically parsing the CIGAR string to add up tpos and return tpos</p>
<h3 id="appendix-c---align-fastq-directly-to-cram"><a aria-hidden="true" tabindex="-1" href="#appendix-c---align-fastq-directly-to-cram"><a href="#appendix-c---align-fastq-directly-to-cram" style="margin-right: 10px">#</a></a>Appendix C - align <code>FASTQ</code> directly to <code>CRAM</code></h3>
<p>This example from the htslib documentation
(<a href="http://www.htslib.org/workflow/fastq.html">http://www.htslib.org/workflow/fastq.html</a>
shows how you can stream directly from <code>FASTQ</code> to <code>CRAM</code> (and generate the
index file .crai too)</p>
<p>If you want, you can make this a little shell script, easy_align_shortreads.sh</p>
<p>easy_align_shortreads.sh</p>
<div class="highlight highlight-shell"><pre><span class="pl-c">#!/bin/bash</span>
minimap2 -t 8 -a -x sr <span class="pl-s"><span class="pl-pds">"</span><span class="pl-smi">$1</span><span class="pl-pds">"</span></span> <span class="pl-s"><span class="pl-pds">"</span><span class="pl-smi">$2</span><span class="pl-pds">"</span></span> <span class="pl-s"><span class="pl-pds">"</span><span class="pl-smi">$3</span><span class="pl-pds">"</span></span>  <span class="pl-k">|</span> \
samtools fixmate -u -m - - <span class="pl-k">|</span> \
samtools sort -u -@2 - <span class="pl-k">|</span> \
samtools markdup -@8 --reference <span class="pl-s"><span class="pl-pds">"</span><span class="pl-smi">$1</span><span class="pl-pds">"</span></span> - --write-index <span class="pl-s"><span class="pl-pds">"</span><span class="pl-smi">$4</span><span class="pl-pds">"</span></span>
</pre></div>
<p>Similar idea for longreads, except just a single fastq file is generally used for longreads</p>
<p>easy_align_longreads.sh</p>
<div class="highlight highlight-shell"><pre><span class="pl-c">#!/bin/bash</span>
minimap2 -t 8 -a <span class="pl-s"><span class="pl-pds">"</span><span class="pl-smi">$1</span><span class="pl-pds">"</span></span> <span class="pl-s"><span class="pl-pds">"</span><span class="pl-smi">$2</span><span class="pl-pds">"</span></span>  <span class="pl-k">|</span> \
samtools fixmate -u -m - - <span class="pl-k">|</span> \
samtools sort -u -@2 - <span class="pl-k">|</span> \
samtools markdup -@8 --reference <span class="pl-s"><span class="pl-pds">"</span><span class="pl-smi">$1</span><span class="pl-pds">"</span></span> - --write-index <span class="pl-s"><span class="pl-pds">"</span><span class="pl-smi">$3</span><span class="pl-pds">"</span></span>
</pre></div>
<p>Then call</p>
<div class="highlight highlight-shell"><pre>bash easy_align_shortreads.sh ref.fa reads1.fq reads2.fq out.cram
bash easy_align_longreads.sh ref.fa reads.fq out.cram

<span class="pl-c">## output BAM instead</span>
bash easy_align_shortreads.sh ref.fa reads1.fq reads2.fq out.bam
bash easy_align_longreads.sh ref.fa reads.fq out.bam
</pre></div>
<p>This same concept works with other common aligners as well like bwa</p>
<p>Bonus: CRAM to bigwig, for looking at CNV/coverage</p>
<div class="highlight highlight-shell"><pre><span class="pl-c">#!/bin/bash</span>
<span class="pl-c"># quickalign.sh ref.fa 1.fq 2.fq out.cram</span>
<span class="pl-c"># produces out.cram and out.bw</span>
samtools faidx <span class="pl-smi">$1</span>
minimap2 -t 8 -a -x sr <span class="pl-s"><span class="pl-pds">"</span><span class="pl-smi">$1</span><span class="pl-pds">"</span></span> <span class="pl-s"><span class="pl-pds">"</span><span class="pl-smi">$2</span><span class="pl-pds">"</span></span> <span class="pl-s"><span class="pl-pds">"</span><span class="pl-smi">$3</span><span class="pl-pds">"</span></span>  <span class="pl-k">|</span> \
samtools fixmate -u -m - - <span class="pl-k">|</span> \
samtools sort -u -@2 - <span class="pl-k">|</span> \
samtools markdup -@8 --reference <span class="pl-s"><span class="pl-pds">"</span><span class="pl-smi">$1</span><span class="pl-pds">"</span></span> - --write-index <span class="pl-s"><span class="pl-pds">"</span><span class="pl-smi">$4</span><span class="pl-pds">"</span></span>


mosdepth <span class="pl-smi">$4</span> -f <span class="pl-smi">$1</span> <span class="pl-smi">$4</span>
gunzip <span class="pl-smi">$4</span>.per-base.bed.gz
bedGraphToBigWig <span class="pl-smi">$4</span>.per-base.bed <span class="pl-smi">$1</span>.fa.fai <span class="pl-smi">$4</span>.bw
</pre></div>
<p>Call as "quickalign.sh ref.fa 1.fq 2.fq out.cram" gives you out.cram, out.cram.crai, and out.cram.bw (coverage)</p>
<h3 id="appendix-d---the-md-tag-and-finding-snps-in-reads"><a aria-hidden="true" tabindex="-1" href="#appendix-d---the-md-tag-and-finding-snps-in-reads"><a href="#appendix-d---the-md-tag-and-finding-snps-in-reads" style="margin-right: 10px">#</a></a>Appendix D - the <code>MD</code> tag and finding SNPs in reads</h3>
<p>The <code>MD</code> tag helps tell you where the mismatches are without looking at the
reference genome. This is useful because as I mentioned, <code>CIGAR</code> can say <code>50M</code>
(50 matches) but some letters inside those 50 matches can be mismatches, it
only says there are no insertions/deletions in those 50 bases, but you have to
determine where in those 50 bases where the mismatches are. The <code>MD</code> tag can
help tell you where those are, but it is somewhat complicated to decode
(<a href="https://vincebuffalo.com/notes/2014/01/17/md-tags-in-bam-files.html">https://vincebuffalo.com/notes/2014/01/17/md-tags-in-bam-files.html</a>).
You have to combine it with the <code>CIGAR</code> to get the position of the mismatches
on the reference genome. If you have a reference genome to look at, you might
just compare all the bases within the 50M to the reference genome and look for
mismatches yourself and forget about the <code>MD</code> tag</p>
<p>The <code>MD</code> tag is also not required to exist, but the command <code>samtools calmd yourfile.bam --reference reference.fa</code> can add <code>MD</code> tags to your <code>BAM</code> file. It
is generally not useful for <code>CRAM</code> because <code>CRAM</code> actually does store
mismatches with the reference genome in it's compression format. Note that
there are also some oddities about <code>MD</code> tag representation leading to
complaints (e.g. <a href="https://github.com/samtools/hts-specs/issues/505">https://github.com/samtools/hts-specs/issues/505</a>) leading more
credence to "doing it yourself" e.g. finding your own mismatches
by comparing the read sequence with the reference, keeping track of where you
are on the read and ref position with the <code>CIGAR</code> string (a la Appendix B)</p></div><div style="margin-top:200px"></div></article></div></main></div><footer style="margin-top:100px"><a href="/">Home</a> <a href="/archive">Blog archive</a> <a href="https://github.com/cmdcolin/">Github</a> <a href="https://twitter.com/cmdcolin">Twitter</a> <a href="/projects">Projects</a> <a href="/photos">Photos</a> <a href="/rss.xml">RSS</a><a href="/about">About</a> </footer></div><script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{"post":{"title":"Structural variants and the SAM format - the long (reads) and short (reads) of it","date":"2022-02-06","slug":"2022-02-06-sv-sam","html":"\u003cp\u003eThe \u003ccode\u003eSAM\u003c/code\u003e specification is pretty amazing\n(\u003ca href=\"https://samtools.github.io/hts-specs/SAMv1.pdf\"\u003ehttps://samtools.github.io/hts-specs/SAMv1.pdf\u003c/a\u003e) but it is also fairly terse\nand abstract. True understanding might come from playing with real world data.\nI will try to relay some things I have learned over the years, with a bit of a\nfocus on how \u003ccode\u003eSAM\u003c/code\u003e file concepts can relate to structural variants.\u003c/p\u003e\n\u003cp\u003eDisclaimer: I'm a developer of JBrowse 2. This document has some screenshots\nand links for it, feel free to try it at \u003ca href=\"https://jbrowse.org\"\u003ehttps://jbrowse.org\u003c/a\u003e.\u003c/p\u003e\n\u003ch2 id=\"basics\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#basics\"\u003e\u003ca href=\"#basics\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eBasics\u003c/h2\u003e\n\u003ch3 id=\"what-is-a-sam-file-and-how-does-it-relate-to-bam-and-cram\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#what-is-a-sam-file-and-how-does-it-relate-to-bam-and-cram\"\u003e\u003ca href=\"#what-is-a-sam-file-and-how-does-it-relate-to-bam-and-cram\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eWhat is a \u003ccode\u003eSAM\u003c/code\u003e file and how does it relate to \u003ccode\u003eBAM\u003c/code\u003e and \u003ccode\u003eCRAM\u003c/code\u003e?\u003c/h3\u003e\n\u003cul\u003e\n\u003cli\u003e\n\u003cp\u003eA \u003ccode\u003eSAM\u003c/code\u003e file \u003cstrong\u003egenerally\u003c/strong\u003e contains \"reads\" from a sequencer, with information\nabout how they are mapped to a reference genome [1][2].\u003c/p\u003e\n\u003c/li\u003e\n\u003cli\u003e\n\u003cp\u003eA \u003ccode\u003eSAM\u003c/code\u003e file is \u003cstrong\u003egenerally\u003c/strong\u003e produced when an aligner takes in raw unaligned reads\n(often stored in \u003ccode\u003eFASTQ\u003c/code\u003e format files) and aligns them to a reference genome [3].\u003c/p\u003e\n\u003c/li\u003e\n\u003cli\u003e\n\u003cp\u003eA \u003ccode\u003eSAM\u003c/code\u003e file is a text format that you can read with your text editor. \u003ccode\u003eBAM\u003c/code\u003e and\n\u003ccode\u003eCRAM\u003c/code\u003e are compressed representations of the \u003ccode\u003eSAM\u003c/code\u003e format.\u003c/p\u003e\n\u003c/li\u003e\n\u003c/ul\u003e\n\u003cp\u003eYou can convert \u003ccode\u003eSAM\u003c/code\u003e to \u003ccode\u003eBAM\u003c/code\u003e with samtools\u003c/p\u003e\n\u003cdiv class=\"highlight highlight-shell\"\u003e\u003cpre\u003esamtools view file.sam -o file.bam\n\u003c/pre\u003e\u003c/div\u003e\n\u003cp\u003eYou can also convert a \u003ccode\u003eBAM\u003c/code\u003e back to \u003ccode\u003eSAM\u003c/code\u003e with samtools view\u003c/p\u003e\n\u003cdiv class=\"highlight highlight-shell\"\u003e\u003cpre\u003esamtools view -h file.bam -o file.sam\n\u003c/pre\u003e\u003c/div\u003e\n\u003cp\u003eThe -h just makes sure to preserve the header.\u003c/p\u003e\n\u003cp\u003eIf you are converting \u003ccode\u003eSAM\u003c/code\u003e to \u003ccode\u003eCRAM\u003c/code\u003e, it may require the -T argument to\nspecify your reference sequence (this is because the \u003ccode\u003eCRAM\u003c/code\u003e is \"reference\ncompressed\")\u003c/p\u003e\n\u003cdiv class=\"highlight highlight-shell\"\u003e\u003cpre\u003esamtools view -T reference.fa file.sam -o file.cram\n\u003c/pre\u003e\u003c/div\u003e\n\u003cp\u003eAlso see Appendix C: piping FASTQ from \u003ccode\u003eminimap2\u003c/code\u003e directly to CRAM\u003c/p\u003e\n\u003cp\u003e[1] \u003ccode\u003eSAM\u003c/code\u003e can contain any type of sequence, not specifically reads. If you\ncreated a \u003cem\u003ede novo\u003c/em\u003e assembly, you could align the contigs of the \u003cem\u003ede novo\u003c/em\u003e assembly\nto a reference genome and store the results in \u003ccode\u003eSAM\u003c/code\u003e.\u003c/p\u003e\n\u003cp\u003e[2] Does not always have to have information about mapping to a reference\ngenome. You can also store unaligned data in \u003ccode\u003eSAM\u003c/code\u003e/\u003ccode\u003eBAM\u003c/code\u003e/\u003ccode\u003eCRAM\u003c/code\u003e (so-called\n\u003ccode\u003euBAM\u003c/code\u003e for example) but most of the time, the reads in \u003ccode\u003eSAM\u003c/code\u003e format are aligned\nto a reference genome.\u003c/p\u003e\n\u003cp\u003e[3] Examples of programs that do alignment include \u003ccode\u003ebwa\u003c/code\u003e, \u003ccode\u003ebowtie\u003c/code\u003e, and \u003ccode\u003eminimap2\u003c/code\u003e\n(there are many others). These programs all can produce \u003ccode\u003eSAM\u003c/code\u003e outputs\u003c/p\u003e\n\u003ch3 id=\"what-is-in-a-sam-file\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#what-is-in-a-sam-file\"\u003e\u003ca href=\"#what-is-in-a-sam-file\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eWhat is in a \u003ccode\u003eSAM\u003c/code\u003e file\u003c/h3\u003e\n\u003cp\u003eA \u003ccode\u003eSAM\u003c/code\u003e file contains a header (\u003ccode\u003eBAM\u003c/code\u003e and \u003ccode\u003eCRAM\u003c/code\u003e files also have the \u003ccode\u003eSAM\u003c/code\u003e\nheader) and a series of records. A record is a single line in a \u003ccode\u003eSAM\u003c/code\u003e file, and\nit generally corresponds to a single read, but as we will see, a split\nalignment may produce multiple records that refer to the same source read.\u003c/p\u003e\n\u003cp\u003eNote: if a read failed to align to the reference genome, it may still be in\nyour \u003ccode\u003eSAM\u003c/code\u003e file, marked as unmapped using the flag column. Sometimes, \"dumpster\ndiving\" (looking at the unmapped records from a \u003ccode\u003eSAM\u003c/code\u003e file) can be used to aid\nstructural variant searches (e.g. there may be novel sequence in there not from\nthe reference genome that could be assembled)\u003c/p\u003e\n\u003ch3 id=\"what-are-tags-in-a-sam-file\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#what-are-tags-in-a-sam-file\"\u003e\u003ca href=\"#what-are-tags-in-a-sam-file\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eWhat are tags in a \u003ccode\u003eSAM\u003c/code\u003e file\u003c/h3\u003e\n\u003cp\u003eA SAM file has a core set of required fields, and then an arbitrary list of\nextra columns called tags. The tags have a two character abbreviation like \u003ccode\u003eMQ\u003c/code\u003e\n(mapping quality) or many others. They can be upper or lower case. Upper case\nare reserved for official usages (except those with X, Y, or Z prefixed). See\n\u003ca href=\"https://samtools.github.io/hts-specs/SAMtags.pdf\"\u003eSAMtags.pdf\u003c/a\u003e for more\ndetails\u003c/p\u003e\n\u003ch3 id=\"what-is-a-cigar-string\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#what-is-a-cigar-string\"\u003e\u003ca href=\"#what-is-a-cigar-string\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eWhat is a \u003ccode\u003eCIGAR\u003c/code\u003e string\u003c/h3\u003e\n\u003cp\u003eA \u003ccode\u003eCIGAR\u003c/code\u003e string is a \"compact idiosyncratic gapped alignment report\". It tells\nyou about insertions, deletions, and clipping. It is a series of \"operators\"\nwith lengths.\u003c/p\u003e\n\u003cp\u003eInsertion example:\u003c/p\u003e\n\u003cp\u003e\u003ccode\u003e50M50I50M\u003c/code\u003e\u003c/p\u003e\n\u003cp\u003eThat would be 50bp of matching bases (\u003ccode\u003e50M\u003c/code\u003e), followed by a 50bp insertion\n(\u003ccode\u003e50I\u003c/code\u003e), followed by another 50bp of matches (\u003ccode\u003e50M\u003c/code\u003e). The 50bp insertion means\nthe read contains 50 bases in the middle which did not match the reference\ngenome that you are comparing the read to.\u003c/p\u003e\n\u003cp\u003eClipping example:\u003c/p\u003e\n\u003cp\u003e\u003ccode\u003e50S50M50S\u003c/code\u003e\u003c/p\u003e\n\u003cp\u003eThis means that 50bp matched (\u003ccode\u003e50M\u003c/code\u003e in the middle of the \u003ccode\u003eCIGAR\u003c/code\u003e string) and\nboth sides of the read are soft clipped. The clipping means the aligner was not\nable to align the reads on either side. You could imagine clipping being like\n\"an insertion on either side of the read\" if you like: basically those bases on\neither side did not align.\u003c/p\u003e\n\u003cp\u003eNotes:\u003c/p\u003e\n\u003cul\u003e\n\u003cli\u003e\n\u003cp\u003e\u003cem\u003eFinding mismatches\u003c/em\u003e: A \u003ccode\u003eCIGAR\u003c/code\u003e string match like \u003ccode\u003e50M\u003c/code\u003e means 50 bases\n\"matched\" the reference genome, but that only means that there are no\ninsertions or deletions in those 50 bases. There could be underlying\nmismatches in the read compared to the reference. Note: there is also\nextended \u003ccode\u003eCIGAR\u003c/code\u003e that replaces \u003ccode\u003eM\u003c/code\u003e with \u003ccode\u003e=\u003c/code\u003e (exact match) and \u003ccode\u003eX\u003c/code\u003e (mismatch).\nAlso see Appendix D on the \u003ccode\u003eMD\u003c/code\u003e tag and finding where the mismatches are, but\nnote that \u003ccode\u003eMD\u003c/code\u003e tag is tricky\u003c/p\u003e\n\u003c/li\u003e\n\u003cli\u003e\n\u003cp\u003e\u003cem\u003eAmbiguity of representation\u003c/em\u003e: A \u003ccode\u003eCIGAR\u003c/code\u003e string with insertions and deletions\ncould be \u003ccode\u003e50M1D1I50M\u003c/code\u003e. This string had a 1bp deletion and a 1bp insertion\nback-to-back. This could be just a mismatch! There is ambiguity in sequence\nalignment representations. Downstream programs must accomodate this.\u003c/p\u003e\n\u003c/li\u003e\n\u003cli\u003e\n\u003cp\u003e\u003cem\u003eSplit records and soft-clipping\u003c/em\u003e: A \u003ccode\u003eCIGAR\u003c/code\u003e string with soft-clipping\n\u003ccode\u003e500S50M\u003c/code\u003e this means that 500 bases of the read were not aligned at this\nposition, but 50 bases were! Note that the alignment might have been a split\nalignment (see section on split alignments below) so another record in the\n\u003ccode\u003eSAM\u003c/code\u003e file, linked by the \u003ccode\u003eSA\u003c/code\u003e (supplementary alignment) tag, might contain\ninfo on where the other 500 bases aligned! (or, they might not map anywhere).\nThe linked split or supplementary alignments all have the same read name\n(\u003ccode\u003eQNAME\u003c/code\u003e).\u003c/p\u003e\n\u003c/li\u003e\n\u003c/ul\u003e\n\u003cp\u003eSee \u003ca href=\"https://samtools.github.io/hts-specs/SAMv1.pdf\"\u003eSAMv1.pdf\u003c/a\u003e for all the\nCIGAR operators.\u003c/p\u003e\n\u003cp\u003eIf you are working with \u003ccode\u003eSAM\u003c/code\u003e data, you will often write loops that directly\nparse CIGAR strings. See Appendix B for handy functions for parsing \u003ccode\u003eCIGAR\u003c/code\u003e\nstrings. Don't fear the \u003ccode\u003eCIGAR\u003c/code\u003e!\u003c/p\u003e\n\u003ch2 id=\"detecting-svs-from-long-reads\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#detecting-svs-from-long-reads\"\u003e\u003ca href=\"#detecting-svs-from-long-reads\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eDetecting SVs from long reads\u003c/h2\u003e\n\u003cp\u003eLong reads offer a wide array of methods for detecting SVs\u003c/p\u003e\n\u003cul\u003e\n\u003cli\u003e\u003cem\u003eSmall insertions/deletions\u003c/em\u003e: Long reads can completely span moderate sized\ninsertions and deletions, indicated by \u003ccode\u003eI\u003c/code\u003e or \u003ccode\u003eD\u003c/code\u003e in a \u003ccode\u003eCIGAR\u003c/code\u003e string.\u003c/li\u003e\n\u003cli\u003e\u003cem\u003eLarge insertions/deletions\u003c/em\u003e: If a long read does not completely span an\ninsertion or deletion, it may be split aligned on either side of the SV or\ncould be soft/hard clipped where it can't align all the way through an\ninsertion.\u003c/li\u003e\n\u003cli\u003e\u003cem\u003eTranslocations\u003c/em\u003e: A split long alignment can span long range or even\ninter-chromosomal translocations, so part of the read maps to one chromosome\nand one part maps to the other\u003c/li\u003e\n\u003cli\u003e\u003cem\u003eInversions\u003c/em\u003e: A split alignment can span an inversion, the long read is split\ninto multiple parts, one part of it aligns in the reverse orientation, while\nthe other part aligns in the forward orientation\u003c/li\u003e\n\u003c/ul\u003e\n\u003cp\u003eNote that there are many different methods for detecting SVs from long reads,\ne.g. not all use mapped reads from SAM files, some use \u003cem\u003ede novo\u003c/em\u003e assembly, but\nit is still useful to be familiar with mapped read methods.\u003c/p\u003e\n\u003ch3 id=\"what-are-splitsupplementarychimeric-alignments\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#what-are-splitsupplementarychimeric-alignments\"\u003e\u003ca href=\"#what-are-splitsupplementarychimeric-alignments\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eWhat are split/supplementary/chimeric alignments?\u003c/h3\u003e\n\u003cp\u003eSplit alignments, or chimeric alignments, are alignments where part of the read\nmaps to one place, and another part to another. For example, part of a long\nread may map to \u003ccode\u003echr1\u003c/code\u003e and part of it maps to \u003ccode\u003echr4\u003c/code\u003e. It is worth reading the\ndefinition of \"Chimeric alignment\" from\n\u003ca href=\"https://samtools.github.io/hts-specs/SAMv1.pdf\"\u003eSAMv1.pdf\u003c/a\u003e when you get the\nchance.\u003c/p\u003e\n\u003cp\u003eAs \u003ca href=\"https://samtools.github.io/hts-specs/SAMv1.pdf\"\u003eSAMv1.pdf\u003c/a\u003e tells us, one\nrecord is marked as \"representative\", sometimes also called the \"primary\"\nrecord, while the other components of the split read are marked\n\"supplementary\", given the 2048 flag. The \"primary\" record generally has a\n\u003ccode\u003eSEQ\u003c/code\u003e field that represents the entirety of the original read's sequence (with\nCIGAR soft clipping operators saying which part of that sequence aligned), and\nthe \"supplementary alignments\" will have \u003ccode\u003eSEQ\u003c/code\u003e field but sometimes just\nsegments of the original read's sequence with CIGAR hard clipping operators\nindicating that it is partial.\u003c/p\u003e\n\u003cp\u003eSupplementary alignments are especially common with long reads, and it can be a\nsignal for structural variants e.g. where two chromosomes are fused together,\nand parts of the read align to multiple chromosomes, or the split alignment may\nalign to either side of a large deletion, or they may be split to align through\nan inversion (part of it aligns to the forward strand, part of it to the\nreverse strand, and again the forward strand)\u003c/p\u003e\n\u003cp\u003eThere is no limitation on how many splits might occur so the split can align to\n3, 4, or more different places. Each part of the split puts a new line in the\nSAM file, and note that all the records also have the same read name, or\n\u003ccode\u003eQNAME\u003c/code\u003e (first column of \u003ccode\u003eSAM\u003c/code\u003e).\u003c/p\u003e\n\u003ch3 id=\"what-are-secondary-alignmentsmulti-mappers\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#what-are-secondary-alignmentsmulti-mappers\"\u003e\u003ca href=\"#what-are-secondary-alignmentsmulti-mappers\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eWhat are secondary alignments/multi-mappers\u003c/h3\u003e\n\u003cp\u003eSecondary alignments generally come from \"multi-mappers\" where the entire read\nmaps equally well (or at least somewhat equally well) to, say, somewhere on\nboth \u003ccode\u003echr4\u003c/code\u003e and \u003ccode\u003echr1\u003c/code\u003e. \"Multi mapping\" results in secondary alignments, while\nsplit reads result in supplementary alignments. See \"Multiple mapping\" in the\n\u003ca href=\"https://samtools.github.io/hts-specs/SAMv1.pdf\"\u003eSAMv1.pdf\u003c/a\u003e for the definition\nof multi-mapping. Note also that secondary alignments sometimes are missing the\n\u003ccode\u003eSEQ\u003c/code\u003e field entirely too, see\n\u003ca href=\"https://github.com/lh3/minimap2/issues/458#issuecomment-516661855\"\u003ehttps://github.com/lh3/minimap2/issues/458#issuecomment-516661855\u003c/a\u003e\u003c/p\u003e\n\u003cp\u003eI wrote a tool called secondary_rewriter to add the \u003ccode\u003eSEQ\u003c/code\u003e field back to\nsecondary alignments, which may help in some cases\u003c/p\u003e\n\u003ch3 id=\"what-is-the-sa-tag\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#what-is-the-sa-tag\"\u003e\u003ca href=\"#what-is-the-sa-tag\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eWhat is the \u003ccode\u003eSA\u003c/code\u003e tag?\u003c/h3\u003e\n\u003cp\u003eThe \u003ccode\u003eSA\u003c/code\u003e tag is outputted on each part of the supplementary/split/chimeric\nalignment, e.g. the primary contains an \u003ccode\u003eSA\u003c/code\u003e tag that refers to information\n(e.g. the location) of where all the supplementary alignments where placed, and\neach of the supplementary alignments also contains an \u003ccode\u003eSA\u003c/code\u003e tag that refers to\nthe primary alignment and each other supplementary alignment.\u003c/p\u003e\n\u003cp\u003eFun fact: The \u003ccode\u003eSA\u003c/code\u003e tag conceptually can result in a 'quadratic explosion' of\ndata, because each part of the split contains references to every other part.\nFor example, if a read is split into 4 pieces, then each record would would\nhave an \u003ccode\u003eSA\u003c/code\u003e tag with 3 segments, so 3*4 segments will be documented in the\n\u003ccode\u003eSA\u003c/code\u003e tag. In many cases, this is not a problem, but if you imagine a finished\nchromosome aligned to a draft assembly, it may get split so many times this\ncould be a factor.\u003c/p\u003e\n\u003cp\u003eSee \u003ca href=\"https://samtools.github.io/hts-specs/SAMtags.pdf\"\u003eSAMtags.pdf\u003c/a\u003e for more\ninfo on the \u003ccode\u003eSA\u003c/code\u003e tag.\u003c/p\u003e\n\u003ch3 id=\"visualizing-split-reads-across-a-breakend-or-translocation\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#visualizing-split-reads-across-a-breakend-or-translocation\"\u003e\u003ca href=\"#visualizing-split-reads-across-a-breakend-or-translocation\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eVisualizing split reads across a breakend or translocation\u003c/h3\u003e\n\u003cp\u003eThis is a specialized JBrowse 2 feature, but if there is an inter-chromosomal\ntranslocation, you can load this into JBrowse and visualize support for this\nevent using our \"breakpoint split view\". This view shows the evidence for the\nreads that are split aligned across an SV, and can show connections between\npaired-end reads across an SV too.\u003c/p\u003e\n\u003cp\u003eWe also have a workflow called the \"SV inspector\" that helps you setup the\n\"breakpoint split views\"\n(\u003ca href=\"https://jbrowse.org/jb2/docs/user_guide/#sv-inspector\"\u003ehttps://jbrowse.org/jb2/docs/user_guide/#sv-inspector\u003c/a\u003e.\nThe SV inspector and Breakpoint split view work best on Breakends (e.g. VCF 4.3\nsection 5.4) and \u003ccode\u003e\u0026#x3C;TRA\u003e\u003c/code\u003e (translocation) events from \u003ccode\u003eVCF\u003c/code\u003e, or \u003ccode\u003eBEDPE\u003c/code\u003e\nformatted SV calls, and you can launch the \"breakpoint split view\" from the \"SV\ninspector\"\u003c/p\u003e\n\u003cp\u003e\u003cimg src=\"/media/breakpoint_split_view.png\" alt=\"\"\u003e\u003c/p\u003e\n\u003ch3 id=\"visualizing-a-read-vs-reference-view-given-a-split-alignment\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#visualizing-a-read-vs-reference-view-given-a-split-alignment\"\u003e\u003ca href=\"#visualizing-a-read-vs-reference-view-given-a-split-alignment\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eVisualizing a 'read vs reference' view given a split alignment\u003c/h3\u003e\n\u003cp\u003eIf we are given the the primary alignment of an arbitrary split read, then we\ncan construct what that split looks like compared to the reference genome.\u003c/p\u003e\n\u003cp\u003eIf we are not given the primary alignment (e.g. we are starting from a\nsupplementary alignment) then we can search the \u003ccode\u003eSA\u003c/code\u003e list for the one that is\nprimary, because at least one will be.\u003c/p\u003e\n\u003cp\u003eNow that we have the primary alignment, it will have the \u003ccode\u003eSEQ\u003c/code\u003e (of the entire\nread, the supplementary alignments typically have a blank \u003ccode\u003eSEQ\u003c/code\u003e!) and the \u003ccode\u003eSA\u003c/code\u003e\ntag containing the \u003ccode\u003eCIGAR\u003c/code\u003e of all the different parts of the split. We can then\nconstruct how the entire read, not just a particular record of the split\nalignment, compares to the genome. In JBrowse 2 we implemented this and it uses\na synteny-style rendering. [1]\u003c/p\u003e\n\u003cp\u003e\u003cimg src=\"/media/linear_alignment.png\" alt=\"\"\u003e\u003c/p\u003e\n\u003cp\u003eFigure showing JBrowse 2 piecing together a long read vs the reference genome\nfrom a single read\u003c/p\u003e\n\u003cp\u003eIn order to do this reconstruction, JBrowse 2 takes the \u003ccode\u003eCIGAR\u003c/code\u003e strings of the\nprimary alignment and each of the pieces of the \u003ccode\u003eSA\u003c/code\u003e tag (it is a semi-colon\nseparated list of chunks), sort them by the amount of softclipping (the\nsoftclipping values will progressively trim off more of the \u003ccode\u003eSEQ\u003c/code\u003e telling you\nit aligned further and further on in the long read), and then this tells me\nwhere each piece of the split alignment came from in the original \u003ccode\u003eSEQ\u003c/code\u003e, so we\ncan plot the alignments of the read vs the reference genome using synteny style\ndisplay.\u003c/p\u003e\n\u003cp\u003e[1] Similar functionality also exists in GenomeRibbon\n\u003ca href=\"https://genomeribbon.org\"\u003ehttps://genomeribbon.org\u003c/a\u003e\u003c/p\u003e\n\u003ch3 id=\"sam-vs-vcf---breakends-vs-split-alignments\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#sam-vs-vcf---breakends-vs-split-alignments\"\u003e\u003ca href=\"#sam-vs-vcf---breakends-vs-split-alignments\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003e\u003ccode\u003eSAM\u003c/code\u003e vs \u003ccode\u003eVCF\u003c/code\u003e - Breakends vs split alignments\u003c/h3\u003e\n\u003cp\u003eAn interesting outcome (to me) is that from a single record in a \u003ccode\u003eSAM\u003c/code\u003e file, I\ncan reconstruct the \"derived\" genome around a region of interest from a single\nread.\u003c/p\u003e\n\u003cp\u003eIf I was to try to do this with the \u003ccode\u003eVCF\u003c/code\u003e Breakend specification (section 5.4\nof \u003ca href=\"https://samtools.github.io/hts-specs/VCFv4.3.pdf\"\u003eVCF4.3.pdf\u003c/a\u003e), it may\nactually be more challenging than from a \u003ccode\u003eSAM\u003c/code\u003e read. This is because a Breakend\nin \u003ccode\u003eVCF\u003c/code\u003e format is only an edge in a graph (and the sequences are nodes).\nTherefore, in order to properly reconstruct a structural variant from a \u003ccode\u003eVCF\u003c/code\u003e\nwith Breakends, I would have to construct a graph and decode paths through it.\u003c/p\u003e\n\u003cp\u003eI like the ability to reconstruct the derived genome from a single read, but\nindividual reads can be noisy (contain errors). That said, \u003cem\u003ede novo\u003c/em\u003e assembled\ncontigs can also be stored in \u003ccode\u003eSAM\u003c/code\u003e format and is significantly less noisy\n(being composed of the aggregate information of many reads).\u003c/p\u003e\n\u003cp\u003eThe point though is that interpretation of the \u003ccode\u003eVCF\u003c/code\u003e breakend specification is\nchallenging due to imposing a sequence graph on the genome, while the \u003ccode\u003eSA\u003c/code\u003e tag\nremains just a simple set of linear alignments that can easily be pieced\ntogether, and you only need to refer to a single record in the \u003ccode\u003eSAM\u003c/code\u003e file to do\nso.\u003c/p\u003e\n\u003cp\u003eI am not aware of a lot of tools that work on the \u003ccode\u003eVCF\u003c/code\u003e Breakend graph, and\nexpect more will need to be created to truly work with this standard. An\ninversion for example may create 4 record in the \u003ccode\u003eVCF\u003c/code\u003e file (see section 5.4 in\nthe \u003ca href=\"https://samtools.github.io/hts-specs/VCFv4.3.pdf\"\u003eVCF4.3.pdf\u003c/a\u003e for\nexample), and needs careful interpretation.\u003c/p\u003e\n\u003ch3 id=\"haplotype-tagged-reads\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#haplotype-tagged-reads\"\u003e\u003ca href=\"#haplotype-tagged-reads\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eHaplotype tagged reads\u003c/h3\u003e\n\u003cp\u003eA new trend has been to create \u003ccode\u003eSAM\u003c/code\u003e/\u003ccode\u003eBAM\u003c/code\u003e/\u003ccode\u003eCRAM\u003c/code\u003e files with tagged reads,\nwhich tells us which haplotype a read was inferred to have come from. This is\ncommonly done with the \u003ccode\u003eHP\u003c/code\u003e tag, which might have \u003ccode\u003eHP=0\u003c/code\u003e and \u003ccode\u003eHP=1\u003c/code\u003e for a\ndiploid genome. Tools like \u003ccode\u003ewhatshap\u003c/code\u003e can add these tags to a \u003ccode\u003eSAM\u003c/code\u003e file, and\nIGV and JBrowse 2 can color and sort by these tags.\u003c/p\u003e\n\u003cp\u003e\u003cimg src=\"/media/color_by_tag.png\" alt=\"\"\u003e\u003c/p\u003e\n\u003cp\u003eScreenshot of JBrowse 2 with the \"Color by tag\" and \"Sort by tag\" setting\nenabled (coloring and sorting by the \u003ccode\u003eHP\u003c/code\u003e tag) letting us see that only one\nhaplotype has a deletion. Tutorial for how to do this in JBrowse 2 here\n\u003ca href=\"https://jbrowse.org/jb2/docs/user_guide/#sort-color-and-filter-by-tag\"\u003ehttps://jbrowse.org/jb2/docs/user_guide/#sort-color-and-filter-by-tag\u003c/a\u003e\u003c/p\u003e\n\u003ch2 id=\"how-do-you-detect-svs-with-paired-end-reads\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#how-do-you-detect-svs-with-paired-end-reads\"\u003e\u003ca href=\"#how-do-you-detect-svs-with-paired-end-reads\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eHow do you detect SVs with paired-end reads?\u003c/h2\u003e\n\u003cp\u003ePaired-end reads are short reads, e.g. 150bp each. This makes them unable to\nrecover some large structural variants.\u003c/p\u003e\n\u003cp\u003eHowever, paired-end reads have a number of attributes that can be used to\ndetect paired end reads\u003c/p\u003e\n\u003ch3 id=\"distance-between-pairs-being-abnormally-large-or-short\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#distance-between-pairs-being-abnormally-large-or-short\"\u003e\u003ca href=\"#distance-between-pairs-being-abnormally-large-or-short\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eDistance between pairs being abnormally large or short\u003c/h3\u003e\n\u003cp\u003eThe distance between pairs is encoded by the \u003ccode\u003eTLEN\u003c/code\u003e column in the \u003ccode\u003eSAM\u003c/code\u003e format.\nThe distance between pairs with good mapping is relatively constant and called\nthe \"insert length\". This comes from how the sequencing is done: paired-end\nsequencing performs sequencing on both ends of a fragment.\u003c/p\u003e\n\u003cp\u003eBut, if you are mapping reads vs the reference genome, and you observe that\nthey are abnormally far apart, say 50kb apart instead of 1kb apart, this may\nindicate there your sample contains a deletion relative to the reference.\u003c/p\u003e\n\u003cp\u003e\u003cimg src=\"/media/aberrant_size.png\" alt=\"\"\u003e\u003c/p\u003e\n\u003cp\u003eScreenshot of JBrowse 1 with \"View as pairs\" enabled, and large insert size\ncolored as red (from\n\u003ca href=\"https://jbrowse.org/docs/paired_reads.html\"\u003ehttps://jbrowse.org/docs/paired_reads.html\u003c/a\u003e.\nNote that some of JBrowse 1's View as pairs features are not yet available in\nJBrowse 2\u003c/p\u003e\n\u003ch3 id=\"an-abundance-of-reads-being-clipped-at-a-particular-position\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#an-abundance-of-reads-being-clipped-at-a-particular-position\"\u003e\u003ca href=\"#an-abundance-of-reads-being-clipped-at-a-particular-position\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eAn abundance of reads being \"clipped\" at a particular position\u003c/h3\u003e\n\u003cp\u003eThis can indicate that part of the reads map well, but then there was an abrupt\nstop to the mapping. This might mean that there is a sequence that was an\ninsertion at that position, or a deletion, or a translocation.\u003c/p\u003e\n\u003cp\u003eThe clipping is indicated by the \u003ccode\u003eCIGAR\u003c/code\u003e string, either at the start or end of\nit by an \u003ccode\u003eS\u003c/code\u003e or an \u003ccode\u003eH\u003c/code\u003e. The \u003ccode\u003eS\u003c/code\u003e indicates \"soft clipping\", and indicates that\nthe sequence of the clipped portion can be found in the \u003ccode\u003eSEQ\u003c/code\u003e field of the\nprimary alignment. The \u003ccode\u003eH\u003c/code\u003e is hard clipped, and the sequence that is hard\nclipped will not appear in the \u003ccode\u003eSEQ\u003c/code\u003e.\u003c/p\u003e\n\u003cp\u003e\u003cimg src=\"/media/clipping_pileup.png\" alt=\"\"\u003e\u003c/p\u003e\n\u003cp\u003eScreenshot of JBrowse 2 showing blue clipping indicator with a \"pileup\" of\nsoft-clipping at a particular position shown in blue. The clipping is an\n\"interbase\" operation (it occurs between base pair coordinates) so it is\nplotted separately from the normal coverage histogram.\u003c/p\u003e\n\u003cp\u003e\u003cimg src=\"/media/show_soft_clipping.png\" alt=\"\"\u003e\u003c/p\u003e\n\u003cp\u003eScreenshot of JBrowse 2 showing an insertion with Nanopore (top), PacBio\n(middle) and Illumina short reads. The long reads may completely span the\ninsertion, so the \u003ccode\u003eCIGAR\u003c/code\u003e string on those have an \u003ccode\u003eI\u003c/code\u003e operator and are indicated\nby the purple triangle above the reads. For the short reads, the reads near the\ninsertion will be clipped since they will not properly map to the reference\ngenome and cannot span the sinsertion. The \"Show soft clipping\" setting in\nJBrowse 2 and IGV can be used to show visually the bases that extend into the\ninsertion (shown on the bottom track).\u003c/p\u003e\n\u003ch3 id=\"unexpected-pair-orientation\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#unexpected-pair-orientation\"\u003e\u003ca href=\"#unexpected-pair-orientation\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eUnexpected pair orientation\u003c/h3\u003e\n\u003cp\u003eWith standard paired end sequencing, the pairs normally point at each other\u003c/p\u003e\n\u003cpre\u003e\u003ccode\u003eforward reverse\n ---\u003e    \u0026#x3C;---\n\u003c/code\u003e\u003c/pre\u003e\n\u003cp\u003eIf the stranded-ness of the pair is off, then it could indicate a structural\nvariant. See Appendix A for a handy function for calculating pair orientation.\u003c/p\u003e\n\u003cp\u003eThis guide from IGV is helpful for interpreting the pair directionality with\npatterns of SVs using \"Color by pair orientation\"\u003c/p\u003e\n\u003cp\u003e\u003ca href=\"https://software.broadinstitute.org/software/igv/interpreting_pair_orientations\"\u003ehttps://software.broadinstitute.org/software/igv/interpreting_pair_orientations\u003c/a\u003e\u003c/p\u003e\n\u003cp\u003e\u003cimg src=\"/media/inverted_duplication.png\" alt=\"\"\u003e\u003c/p\u003e\n\u003cp\u003eFigure: JBrowse 2 showing an inverted (tandem) duplication in 1000 genomes\ndata. It uses the same coloring as IGV for pair orientation. The tandem\nduplication can produce green arrows which have reads pointing in opposite\ndirections e.g. \u003ccode\u003e\u0026#x3C;--\u003c/code\u003e and \u003ccode\u003e--\u003e\u003c/code\u003e, while blue arrows which can indicate an\ninversion point in the same direction e.g. \u003ccode\u003e--\u003e\u003c/code\u003e and \u003ccode\u003e--\u003e\u003c/code\u003e\u003c/p\u003e\n\u003ch3 id=\"caveat-about-tlen\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#caveat-about-tlen\"\u003e\u003ca href=\"#caveat-about-tlen\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eCaveat about TLEN\u003c/h3\u003e\n\u003cp\u003eNote that \u003ccode\u003eTLEN\u003c/code\u003e is a field in the SAM format that is somewhat ill defined,\nat least in the sense that different tools may use it differently\n\u003ca href=\"https://github.com/pysam-developers/pysam/issues/667#issuecomment-381521767\"\u003ehttps://github.com/pysam-developers/pysam/issues/667#issuecomment-381521767\u003c/a\u003e\u003c/p\u003e\n\u003cp\u003eIf needed, you can calculate \u003ccode\u003eTLEN\u003c/code\u003e yourself if you process the file yourself\n(e.g. process all reads, get the actual records for the pairs, and calculate\ndistance) but I have not had trouble with relying on the \u003ccode\u003eTLEN\u003c/code\u003e from the data\nfiles themselves.\u003c/p\u003e\n\u003ch2 id=\"calling-copy-number-variants-with-your-short-or-long-reads\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#calling-copy-number-variants-with-your-short-or-long-reads\"\u003e\u003ca href=\"#calling-copy-number-variants-with-your-short-or-long-reads\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eCalling copy number variants with your short or long reads\u003c/h2\u003e\n\u003cp\u003eAnother type of SV that you can get from your \u003ccode\u003eSAM\u003c/code\u003e files are copy number\nvariants (CNVs). By looking at the depth-of-coverage for your data files, you\ncan look for abnormalities that may indicate copy number variants. By using a\ntool like \u003ccode\u003emosdepth\u003c/code\u003e, you can quickly get a file showing the coverage across\nthe genome.\u003c/p\u003e\n\u003cp\u003eBe aware that if you are comparing the coverage counts from different tools,\nthat they have different defaults that may affect comparison. Some discard\n\u003ccode\u003eQC_FAIL\u003c/code\u003e, \u003ccode\u003eDUP\u003c/code\u003e, and \u003ccode\u003eSECONDARY\u003c/code\u003e flagged reads. This is probably appropriate,\nand corresponds to what most genome browsers will display (see\n\u003ca href=\"https://gist.github.com/cmdcolin/9f677eca28448d8a7c5d6e9917fc56af\"\u003ehttps://gist.github.com/cmdcolin/9f677eca28448d8a7c5d6e9917fc56af\u003c/a\u003e for a short\nsummary of depth calculated from different tools)\u003c/p\u003e\n\u003cp\u003eNote that both long and short reads can be used for CNV detection. Long reads\nmay give more accurate measurements also, with their better ability to map\nsmoothly through difficult regions of the genome.\u003c/p\u003e\n\u003cp\u003e\u003cimg src=\"/media/coverage_cnv.png\" alt=\"\"\u003e\u003c/p\u003e\n\u003cp\u003eScreenshot showing coverage in \u003ccode\u003eBigWig\u003c/code\u003e format from nanopore reads on normal\nand tumor tissue from a melanoma cancer cell line (COLO829) plotted using\nJBrowse 2. This coverage data is calculated from nanopore sequencing from\n\u003ca href=\"https://www.biorxiv.org/content/10.1101/2020.10.15.340497v1.full\"\u003ehere\u003c/a\u003e using\n\u003ca href=\"https://github.com/brentp/mosdepth\"\u003e\u003ccode\u003emosdepth\u003c/code\u003e\u003c/a\u003e, converted from \u003ccode\u003eBedGraph\u003c/code\u003e to\n\u003ccode\u003eBigWig\u003c/code\u003e, and loaded into JBrowse 2. See\n(\u003ca href=\"https://jbrowse.org/code/jb2/v1.6.4/?config=test_data%2Fconfig_demo.json\u0026#x26;session=share-MZj3d18lzH\u0026#x26;password=3X7bS\"\u003edemo\u003c/a\u003e\nand\n\u003ca href=\"https://jbrowse.org/jb2/docs/user_guide/#viewing-whole-genome-coverage-for-profiling-cnv\"\u003etutorial\u003c/a\u003e)\u003c/p\u003e\n\u003ch2 id=\"the-future-with-graph-genomes-and-de-novo-assemblies\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#the-future-with-graph-genomes-and-de-novo-assemblies\"\u003e\u003ca href=\"#the-future-with-graph-genomes-and-de-novo-assemblies\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eThe future, with graph genomes and \u003cem\u003ede novo\u003c/em\u003e assemblies\u003c/h2\u003e\n\u003cp\u003eCurrently, SV visualization is highly based on comparing data versus a\nreference genome (and the \u003ccode\u003eSAM\u003c/code\u003e format is a signature of this: it stores data\nin terms of reference genome coordinates). In the future, SV visualization may\nlook more similar to comparative genomics, where we compare an SV to a\npopulation specific reference from a graph genomes or something like this.\u003c/p\u003e\n\u003cp\u003eIt is known that \u003cem\u003ede novo\u003c/em\u003e assembly has more power to detect SVs than some read\noperations (\u003ca href=\"https://twitter.com/lh3lh3/status/1362921612690010118/photo/1\"\u003ehttps://twitter.com/lh3lh3/status/1362921612690010118/photo/1\u003c/a\u003e\nas \u003cem\u003ede novo\u003c/em\u003e assembled genomes improve and become more widespread, we may see a\nshift in how SVs are called\u003c/p\u003e\n\u003cp\u003eI would also like to see improved ability to do fast or 'on the fly' gene\nprediction on the \u003cem\u003ede novo\u003c/em\u003e assembled genomes, and we can see what SNPs or\nmodified splicing might look like in copies of genes (e.g. derived regions of\nthe CNV duplications).\u003c/p\u003e\n\u003cp\u003eFun fact: the\n\u003ca href=\"https://github.com/lh3/gfatools/blob/master/doc/rGFA.md#the-graph-alignment-format-gaf\"\u003e\u003ccode\u003eGAF\u003c/code\u003e\u003c/a\u003e\n(graphical alignment format) is a strict superset of\n\u003ca href=\"https://github.com/lh3/miniasm/blob/master/PAF.md\"\u003e\u003ccode\u003ePAF\u003c/code\u003e\u003c/a\u003e (pairwise alignment\nformat) by storing graph node labels in the \u003ccode\u003etarget name\u003c/code\u003e slot of \u003ccode\u003ePAF\u003c/code\u003e, and\ncan refer to an \u003ccode\u003erGFA\u003c/code\u003e (reference genome graph)! Looking forward to the graph\ngenome world.\u003c/p\u003e\n\u003ch2 id=\"conclusion\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#conclusion\"\u003e\u003ca href=\"#conclusion\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eConclusion\u003c/h2\u003e\n\u003cp\u003eAlgorithms that actually call structural variants face many challenges, but\nunderstanding how the reads are encoded in SAM format, and seeing what they\nlook like in the genome browser is a useful first step to gaining a better\nunderstanding.\u003c/p\u003e\n\u003cp\u003eIn summary, some of the signatures of SVs may include:\u003c/p\u003e\n\u003cul\u003e\n\u003cli\u003eAberrant insert size (\u003ccode\u003eTLEN\u003c/code\u003e) detection (longer for deletion, shorter for\ninsertion)\u003c/li\u003e\n\u003cli\u003eAberrant pair orientation (pairs are not pointing at each other)\u003c/li\u003e\n\u003cli\u003eSplit-read detection (\u003ccode\u003eSA\u003c/code\u003e tag)\u003c/li\u003e\n\u003cli\u003e\u003ccode\u003eCIGAR\u003c/code\u003e string processing (\u003ccode\u003eD\u003c/code\u003e operator for deletions, \u003ccode\u003eI\u003c/code\u003e operator for\ninsertions)\u003c/li\u003e\n\u003cli\u003eOver-abundance of clipping (\u003ccode\u003eS\u003c/code\u003e or \u003ccode\u003eH\u003c/code\u003e operators in \u003ccode\u003eCIGAR\u003c/code\u003e)\u003c/li\u003e\n\u003cli\u003eDepth of coverage changes for CNVs\u003c/li\u003e\n\u003cli\u003eAligning \u003cem\u003ede novo\u003c/em\u003e assembly vs a reference genome\n(\u003ca href=\"https://twitter.com/lh3lh3/status/1362921612690010118/photo/1\"\u003ehttps://twitter.com/lh3lh3/status/1362921612690010118/photo/1\u003c/a\u003e) which can\noutput \u003ccode\u003eSAM\u003c/code\u003e, but it can also output\n\u003ca href=\"https://github.com/lh3/miniasm/blob/master/PAF.md\"\u003e\u003ccode\u003ePAF\u003c/code\u003e\u003c/a\u003e format (which can\nbe loaded in JBrowse 2 in the synteny views). Techniques of detecting SVs on\nPAF will be fundamentally pretty similar to the techniques listed above but\nmay look a bit different (see \u003ccode\u003ecs\u003c/code\u003e tag in \u003ccode\u003ePAF\u003c/code\u003e for example, it is a modified\n\u003ccode\u003eCIGAR\u003c/code\u003e-like string)\u003c/li\u003e\n\u003c/ul\u003e\n\u003cp\u003eIf you have any ideas I should include here, let me know!\u003c/p\u003e\n\u003ch3 id=\"appendix-a-parsing-cigar-strings\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#appendix-a-parsing-cigar-strings\"\u003e\u003ca href=\"#appendix-a-parsing-cigar-strings\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eAppendix A: Parsing \u003ccode\u003eCIGAR\u003c/code\u003e strings\u003c/h3\u003e\n\u003cp\u003eThis is code that can help determine the pair orientation from a single BAM\nrecord. Might be too much detail but follow along\u003c/p\u003e\n\u003cdiv class=\"highlight highlight-ts\"\u003e\u003cpre\u003e\u003cspan class=\"pl-c\"\u003e// @param flags - flags from a single read\u003c/span\u003e\n\u003cspan class=\"pl-c\"\u003e// @param ref - the string of the reference sequence, just used to determine if it matches rnext\u003c/span\u003e\n\u003cspan class=\"pl-c\"\u003e// @param rnext - the string of the RNEXT, just used to determine if it matches ref\u003c/span\u003e\n\u003cspan class=\"pl-c\"\u003e// @param tlen - the TLEN field from SAM\u003c/span\u003e\n\u003cspan class=\"pl-c\"\u003e// @return e.g. F1R2 normal paired end orientation\u003c/span\u003e\n\u003cspan class=\"pl-k\"\u003efunction\u003c/span\u003e \u003cspan class=\"pl-en\"\u003egetPairOrientation\u003c/span\u003e(\n  \u003cspan class=\"pl-v\"\u003eflags\u003c/span\u003e\u003cspan class=\"pl-k\"\u003e:\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003enumber\u003c/span\u003e,\n  \u003cspan class=\"pl-v\"\u003eref\u003c/span\u003e\u003cspan class=\"pl-k\"\u003e:\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003estring\u003c/span\u003e,\n  \u003cspan class=\"pl-v\"\u003ernext\u003c/span\u003e\u003cspan class=\"pl-k\"\u003e:\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003estring\u003c/span\u003e,\n  \u003cspan class=\"pl-v\"\u003etlen\u003c/span\u003e\u003cspan class=\"pl-k\"\u003e:\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003enumber\u003c/span\u003e,\n) {\n  \u003cspan class=\"pl-c\"\u003e// this read is not unmapped \u0026#x26;\u0026#x26;\u003c/span\u003e\n  \u003cspan class=\"pl-c\"\u003e// this read's mate is also not unmapped \u0026#x26;\u0026#x26;\u003c/span\u003e\n  \u003cspan class=\"pl-c\"\u003e// this read's mate is on the same reference genome\u003c/span\u003e\n  \u003cspan class=\"pl-k\"\u003eif\u003c/span\u003e (\u003cspan class=\"pl-k\"\u003e!\u003c/span\u003e\u003cspan class=\"pl-smi\"\u003eflags\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e\u0026#x26;\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003e4\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e\u0026#x26;\u0026#x26;\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e!\u003c/span\u003e\u003cspan class=\"pl-smi\"\u003eflags\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e\u0026#x26;\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003e8\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e\u0026#x26;\u0026#x26;\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003eref\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e===\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003ernext\u003c/span\u003e) {\n    \u003cspan class=\"pl-k\"\u003econst\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003es1\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003eflags\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e\u0026#x26;\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003e16\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e?\u003c/span\u003e \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003eR\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003e\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e:\u003c/span\u003e \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003eF\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003e\u003c/span\u003e\n    \u003cspan class=\"pl-k\"\u003econst\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003es2\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003eflags\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e\u0026#x26;\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003e32\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e?\u003c/span\u003e \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003eR\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003e\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e:\u003c/span\u003e \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003eF\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003e\u003c/span\u003e\n    \u003cspan class=\"pl-k\"\u003elet\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003eo1\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003e \u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003e\u003c/span\u003e\n    \u003cspan class=\"pl-k\"\u003elet\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003eo2\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003e \u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003e\u003c/span\u003e\n\n    \u003cspan class=\"pl-c\"\u003e// if first in pair\u003c/span\u003e\n    \u003cspan class=\"pl-k\"\u003eif\u003c/span\u003e (\u003cspan class=\"pl-smi\"\u003eflags\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e\u0026#x26;\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003e64\u003c/span\u003e) {\n      \u003cspan class=\"pl-smi\"\u003eo1\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003e1\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003e\u003c/span\u003e\n      \u003cspan class=\"pl-smi\"\u003eo2\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003e2\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003e\u003c/span\u003e\n    }\n\n    \u003cspan class=\"pl-c\"\u003e// else if second in pair\u003c/span\u003e\n    \u003cspan class=\"pl-k\"\u003eelse\u003c/span\u003e \u003cspan class=\"pl-k\"\u003eif\u003c/span\u003e (\u003cspan class=\"pl-smi\"\u003eflags\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e\u0026#x26;\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003e128\u003c/span\u003e) {\n      \u003cspan class=\"pl-smi\"\u003eo1\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003e2\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003e\u003c/span\u003e\n      \u003cspan class=\"pl-smi\"\u003eo2\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003e1\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003e\u003c/span\u003e\n    }\n\n    \u003cspan class=\"pl-k\"\u003econst\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003etmp\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e []\n    \u003cspan class=\"pl-k\"\u003eif\u003c/span\u003e (\u003cspan class=\"pl-smi\"\u003etlen\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e\u003e\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003e0\u003c/span\u003e) {\n      \u003cspan class=\"pl-smi\"\u003etmp\u003c/span\u003e[\u003cspan class=\"pl-c1\"\u003e0\u003c/span\u003e] \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003es1\u003c/span\u003e\n      \u003cspan class=\"pl-smi\"\u003etmp\u003c/span\u003e[\u003cspan class=\"pl-c1\"\u003e1\u003c/span\u003e] \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003eo1\u003c/span\u003e\n      \u003cspan class=\"pl-smi\"\u003etmp\u003c/span\u003e[\u003cspan class=\"pl-c1\"\u003e2\u003c/span\u003e] \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003es2\u003c/span\u003e\n      \u003cspan class=\"pl-smi\"\u003etmp\u003c/span\u003e[\u003cspan class=\"pl-c1\"\u003e3\u003c/span\u003e] \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003eo2\u003c/span\u003e\n    } \u003cspan class=\"pl-k\"\u003eelse\u003c/span\u003e {\n      \u003cspan class=\"pl-smi\"\u003etmp\u003c/span\u003e[\u003cspan class=\"pl-c1\"\u003e2\u003c/span\u003e] \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003es1\u003c/span\u003e\n      \u003cspan class=\"pl-smi\"\u003etmp\u003c/span\u003e[\u003cspan class=\"pl-c1\"\u003e3\u003c/span\u003e] \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003eo1\u003c/span\u003e\n      \u003cspan class=\"pl-smi\"\u003etmp\u003c/span\u003e[\u003cspan class=\"pl-c1\"\u003e0\u003c/span\u003e] \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003es2\u003c/span\u003e\n      \u003cspan class=\"pl-smi\"\u003etmp\u003c/span\u003e[\u003cspan class=\"pl-c1\"\u003e1\u003c/span\u003e] \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003eo2\u003c/span\u003e\n    }\n    \u003cspan class=\"pl-k\"\u003ereturn\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003etmp\u003c/span\u003e.\u003cspan class=\"pl-c1\"\u003ejoin\u003c/span\u003e(\u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e''\u003c/span\u003e\u003c/span\u003e)\n  }\n  \u003cspan class=\"pl-k\"\u003ereturn\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003enull\u003c/span\u003e\n}\n\u003c/pre\u003e\u003c/div\u003e\n\u003cp\u003eThen this can be broken down further by orientation type\u003c/p\u003e\n\u003cp\u003ePaired end reads are \"fr\"\nMate pair reads are \"rf\"\u003c/p\u003e\n\u003cp\u003eSo you can interpret e.g. F1R2 in relation to being a paired end read (fr) or mate pair (rf) below and with this link \u003ca href=\"https://software.broadinstitute.org/software/igv/interpreting_pair_orientations\"\u003ehttps://software.broadinstitute.org/software/igv/interpreting_pair_orientations\u003c/a\u003e\u003c/p\u003e\n\u003cdiv class=\"highlight highlight-json\"\u003e\u003cpre\u003e{\n  \u003cspan class=\"pl-ent\"\u003e\"fr\"\u003c/span\u003e: {\n    \u003cspan class=\"pl-ent\"\u003e\"F1R2\"\u003c/span\u003e: \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003eLR\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e,\n    \u003cspan class=\"pl-ent\"\u003e\"F2R1\"\u003c/span\u003e: \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003eLR\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e,\n\n    \u003cspan class=\"pl-ent\"\u003e\"F1F2\"\u003c/span\u003e: \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003eLL\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e,\n    \u003cspan class=\"pl-ent\"\u003e\"F2F1\"\u003c/span\u003e: \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003eLL\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e,\n\n    \u003cspan class=\"pl-ent\"\u003e\"R1R2\"\u003c/span\u003e: \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003eRR\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e,\n    \u003cspan class=\"pl-ent\"\u003e\"R2R1\"\u003c/span\u003e: \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003eRR\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e,\n\n    \u003cspan class=\"pl-ent\"\u003e\"R1F2\"\u003c/span\u003e: \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003eRL\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e,\n    \u003cspan class=\"pl-ent\"\u003e\"R2F1\"\u003c/span\u003e: \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003eRL\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e\n  },\n\n  \u003cspan class=\"pl-ent\"\u003e\"rf\"\u003c/span\u003e: {\n    \u003cspan class=\"pl-ent\"\u003e\"R1F2\"\u003c/span\u003e: \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003eLR\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e,\n    \u003cspan class=\"pl-ent\"\u003e\"R2F1\"\u003c/span\u003e: \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003eLR\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e,\n\n    \u003cspan class=\"pl-ent\"\u003e\"R1R2\"\u003c/span\u003e: \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003eLL\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e,\n    \u003cspan class=\"pl-ent\"\u003e\"R2R1\"\u003c/span\u003e: \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003eLL\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e,\n\n    \u003cspan class=\"pl-ent\"\u003e\"F1F2\"\u003c/span\u003e: \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003eRR\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e,\n    \u003cspan class=\"pl-ent\"\u003e\"F2F1\"\u003c/span\u003e: \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003eRR\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e,\n\n    \u003cspan class=\"pl-ent\"\u003e\"F1R2\"\u003c/span\u003e: \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003eRL\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e,\n    \u003cspan class=\"pl-ent\"\u003e\"F2R1\"\u003c/span\u003e: \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003eRL\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e\n  }\n}\n\u003c/pre\u003e\u003c/div\u003e\n\u003ch3 id=\"appendix-b---cigar-parsing\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#appendix-b---cigar-parsing\"\u003e\u003ca href=\"#appendix-b---cigar-parsing\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eAppendix B - \u003ccode\u003eCIGAR\u003c/code\u003e parsing\u003c/h3\u003e\n\u003cdiv class=\"highlight highlight-ts\"\u003e\u003cpre\u003e\u003cspan class=\"pl-c\"\u003e// @param cigar: CIGAR string in text form\u003c/span\u003e\n\u003cspan class=\"pl-c\"\u003e// @returns an array of elements like ['30','M', '2','I', '50','M', '40','D']\u003c/span\u003e\n\u003cspan class=\"pl-c\"\u003e// which you can consume in a loop two elements at a time\u003c/span\u003e\n\u003cspan class=\"pl-k\"\u003efunction\u003c/span\u003e \u003cspan class=\"pl-en\"\u003eparseCigar\u003c/span\u003e(\u003cspan class=\"pl-v\"\u003ecigar\u003c/span\u003e\u003cspan class=\"pl-k\"\u003e:\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003estring\u003c/span\u003e) {\n  \u003cspan class=\"pl-k\"\u003ereturn\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003ecigar\u003c/span\u003e.\u003cspan class=\"pl-c1\"\u003esplit\u003c/span\u003e(\u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-sr\"\u003e\u003cspan class=\"pl-pds\"\u003e/\u003c/span\u003e(\u003cspan class=\"pl-c1\"\u003e[MIDNSHPX=]\u003c/span\u003e)\u003cspan class=\"pl-pds\"\u003e/\u003c/span\u003e\u003c/span\u003e\u003c/span\u003e)\n}\n\u003c/pre\u003e\u003c/div\u003e\n\u003cp\u003eThen parse the returned array two at a time\u003c/p\u003e\n\u003cdiv class=\"highlight highlight-ts\"\u003e\u003cpre\u003e\u003cspan class=\"pl-c\"\u003e// this function does nothing, but is informative for how to parse interpret a\u003c/span\u003e\n\u003cspan class=\"pl-c\"\u003e// CIGAR string\u003c/span\u003e\n\u003cspan class=\"pl-c\"\u003e// @param cigar:CIGAR string from record\u003c/span\u003e\n\u003cspan class=\"pl-c\"\u003e// @param readSeq: the SEQ from record\u003c/span\u003e\n\u003cspan class=\"pl-c\"\u003e// @param refSeq: the reference sequence underlying the read\u003c/span\u003e\n\u003cspan class=\"pl-k\"\u003efunction\u003c/span\u003e \u003cspan class=\"pl-en\"\u003einterpretCigar\u003c/span\u003e(\u003cspan class=\"pl-v\"\u003ecigar\u003c/span\u003e\u003cspan class=\"pl-k\"\u003e:\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003estring\u003c/span\u003e, \u003cspan class=\"pl-v\"\u003ereadSeq\u003c/span\u003e\u003cspan class=\"pl-k\"\u003e:\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003estring\u003c/span\u003e, \u003cspan class=\"pl-v\"\u003erefSeq\u003c/span\u003e\u003cspan class=\"pl-k\"\u003e:\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003estring\u003c/span\u003e) {\n  \u003cspan class=\"pl-k\"\u003econst\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003eopts\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e \u003cspan class=\"pl-en\"\u003eparseCigar\u003c/span\u003e(\u003cspan class=\"pl-smi\"\u003ecigar\u003c/span\u003e)\n  \u003cspan class=\"pl-k\"\u003elet\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003eqpos\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003e0\u003c/span\u003e \u003cspan class=\"pl-c\"\u003e// query position, position on the read\u003c/span\u003e\n  \u003cspan class=\"pl-k\"\u003elet\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003etpos\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003e0\u003c/span\u003e \u003cspan class=\"pl-c\"\u003e// target position, position on the reference sequence\u003c/span\u003e\n  \u003cspan class=\"pl-k\"\u003efor\u003c/span\u003e (\u003cspan class=\"pl-k\"\u003elet\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003ei\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003e0\u003c/span\u003e; \u003cspan class=\"pl-smi\"\u003ei\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e\u0026#x3C;\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003eops\u003c/span\u003e.\u003cspan class=\"pl-c1\"\u003elength\u003c/span\u003e; \u003cspan class=\"pl-smi\"\u003ei\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e+=\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003e2\u003c/span\u003e) {\n    \u003cspan class=\"pl-k\"\u003econst\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003elength\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e+\u003c/span\u003e\u003cspan class=\"pl-smi\"\u003eopts\u003c/span\u003e[\u003cspan class=\"pl-smi\"\u003ei\u003c/span\u003e]\n    \u003cspan class=\"pl-k\"\u003econst\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003eoperator\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003eopts\u003c/span\u003e[\u003cspan class=\"pl-smi\"\u003ei\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e+\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003e1\u003c/span\u003e]\n    \u003cspan class=\"pl-c\"\u003e// do things. refer to the CIGAR chart in SAMv1.pdf for which operators\u003c/span\u003e\n    \u003cspan class=\"pl-c\"\u003e// \"consume reference\" to see whether to increment\u003c/span\u003e\n    \u003cspan class=\"pl-k\"\u003eif\u003c/span\u003e (\u003cspan class=\"pl-smi\"\u003eop\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e===\u003c/span\u003e \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003eM\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003e\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e||\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003eop\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e===\u003c/span\u003e \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003e=\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003e\u003c/span\u003e) {\n      \u003cspan class=\"pl-c\"\u003e// matches consume query and reference\u003c/span\u003e\n      \u003cspan class=\"pl-k\"\u003econst\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003erefMatch\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003erefSeq\u003c/span\u003e.\u003cspan class=\"pl-c1\"\u003eslice\u003c/span\u003e(\u003cspan class=\"pl-smi\"\u003etpos\u003c/span\u003e, \u003cspan class=\"pl-smi\"\u003etpos\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e+\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003elen\u003c/span\u003e)\n      \u003cspan class=\"pl-k\"\u003econst\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003ereadMatch\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003ereadSeq\u003c/span\u003e.\u003cspan class=\"pl-c1\"\u003eslice\u003c/span\u003e(\u003cspan class=\"pl-smi\"\u003eqpos\u003c/span\u003e, \u003cspan class=\"pl-smi\"\u003eqpos\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e+\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003elen\u003c/span\u003e)\n      \u003cspan class=\"pl-k\"\u003efor\u003c/span\u003e (\u003cspan class=\"pl-k\"\u003elet\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003ei\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003e0\u003c/span\u003e; \u003cspan class=\"pl-smi\"\u003ei\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e\u0026#x3C;\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003elen\u003c/span\u003e; \u003cspan class=\"pl-smi\"\u003ei\u003c/span\u003e\u003cspan class=\"pl-k\"\u003e++\u003c/span\u003e) {\n        \u003cspan class=\"pl-k\"\u003eif\u003c/span\u003e (\u003cspan class=\"pl-smi\"\u003erefMatch\u003c/span\u003e[\u003cspan class=\"pl-smi\"\u003ei\u003c/span\u003e] \u003cspan class=\"pl-k\"\u003e!==\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003ereadMatch\u003c/span\u003e[\u003cspan class=\"pl-smi\"\u003ei\u003c/span\u003e]) {\n          \u003cspan class=\"pl-c\"\u003e// SNP at this position\u003c/span\u003e\n        }\n      }\n      \u003cspan class=\"pl-smi\"\u003eqpos\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e+=\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003elen\u003c/span\u003e\n      \u003cspan class=\"pl-smi\"\u003etpos\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e+=\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003elen\u003c/span\u003e\n    }\n    \u003cspan class=\"pl-k\"\u003eif\u003c/span\u003e (\u003cspan class=\"pl-smi\"\u003eop\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e===\u003c/span\u003e \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003eI\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003e\u003c/span\u003e) {\n      \u003cspan class=\"pl-c\"\u003e// insertions only consume query\u003c/span\u003e\n      \u003cspan class=\"pl-c\"\u003e// sequence of the insertion from the read is\u003c/span\u003e\n      \u003cspan class=\"pl-k\"\u003econst\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003einsSeq\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003ereadSeq\u003c/span\u003e.\u003cspan class=\"pl-c1\"\u003eslice\u003c/span\u003e(\u003cspan class=\"pl-smi\"\u003eqpos\u003c/span\u003e, \u003cspan class=\"pl-smi\"\u003eqpos\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e+\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003elen\u003c/span\u003e)\n      \u003cspan class=\"pl-smi\"\u003eqpos\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e+=\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003elen\u003c/span\u003e\n    }\n    \u003cspan class=\"pl-k\"\u003eif\u003c/span\u003e (\u003cspan class=\"pl-smi\"\u003eop\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e===\u003c/span\u003e \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003eD\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003e\u003c/span\u003e) {\n      \u003cspan class=\"pl-c\"\u003e// deletions only consume reference\u003c/span\u003e\n      \u003cspan class=\"pl-c\"\u003e// sequence of the deletion from the reference is\u003c/span\u003e\n      \u003cspan class=\"pl-k\"\u003econst\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003edelSeq\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003erefSeq\u003c/span\u003e.\u003cspan class=\"pl-c1\"\u003eslice\u003c/span\u003e(\u003cspan class=\"pl-smi\"\u003etpos\u003c/span\u003e, \u003cspan class=\"pl-smi\"\u003etpos\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e+\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003elen\u003c/span\u003e)\n      \u003cspan class=\"pl-smi\"\u003etpad\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e+=\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003elen\u003c/span\u003e\n    }\n    \u003cspan class=\"pl-k\"\u003eif\u003c/span\u003e (\u003cspan class=\"pl-smi\"\u003eop\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e===\u003c/span\u003e \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003eN\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003e\u003c/span\u003e) {\n      \u003cspan class=\"pl-c\"\u003e// skips only consume reference\u003c/span\u003e\n      \u003cspan class=\"pl-c\"\u003e// skips are similar to deletions but are related to spliced alignments\u003c/span\u003e\n      \u003cspan class=\"pl-smi\"\u003etpad\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e+=\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003elen\u003c/span\u003e\n    }\n    \u003cspan class=\"pl-k\"\u003eif\u003c/span\u003e (\u003cspan class=\"pl-smi\"\u003eop\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e===\u003c/span\u003e \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003eX\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003e\u003c/span\u003e) {\n      \u003cspan class=\"pl-c\"\u003e// mismatch using the extended CIGAR format\u003c/span\u003e\n      \u003cspan class=\"pl-c\"\u003e// could lookup the mismatch letter in a string containing the reference\u003c/span\u003e\n      \u003cspan class=\"pl-k\"\u003econst\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003emismatch\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003erefSeq\u003c/span\u003e.\u003cspan class=\"pl-c1\"\u003eslice\u003c/span\u003e(\u003cspan class=\"pl-smi\"\u003etpos\u003c/span\u003e, \u003cspan class=\"pl-smi\"\u003etpos\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e+\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003elen\u003c/span\u003e)\n      \u003cspan class=\"pl-smi\"\u003eqpos\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e+=\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003elen\u003c/span\u003e\n      \u003cspan class=\"pl-smi\"\u003etpos\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e+=\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003elen\u003c/span\u003e\n    }\n    \u003cspan class=\"pl-k\"\u003eif\u003c/span\u003e (\u003cspan class=\"pl-smi\"\u003eop\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e===\u003c/span\u003e \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003eH\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003e\u003c/span\u003e) {\n      \u003cspan class=\"pl-c\"\u003e// does not consume query or reference\u003c/span\u003e\n      \u003cspan class=\"pl-c\"\u003e// hardclip is just an indicator\u003c/span\u003e\n    }\n    \u003cspan class=\"pl-k\"\u003eif\u003c/span\u003e (\u003cspan class=\"pl-smi\"\u003eop\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e===\u003c/span\u003e \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003eS\u003cspan class=\"pl-pds\"\u003e'\u003c/span\u003e\u003c/span\u003e) {\n      \u003cspan class=\"pl-c\"\u003e// softclip consumes query\u003c/span\u003e\n      \u003cspan class=\"pl-c\"\u003e// below gets the entire soft clipped portion\u003c/span\u003e\n      \u003cspan class=\"pl-k\"\u003econst\u003c/span\u003e \u003cspan class=\"pl-c1\"\u003esoftClipStr\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e=\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003ereadSeq\u003c/span\u003e.\u003cspan class=\"pl-c1\"\u003eslice\u003c/span\u003e(\u003cspan class=\"pl-smi\"\u003eqpos\u003c/span\u003e, \u003cspan class=\"pl-smi\"\u003eqpos\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e+\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003elen\u003c/span\u003e)\n      \u003cspan class=\"pl-smi\"\u003eqpos\u003c/span\u003e \u003cspan class=\"pl-k\"\u003e+=\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003elen\u003c/span\u003e\n    }\n  }\n}\n\u003c/pre\u003e\u003c/div\u003e\n\u003cp\u003eNote for example, that to determine how long a record is on the reference\nsequence, you have to combine the records start position with the CIGAR string,\nbasically parsing the CIGAR string to add up tpos and return tpos\u003c/p\u003e\n\u003ch3 id=\"appendix-c---align-fastq-directly-to-cram\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#appendix-c---align-fastq-directly-to-cram\"\u003e\u003ca href=\"#appendix-c---align-fastq-directly-to-cram\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eAppendix C - align \u003ccode\u003eFASTQ\u003c/code\u003e directly to \u003ccode\u003eCRAM\u003c/code\u003e\u003c/h3\u003e\n\u003cp\u003eThis example from the htslib documentation\n(\u003ca href=\"http://www.htslib.org/workflow/fastq.html\"\u003ehttp://www.htslib.org/workflow/fastq.html\u003c/a\u003e\nshows how you can stream directly from \u003ccode\u003eFASTQ\u003c/code\u003e to \u003ccode\u003eCRAM\u003c/code\u003e (and generate the\nindex file .crai too)\u003c/p\u003e\n\u003cp\u003eIf you want, you can make this a little shell script, easy_align_shortreads.sh\u003c/p\u003e\n\u003cp\u003eeasy_align_shortreads.sh\u003c/p\u003e\n\u003cdiv class=\"highlight highlight-shell\"\u003e\u003cpre\u003e\u003cspan class=\"pl-c\"\u003e#!/bin/bash\u003c/span\u003e\nminimap2 -t 8 -a -x sr \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003cspan class=\"pl-smi\"\u003e$1\u003c/span\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003cspan class=\"pl-smi\"\u003e$2\u003c/span\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003cspan class=\"pl-smi\"\u003e$3\u003c/span\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e  \u003cspan class=\"pl-k\"\u003e|\u003c/span\u003e \\\nsamtools fixmate -u -m - - \u003cspan class=\"pl-k\"\u003e|\u003c/span\u003e \\\nsamtools sort -u -@2 - \u003cspan class=\"pl-k\"\u003e|\u003c/span\u003e \\\nsamtools markdup -@8 --reference \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003cspan class=\"pl-smi\"\u003e$1\u003c/span\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e - --write-index \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003cspan class=\"pl-smi\"\u003e$4\u003c/span\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e\n\u003c/pre\u003e\u003c/div\u003e\n\u003cp\u003eSimilar idea for longreads, except just a single fastq file is generally used for longreads\u003c/p\u003e\n\u003cp\u003eeasy_align_longreads.sh\u003c/p\u003e\n\u003cdiv class=\"highlight highlight-shell\"\u003e\u003cpre\u003e\u003cspan class=\"pl-c\"\u003e#!/bin/bash\u003c/span\u003e\nminimap2 -t 8 -a \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003cspan class=\"pl-smi\"\u003e$1\u003c/span\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003cspan class=\"pl-smi\"\u003e$2\u003c/span\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e  \u003cspan class=\"pl-k\"\u003e|\u003c/span\u003e \\\nsamtools fixmate -u -m - - \u003cspan class=\"pl-k\"\u003e|\u003c/span\u003e \\\nsamtools sort -u -@2 - \u003cspan class=\"pl-k\"\u003e|\u003c/span\u003e \\\nsamtools markdup -@8 --reference \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003cspan class=\"pl-smi\"\u003e$1\u003c/span\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e - --write-index \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003cspan class=\"pl-smi\"\u003e$3\u003c/span\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e\n\u003c/pre\u003e\u003c/div\u003e\n\u003cp\u003eThen call\u003c/p\u003e\n\u003cdiv class=\"highlight highlight-shell\"\u003e\u003cpre\u003ebash easy_align_shortreads.sh ref.fa reads1.fq reads2.fq out.cram\nbash easy_align_longreads.sh ref.fa reads.fq out.cram\n\n\u003cspan class=\"pl-c\"\u003e## output BAM instead\u003c/span\u003e\nbash easy_align_shortreads.sh ref.fa reads1.fq reads2.fq out.bam\nbash easy_align_longreads.sh ref.fa reads.fq out.bam\n\u003c/pre\u003e\u003c/div\u003e\n\u003cp\u003eThis same concept works with other common aligners as well like bwa\u003c/p\u003e\n\u003cp\u003eBonus: CRAM to bigwig, for looking at CNV/coverage\u003c/p\u003e\n\u003cdiv class=\"highlight highlight-shell\"\u003e\u003cpre\u003e\u003cspan class=\"pl-c\"\u003e#!/bin/bash\u003c/span\u003e\n\u003cspan class=\"pl-c\"\u003e# quickalign.sh ref.fa 1.fq 2.fq out.cram\u003c/span\u003e\n\u003cspan class=\"pl-c\"\u003e# produces out.cram and out.bw\u003c/span\u003e\nsamtools faidx \u003cspan class=\"pl-smi\"\u003e$1\u003c/span\u003e\nminimap2 -t 8 -a -x sr \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003cspan class=\"pl-smi\"\u003e$1\u003c/span\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003cspan class=\"pl-smi\"\u003e$2\u003c/span\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003cspan class=\"pl-smi\"\u003e$3\u003c/span\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e  \u003cspan class=\"pl-k\"\u003e|\u003c/span\u003e \\\nsamtools fixmate -u -m - - \u003cspan class=\"pl-k\"\u003e|\u003c/span\u003e \\\nsamtools sort -u -@2 - \u003cspan class=\"pl-k\"\u003e|\u003c/span\u003e \\\nsamtools markdup -@8 --reference \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003cspan class=\"pl-smi\"\u003e$1\u003c/span\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e - --write-index \u003cspan class=\"pl-s\"\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003cspan class=\"pl-smi\"\u003e$4\u003c/span\u003e\u003cspan class=\"pl-pds\"\u003e\"\u003c/span\u003e\u003c/span\u003e\n\n\nmosdepth \u003cspan class=\"pl-smi\"\u003e$4\u003c/span\u003e -f \u003cspan class=\"pl-smi\"\u003e$1\u003c/span\u003e \u003cspan class=\"pl-smi\"\u003e$4\u003c/span\u003e\ngunzip \u003cspan class=\"pl-smi\"\u003e$4\u003c/span\u003e.per-base.bed.gz\nbedGraphToBigWig \u003cspan class=\"pl-smi\"\u003e$4\u003c/span\u003e.per-base.bed \u003cspan class=\"pl-smi\"\u003e$1\u003c/span\u003e.fa.fai \u003cspan class=\"pl-smi\"\u003e$4\u003c/span\u003e.bw\n\u003c/pre\u003e\u003c/div\u003e\n\u003cp\u003eCall as \"quickalign.sh ref.fa 1.fq 2.fq out.cram\" gives you out.cram, out.cram.crai, and out.cram.bw (coverage)\u003c/p\u003e\n\u003ch3 id=\"appendix-d---the-md-tag-and-finding-snps-in-reads\"\u003e\u003ca aria-hidden=\"true\" tabindex=\"-1\" href=\"#appendix-d---the-md-tag-and-finding-snps-in-reads\"\u003e\u003ca href=\"#appendix-d---the-md-tag-and-finding-snps-in-reads\" style=\"margin-right: 10px\"\u003e#\u003c/a\u003e\u003c/a\u003eAppendix D - the \u003ccode\u003eMD\u003c/code\u003e tag and finding SNPs in reads\u003c/h3\u003e\n\u003cp\u003eThe \u003ccode\u003eMD\u003c/code\u003e tag helps tell you where the mismatches are without looking at the\nreference genome. This is useful because as I mentioned, \u003ccode\u003eCIGAR\u003c/code\u003e can say \u003ccode\u003e50M\u003c/code\u003e\n(50 matches) but some letters inside those 50 matches can be mismatches, it\nonly says there are no insertions/deletions in those 50 bases, but you have to\ndetermine where in those 50 bases where the mismatches are. The \u003ccode\u003eMD\u003c/code\u003e tag can\nhelp tell you where those are, but it is somewhat complicated to decode\n(\u003ca href=\"https://vincebuffalo.com/notes/2014/01/17/md-tags-in-bam-files.html\"\u003ehttps://vincebuffalo.com/notes/2014/01/17/md-tags-in-bam-files.html\u003c/a\u003e).\nYou have to combine it with the \u003ccode\u003eCIGAR\u003c/code\u003e to get the position of the mismatches\non the reference genome. If you have a reference genome to look at, you might\njust compare all the bases within the 50M to the reference genome and look for\nmismatches yourself and forget about the \u003ccode\u003eMD\u003c/code\u003e tag\u003c/p\u003e\n\u003cp\u003eThe \u003ccode\u003eMD\u003c/code\u003e tag is also not required to exist, but the command \u003ccode\u003esamtools calmd yourfile.bam --reference reference.fa\u003c/code\u003e can add \u003ccode\u003eMD\u003c/code\u003e tags to your \u003ccode\u003eBAM\u003c/code\u003e file. It\nis generally not useful for \u003ccode\u003eCRAM\u003c/code\u003e because \u003ccode\u003eCRAM\u003c/code\u003e actually does store\nmismatches with the reference genome in it's compression format. Note that\nthere are also some oddities about \u003ccode\u003eMD\u003c/code\u003e tag representation leading to\ncomplaints (e.g. \u003ca href=\"https://github.com/samtools/hts-specs/issues/505\"\u003ehttps://github.com/samtools/hts-specs/issues/505\u003c/a\u003e) leading more\ncredence to \"doing it yourself\" e.g. finding your own mismatches\nby comparing the read sequence with the reference, keeping track of where you\nare on the read and ref position with the \u003ccode\u003eCIGAR\u003c/code\u003e string (a la Appendix B)\u003c/p\u003e"}},"__N_SSG":true},"page":"/posts/[slug]","query":{"slug":"2022-02-06-sv-sam"},"buildId":"hr7NfWqLWdv1Kvd3x6-Z_","isFallback":false,"gsp":true,"scriptLoader":[]}</script></body></html>