(self.webpackChunk_N_E=self.webpackChunk_N_E||[]).push([[8463],{3905:function(e,t,n){"use strict";n.d(t,{kt:function(){return f}});var r=n(7294);function a(e,t,n){return t in e?Object.defineProperty(e,t,{value:n,enumerable:!0,configurable:!0,writable:!0}):e[t]=n,e}function o(e,t){var n=Object.keys(e);if(Object.getOwnPropertySymbols){var r=Object.getOwnPropertySymbols(e);t&&(r=r.filter((function(t){return Object.getOwnPropertyDescriptor(e,t).enumerable}))),n.push.apply(n,r)}return n}function i(e){for(var t=1;t<arguments.length;t++){var n=null!=arguments[t]?arguments[t]:{};t%2?o(Object(n),!0).forEach((function(t){a(e,t,n[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(n)):o(Object(n)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(n,t))}))}return e}function l(e,t){if(null==e)return{};var n,r,a=function(e,t){if(null==e)return{};var n,r,a={},o=Object.keys(e);for(r=0;r<o.length;r++)n=o[r],t.indexOf(n)>=0||(a[n]=e[n]);return a}(e,t);if(Object.getOwnPropertySymbols){var o=Object.getOwnPropertySymbols(e);for(r=0;r<o.length;r++)n=o[r],t.indexOf(n)>=0||Object.prototype.propertyIsEnumerable.call(e,n)&&(a[n]=e[n])}return a}var s=r.createContext({}),c=function(e){var t=r.useContext(s),n=t;return e&&(n="function"===typeof e?e(t):i(i({},t),e)),n},u={inlineCode:"code",wrapper:function(e){var t=e.children;return r.createElement(r.Fragment,{},t)}},p=r.forwardRef((function(e,t){var n=e.components,a=e.mdxType,o=e.originalType,s=e.parentName,p=l(e,["components","mdxType","originalType","parentName"]),f=c(n),d=a,h=f["".concat(s,".").concat(d)]||f[d]||u[d]||o;return n?r.createElement(h,i(i({ref:t},p),{},{components:n})):r.createElement(h,i({ref:t},p))}));function f(e,t){var n=arguments,a=t&&t.mdxType;if("string"===typeof e||a){var o=n.length,i=new Array(o);i[0]=p;var l={};for(var s in t)hasOwnProperty.call(t,s)&&(l[s]=t[s]);l.originalType=e,l.mdxType="string"===typeof e?e:a,i[1]=l;for(var c=2;c<o;c++)i[c]=n[c];return r.createElement.apply(null,i)}return r.createElement.apply(null,n)}p.displayName="MDXCreateElement"},4995:function(e,t,n){"use strict";n.r(t),n.d(t,{default:function(){return a}});var r=n(5893);function a(e){var t=e.children;return(0,r.jsxs)("div",{children:[(0,r.jsx)("a",{href:"/",children:"Return home"}),(0,r.jsx)("div",{className:"blog",children:t})]})}},1265:function(e,t,n){"use strict";n.r(t),n.d(t,{default:function(){return u}});var r=n(159),a=n(219),o=(n(7294),n(3905)),i=n(4995),l=["components"],s={},c=function(e){var t=e.children;return(0,o.kt)(i.default,null,t)};function u(e){var t=e.components,n=(0,a.Z)(e,l);return(0,o.kt)(c,(0,r.Z)({},s,n,{components:t,mdxType:"MDXLayout"}),(0,o.kt)("h1",null,"Analysis of chip-seq data"),(0,o.kt)("p",null,(0,o.kt)("strong",{parentName:"p"},"Intro")),(0,o.kt)("p",null,"I have had a goal for a long time to analyze ChIP-seq data from multiple\nexperiments, and i have had trouble due to my poor understanding of\nstatistics. I have recently taken several classes that vastly improve my\npractical understanding of subjects including the Data Analysis class on\ncoursera, the Computing for Data analysis class, and the Experimental\ngenome science class, all on coursera!! Currently, the data anlysis\napproach has been invaluable, and i looked at the performance of\nNormdiff scores for comparing experiments (Zheng et al. 2010)."),(0,o.kt)("p",null,(0,o.kt)("strong",{parentName:"p"},"Methods")),(0,o.kt)("p",null,"I did this preliminary data analysis to compare two biological\nreplicates. I used the wiggle files produced using MACS that represent\nthe number of reads that overlap each genome position. \xa0Then i\ncalculated the NormDiff using a short R script to get the background\nsubtracted and normalized scores."),(0,o.kt)("p",null,'After this, I analyzed the effectiveness of the normalization by\ncreating an "MA plot". MA graphs the results of two experiments by using\nthe log product vs log ratio, or alternatively, using the added effect\nvs the difference of two experiments (which is what I used). The idea is\nthat if the normalization is good, the differences or log ratios should\nnot be biased.'),(0,o.kt)("p",null,(0,o.kt)("strong",{parentName:"p"},"Results"),"\\\n\\\nAs it turns out, the normalization is very good and the best fit line\nalmost flat (slope 0.05, 95% confidence interval +- 0.001). One of the\ncool things of the NormDiff algorithm is that it estimates the variance\nfrom a sliding window in the data, and this appears to stabilize the\nnormalization very effectively (see Figure 1).\\\n\\\nFigure 1. MA plots showing line of best fit for S96 replicates. (a)\nNormDiff with local variance parameter (b) NormDiff with global variance\nparameter"),(0,o.kt)("p",null,(0,o.kt)("img",{parentName:"p",src:"../../media/44474299692_0.png",alt:"image"})),(0,o.kt)("p",null,(0,o.kt)("img",{parentName:"p",src:"../../media/44474299692_1.png",alt:"image"})),(0,o.kt)("p",null,(0,o.kt)("strong",{parentName:"p"},"Conclusion"),"\\\n\\\n\xa0I am fairly pleased with the results so I am making the data available\nand the R code so you can reproduce the results if you want!! Next, I\nwant to look at the clustering and peak finding, and possibly evaluate\nother models of normalization\\\n\\\n",(0,o.kt)("a",{parentName:"p",href:"https://dl.dropbox.com/u/11267/analyzewig.R"},"https://dl.dropbox.com/u/11267/analyzewig.R")),(0,o.kt)("p",null,(0,o.kt)("a",{parentName:"p",href:"https://dl.dropbox.com/u/11267/wiggledata.rda"},"https://dl.dropbox.com/u/11267/wiggledata.rda")," (5MB file)"),(0,o.kt)("p",null,"::: {#footer}\n","[ March 3rd, 2013 2:10pm ]","{#timestamp} ","[bioinformatics]","{.tag} ","[data\nanalysis]","{.tag} ","[statistics]","{.tag}\n:::"))}u.isMDXComponent=!0},6843:function(e,t,n){(window.__NEXT_P=window.__NEXT_P||[]).push(["/2013-03-03",function(){return n(1265)}])},159:function(e,t,n){"use strict";function r(){return(r=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var n=arguments[t];for(var r in n)Object.prototype.hasOwnProperty.call(n,r)&&(e[r]=n[r])}return e}).apply(this,arguments)}n.d(t,{Z:function(){return r}})},219:function(e,t,n){"use strict";function r(e,t){if(null==e)return{};var n,r,a=function(e,t){if(null==e)return{};var n,r,a={},o=Object.keys(e);for(r=0;r<o.length;r++)n=o[r],t.indexOf(n)>=0||(a[n]=e[n]);return a}(e,t);if(Object.getOwnPropertySymbols){var o=Object.getOwnPropertySymbols(e);for(r=0;r<o.length;r++)n=o[r],t.indexOf(n)>=0||Object.prototype.propertyIsEnumerable.call(e,n)&&(a[n]=e[n])}return a}n.d(t,{Z:function(){return r}})}},function(e){e.O(0,[9774,2888,179],(function(){return t=6843,e(e.s=t);var t}));var t=e.O();_N_E=t}]);