(self.webpackChunk_N_E=self.webpackChunk_N_E||[]).push([[9474],{3905:function(e,t,n){"use strict";n.d(t,{kt:function(){return d}});var r=n(7294);function a(e,t,n){return t in e?Object.defineProperty(e,t,{value:n,enumerable:!0,configurable:!0,writable:!0}):e[t]=n,e}function i(e,t){var n=Object.keys(e);if(Object.getOwnPropertySymbols){var r=Object.getOwnPropertySymbols(e);t&&(r=r.filter((function(t){return Object.getOwnPropertyDescriptor(e,t).enumerable}))),n.push.apply(n,r)}return n}function o(e){for(var t=1;t<arguments.length;t++){var n=null!=arguments[t]?arguments[t]:{};t%2?i(Object(n),!0).forEach((function(t){a(e,t,n[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(n)):i(Object(n)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(n,t))}))}return e}function s(e,t){if(null==e)return{};var n,r,a=function(e,t){if(null==e)return{};var n,r,a={},i=Object.keys(e);for(r=0;r<i.length;r++)n=i[r],t.indexOf(n)>=0||(a[n]=e[n]);return a}(e,t);if(Object.getOwnPropertySymbols){var i=Object.getOwnPropertySymbols(e);for(r=0;r<i.length;r++)n=i[r],t.indexOf(n)>=0||Object.prototype.propertyIsEnumerable.call(e,n)&&(a[n]=e[n])}return a}var l=r.createContext({}),f=function(e){var t=r.useContext(l),n=t;return e&&(n="function"===typeof e?e(t):o(o({},t),e)),n},p={inlineCode:"code",wrapper:function(e){var t=e.children;return r.createElement(r.Fragment,{},t)}},c=r.forwardRef((function(e,t){var n=e.components,a=e.mdxType,i=e.originalType,l=e.parentName,c=s(e,["components","mdxType","originalType","parentName"]),d=f(n),h=a,u=d["".concat(l,".").concat(h)]||d[h]||p[h]||i;return n?r.createElement(u,o(o({ref:t},c),{},{components:n})):r.createElement(u,o({ref:t},c))}));function d(e,t){var n=arguments,a=t&&t.mdxType;if("string"===typeof e||a){var i=n.length,o=new Array(i);o[0]=c;var s={};for(var l in t)hasOwnProperty.call(t,l)&&(s[l]=t[l]);s.originalType=e,s.mdxType="string"===typeof e?e:a,o[1]=s;for(var f=2;f<i;f++)o[f]=n[f];return r.createElement.apply(null,o)}return r.createElement.apply(null,n)}c.displayName="MDXCreateElement"},4995:function(e,t,n){"use strict";n.r(t),n.d(t,{default:function(){return a}});var r=n(5893);function a(e){var t=e.children;return(0,r.jsxs)("div",{children:[(0,r.jsx)("a",{href:"/",children:"Return home"}),(0,r.jsx)("div",{className:"blog",children:t})]})}},502:function(e,t,n){"use strict";n.r(t),n.d(t,{default:function(){return p}});var r=n(159),a=n(219),i=(n(7294),n(3905)),o=n(4995),s=["components"],l={},f=function(e){var t=e.children;return(0,i.kt)(o.default,null,t)};function p(e){var t=e.components,n=(0,a.Z)(e,s);return(0,i.kt)(f,(0,r.Z)({},l,n,{components:t,mdxType:"MDXLayout"}),(0,i.kt)("p",null,"Hello all!"),(0,i.kt)("p",null,"I just completed my undergrad senior thesis recently and I wanted to\nshare a little bit with you. All in all, it was a amazing experience! I\nhad a lot of help with my thesis from my advisors and other students but\nit was a very independent project for me also, and I feel very\nprivileged to have gotten involved in such a cool field--bioinformatics!"),(0,i.kt)("p",null,"The thesis is titled \"Comparison of transcription factor binding sites\nfrom ChIP-seq experiments\" and I can make it available on request or\nonce it's published on my university's website. It contains a lot of\nstuff that I have\n",(0,i.kt)("a",{parentName:"p",href:"http://searchvoidstar.tumblr.com/post/46138839023/comparison-of-chip-seq-data"},"already"),"\n",(0,i.kt)("a",{parentName:"p",href:"http://searchvoidstar.tumblr.com/post/44474299692/analysis-of-chip-seq-data"},"talked"),"\nabout on this blog already, but a lot of new stuff too!"),(0,i.kt)("p",null,"For my final project, one of the main things that I did was to use the\nlimma package from Bioconductor as a way to compare ChIP-seq data. My\noriginal goal was to use some standard t-tests to do comparisons, but I\nfound digging into the literature that there are improvements including\nthe moderated t-test from the limma package."),(0,i.kt)("p",null,"As a starting point, I use the binned read counts that are outputted as\nwiggle files from MACS, find the bins that are differentially bound by\ncomparing the entire datasets using limma (Figure 1). This reveals\nintuitively that the read counts that are high in one experiment but low\nin the other are identified as differential."),(0,i.kt)("p",null,(0,i.kt)("img",{parentName:"p",src:"https://64.media.tumblr.com/04c09c8fd00e773cb6fcaba1f3dbf21c/fd8ac0b514f0d0be-3a/s540x810/d7f54407189d1460ba3f63a63cacd623a789a1b0.png",alt:null})),(0,i.kt)("p",null,(0,i.kt)("strong",{parentName:"p"},"Figure 1. Differential ChIP-seq read counts from two ChIP-seq\nexperiments that are identified using limma")),(0,i.kt)("p",null,"The next thing that I did was to merge all the nearby binding sites\ntogether. I used a very simple merge approach using BEDTools by merging\nnearby positions only \\<20bp apart. It turns out that the merging\nperforms very well and many nearby binding sites are merged together\nfrom the high-confidence sites that limma identifies (Figure 2)"),(0,i.kt)("p",null,(0,i.kt)("img",{parentName:"p",src:"https://64.media.tumblr.com/64c92c49c15639d366e2026afcf63f27/fd8ac0b514f0d0be-3f/s540x810/182d1f4f0619d8e8caf3ef80a6d8be6a7aa801b5.png",alt:null})),(0,i.kt)("p",null,(0,i.kt)("strong",{parentName:"p"},"Figure 2. The ratio of the merged peaks (output) to the inputted bins\n(input)")," The decreasing ratio is a sign of many nearby bins that are\nidentified by limma being merged together at different p-value\nthresholds."),(0,i.kt)("p",null,"The next thing I did was inspect the 'signals' to show that these peaks\nresemble intuitively differential peaks. Indeed, many of these these\npeaks are highly significant and highly differentially bound (Figure 3).\nI attribute the success of this approach to the limma package. In my\npaper, I also looked at how my method compares with other software such\nas MACS and DIME for finding differential binding sites."),(0,i.kt)("p",null,(0,i.kt)("img",{parentName:"p",src:"https://64.media.tumblr.com/fbb195d0b4e3be71f2f43f4516757f8e/fd8ac0b514f0d0be-bf/s540x810/53bbe7e5cb1a2f2150b53974e77deeb40d8a8949.png",alt:null})),(0,i.kt)("p",null,(0,i.kt)("strong",{parentName:"p"},"Figure 3 An example of a differential binding site"),", showing a peak\nthat is reproduced in the replicates from one strain but that is not\npresent in the\xa0 replicates of the other strain."),(0,i.kt)("p",null,"These differential binding sites appear in many different forms, and one\nquestion is what effect do these binding sites have? These binding sites\nessentially regulate the gene expression of nearby genes, so I looked at\nthe gene expression data that was also made available and assessed the\ndifferential gene expression vs the differential binding (Figure 4).\nThis method is inspired by plotMM from the ",(0,i.kt)("a",{parentName:"p",href:"http://www.bioconductor.org/packages/2.12/bioc/html/Rcade.html"},"Rcade\npackage"),"\nfrom Bioconductor. The gene expression data was obtained with help from\nthe ",(0,i.kt)("a",{parentName:"p",href:"http://www.ncbi.nlm.nih.gov/geo/geo2r/"},"GEO2R interface")," that is\nfairly new interface for GEOquery on the NCBI website!"),(0,i.kt)("p",null,(0,i.kt)("img",{parentName:"p",src:"https://64.media.tumblr.com/f4eb4cb3d0f6f029b4a281ebc9e6dbae/fd8ac0b514f0d0be-0f/s540x810/585ae9f846b904b086266c945763ddbc8faf39a1.png",alt:null})),(0,i.kt)("p",null,(0,i.kt)("strong",{parentName:"p"},"Figure 4. The differences in binding sites are associated with changes\nto gene expression also.\xa0"),"Comparison of the log ratio of differential\nbinding sites vs the log ratio of the gene expression data. The lightly\ncolored points are the gene expression from a random gene, and the dark\ncolored points are gene expression for the nearest gene to the binding\nsite"),(0,i.kt)("p",null,"Overall, I was able to really deeply explore one approach for comparing\nChIP-seq data. I think the availability of high-throughput sequencing\ndata is very cool on the GEO database, and many new types of\nbioinformatics tools are needed to help analyze it! I don't think my\nwork will stop just because I finished my project, and I will hope to\nfind new and interesting things for the future!"),(0,i.kt)("p",null,"PS: there are some ",(0,i.kt)("a",{parentName:"p",href:"http://statsandgenomes.wordpress.com/2011/11/23/bioinformatics-seminar-professor-gordon-smyth-variance-models-for-rna-seq/"},"interesting\ncomparisons"),"\nof limma and edgeR that I did not consider in my project but which would\nbe interesting to analyze."),(0,i.kt)("p",null,"::: {#footer}\n","[ May 12th, 2013 4:53pm ]","{#timestamp} ","[bioinformatics]","{.tag}\n","[statistics]","{.tag} ","[chip-seq]","{.tag} ","[Rstats]","{.tag} ","[genomics]","{.tag}\n:::"))}p.isMDXComponent=!0},7775:function(e,t,n){(window.__NEXT_P=window.__NEXT_P||[]).push(["/2013-05-12",function(){return n(502)}])},159:function(e,t,n){"use strict";function r(){return(r=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var n=arguments[t];for(var r in n)Object.prototype.hasOwnProperty.call(n,r)&&(e[r]=n[r])}return e}).apply(this,arguments)}n.d(t,{Z:function(){return r}})},219:function(e,t,n){"use strict";function r(e,t){if(null==e)return{};var n,r,a=function(e,t){if(null==e)return{};var n,r,a={},i=Object.keys(e);for(r=0;r<i.length;r++)n=i[r],t.indexOf(n)>=0||(a[n]=e[n]);return a}(e,t);if(Object.getOwnPropertySymbols){var i=Object.getOwnPropertySymbols(e);for(r=0;r<i.length;r++)n=i[r],t.indexOf(n)>=0||Object.prototype.propertyIsEnumerable.call(e,n)&&(a[n]=e[n])}return a}n.d(t,{Z:function(){return r}})}},function(e){e.O(0,[9774,2888,179],(function(){return t=7775,e(e.s=t);var t}));var t=e.O();_N_E=t}]);