(self.webpackChunk_N_E=self.webpackChunk_N_E||[]).push([[9604],{3905:function(e,t,n){"use strict";n.d(t,{kt:function(){return f}});var r=n(7294);function o(e,t,n){return t in e?Object.defineProperty(e,t,{value:n,enumerable:!0,configurable:!0,writable:!0}):e[t]=n,e}function a(e,t){var n=Object.keys(e);if(Object.getOwnPropertySymbols){var r=Object.getOwnPropertySymbols(e);t&&(r=r.filter((function(t){return Object.getOwnPropertyDescriptor(e,t).enumerable}))),n.push.apply(n,r)}return n}function i(e){for(var t=1;t<arguments.length;t++){var n=null!=arguments[t]?arguments[t]:{};t%2?a(Object(n),!0).forEach((function(t){o(e,t,n[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(n)):a(Object(n)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(n,t))}))}return e}function s(e,t){if(null==e)return{};var n,r,o=function(e,t){if(null==e)return{};var n,r,o={},a=Object.keys(e);for(r=0;r<a.length;r++)n=a[r],t.indexOf(n)>=0||(o[n]=e[n]);return o}(e,t);if(Object.getOwnPropertySymbols){var a=Object.getOwnPropertySymbols(e);for(r=0;r<a.length;r++)n=a[r],t.indexOf(n)>=0||Object.prototype.propertyIsEnumerable.call(e,n)&&(o[n]=e[n])}return o}var u=r.createContext({}),l=function(e){var t=r.useContext(u),n=t;return e&&(n="function"===typeof e?e(t):i(i({},t),e)),n},c={inlineCode:"code",wrapper:function(e){var t=e.children;return r.createElement(r.Fragment,{},t)}},p=r.forwardRef((function(e,t){var n=e.components,o=e.mdxType,a=e.originalType,u=e.parentName,p=s(e,["components","mdxType","originalType","parentName"]),f=l(n),h=o,m=f["".concat(u,".").concat(h)]||f[h]||c[h]||a;return n?r.createElement(m,i(i({ref:t},p),{},{components:n})):r.createElement(m,i({ref:t},p))}));function f(e,t){var n=arguments,o=t&&t.mdxType;if("string"===typeof e||o){var a=n.length,i=new Array(a);i[0]=p;var s={};for(var u in t)hasOwnProperty.call(t,u)&&(s[u]=t[u]);s.originalType=e,s.mdxType="string"===typeof e?e:o,i[1]=s;for(var l=2;l<a;l++)i[l]=n[l];return r.createElement.apply(null,i)}return r.createElement.apply(null,n)}p.displayName="MDXCreateElement"},4995:function(e,t,n){"use strict";n.r(t),n.d(t,{default:function(){return o}});var r=n(5893);function o(e){var t=e.children;return(0,r.jsxs)("div",{children:[(0,r.jsx)("a",{href:"/",children:"Return home"}),(0,r.jsx)("div",{className:"blog",children:t})]})}},7886:function(e,t,n){"use strict";n.r(t),n.d(t,{default:function(){return c}});var r=n(159),o=n(219),a=(n(7294),n(3905)),i=n(4995),s=["components"],u={},l=function(e){var t=e.children;return(0,a.kt)(i.default,null,t)};function c(e){var t=e.components,n=(0,o.Z)(e,s);return(0,a.kt)(l,(0,r.Z)({},u,n,{components:t,mdxType:"MDXLayout"}),(0,a.kt)("h1",null,"Motif finding from chip-seq"),(0,a.kt)("p",null,(0,a.kt)("strong",{parentName:"p"},"Introduction")),(0,a.kt)("p",null,"Recently I have been involved in a discussion about MACS on the mailing\nlist, and I wanted to do some basic experiments to look at motif finding\nin ChIP-seq."),(0,a.kt)("p",null,(0,a.kt)("strong",{parentName:"p"},"Methods")),(0,a.kt)("p",null,"I used very basic methods and I have ",(0,a.kt)("a",{parentName:"p",href:"https://gist.github.com/tonto/8312029"},"posted some gist\nfiles")," on GitHub that essentially\nfollows from ",(0,a.kt)("a",{parentName:"p",href:"http://www.bioconductor.org/help/workflows/gene-regulation-tfbs/"},"Bioconductor for motif\nfinding"),".\nI also made up some code for plotting the read coverage of the peaks\nusing a combination of bedtools and R."),(0,a.kt)("p",null,(0,a.kt)("strong",{parentName:"p"},"Results")),(0,a.kt)("p",null,"The results from plotting the coverage are pretty much as expected\n(Figure 1), but I think it opens up some questions when we also look at\nthe motif map (Figure 2). The enrichment of ChIP-seq reads shows a clear\npattern over the peak summits, but there are some areas where either\nmappability was bad (dark stripes) or my algorithm for calculating\ncoverage has some bugs."),(0,a.kt)("p",null,(0,a.kt)("img",{parentName:"p",src:"../../media/72637788100_0.png",alt:null})),(0,a.kt)("p",null,(0,a.kt)("strong",{parentName:"p"},"Figure 1.")," Plot of log read counts in 2000bp windows around the\nsummits of ChIP-seq peaks from a deeply sequenced STE12 transcription\nfactor pulldown in yeast (Zheng et al, 2010)"),(0,a.kt)("p",null,"I also looked at the instances of motifs in the same 2000bp window\naround the summits (Figure 2). The motif finding produced much less\ndefinitive results, and I was actually verycurious about this. There are\nalmost perfect matches to the motif scattered throughout the sequences\nthat don't necessarily correspond to the enrichment plot."),(0,a.kt)("p",null,(0,a.kt)("img",{parentName:"p",src:"../../media/72637788100_1.png",alt:null})),(0,a.kt)("p",null,(0,a.kt)("strong",{parentName:"p"},"Figure 2.")," Motif matches from a very basic motif searcher in the\npackage Biostrings using matchPWM. PWM is downloaded from JASPAR for\nSTE12, which corresponds to the same transcription factor targetted in\nthis ChIP-seq."),(0,a.kt)("p",null,(0,a.kt)("strong",{parentName:"p"},"Discussion")),(0,a.kt)("p",null,"Given the disparity between the datasets, I see a couple possibilities\n(as a nive bioinformatics person, biologists please correct me)"),(0,a.kt)("p",null,"1",")"," We used too basic of a motif finder, and it could not distinguish\nthe motifs clearly and they may actually be clustered near the peak\nsummits"),(0,a.kt)("p",null,"2",")"," Motifs may be enriched in the promoter regions, but more complex\nregulation causes true ChIP-seq binding"),(0,a.kt)("p",null,"3",")"," The motif is simply found very frequently throughout the genome, and\nnot just in the promoter region, and so like the above we need something\nbetter than the basic motif finding that we have to figure out chip-seq\nregulation"),(0,a.kt)("p",null,"4",")"," Edit: there may also be co-regulation (which is very possible given\nthe trans-regulators that were found in Zheng et al.'s results)"),(0,a.kt)("p",null,"For posterity, here is the seqLogo for our JASPAR STE12 PWM"),(0,a.kt)("p",null,(0,a.kt)("img",{parentName:"p",src:"../../media/72637788100_2.png",alt:null})),(0,a.kt)("p",null,(0,a.kt)("strong",{parentName:"p"},"Conclusion")),(0,a.kt)("p",null,"I have a lot of questions about this data, and I am wondering how we can\nuse it to further our knowledge of the behavior of the transcription\nfactors for regulating the genome!"),(0,a.kt)("p",null,"::: {#footer}\n","[ January 8th, 2014 12:57am ]","{#timestamp} ","[chip-seq]","{.tag}\n","[bioinformatics]","{.tag} ","[questions]","{.tag} ","[genomics]","{.tag} ","[fun]","{.tag}\n:::"))}c.isMDXComponent=!0},9019:function(e,t,n){(window.__NEXT_P=window.__NEXT_P||[]).push(["/2014-01-08",function(){return n(7886)}])},159:function(e,t,n){"use strict";function r(){return(r=Object.assign||function(e){for(var t=1;t<arguments.length;t++){var n=arguments[t];for(var r in n)Object.prototype.hasOwnProperty.call(n,r)&&(e[r]=n[r])}return e}).apply(this,arguments)}n.d(t,{Z:function(){return r}})},219:function(e,t,n){"use strict";function r(e,t){if(null==e)return{};var n,r,o=function(e,t){if(null==e)return{};var n,r,o={},a=Object.keys(e);for(r=0;r<a.length;r++)n=a[r],t.indexOf(n)>=0||(o[n]=e[n]);return o}(e,t);if(Object.getOwnPropertySymbols){var a=Object.getOwnPropertySymbols(e);for(r=0;r<a.length;r++)n=a[r],t.indexOf(n)>=0||Object.prototype.propertyIsEnumerable.call(e,n)&&(o[n]=e[n])}return o}n.d(t,{Z:function(){return r}})}},function(e){e.O(0,[9774,2888,179],(function(){return t=9019,e(e.s=t);var t}));var t=e.O();_N_E=t}]);